// @fundamental-ngx/core@0.13.2/esm5/lib/shellbar/public_api.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shellbar/public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { ShellbarModule } from './shellbar.module';
export {} from './model/shellbar-menu-item';
export {} from './model/shellbar-user';
export {} from './model/shellbar-user-menu';
export { ProductMenuComponent } from './product-menu/product-menu.component';
export { ShellbarActionsComponent } from './shellbar-actions/shellbar-actions.component';
export { ShellbarComponent } from './shellbar.component';
export { ShellbarActionComponent } from './shellbar-action/shellbar-action.component';
export { ShellbarLogoComponent } from './shellbar-logo/shellbar-logo.component';
export { ShellbarSubtitleComponent } from './shellbar-subtitle/shellbar-subtitle.component';
export { ShellbarTitleComponent } from './shellbar-title/shellbar-title.component';
export { ShellbarUserMenuComponent } from './user-menu/shellbar-user-menu.component';
