// @fundamental-ngx/core@0.13.2/esm2015/lib/nested-list/nested-list-state.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list-state.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
export class NestedListStateService {
    constructor() {
        /**
         * @hidden
         * The condensed state is modified by the parent and read by nested lists.
         */
        this.condensed = false;
        /**
         * Event, that is thrown always, when the link's selected state is changed, triggers
         */
        this.refresh$ = new Subject();
    }
    /**
     * @hidden
     * Recursive method, that adds selected state to parents, when any of the children is selected
     * @private
     * @param {?} item
     * @return {?}
     */
    selected(item) {
        item.allChildrenItems.forEach((/**
         * @param {?} _item
         * @return {?}
         */
        _item => _item.linkItem.controlSelected = _item.linkItem.selected));
        /** @type {?} */
        const hasChildSelected = !!item.allChildrenItems.find((/**
         * @param {?} _item
         * @return {?}
         */
        _item => (_item.linkItem && _item.linkItem.selected || this.selected(_item))));
        if (hasChildSelected) {
            item.linkItem.controlSelected = hasChildSelected;
        }
        return hasChildSelected;
    }
    /**
     * @hidden
     * Recursive method, that removes all of controlSelected properties.
     * @private
     * @param {?} item
     * @return {?}
     */
    resetSelected(item) {
        if (item.linkItem) {
            item.linkItem.controlSelected = false;
        }
        item.allChildrenItems.forEach((/**
         * @param {?} _item
         * @return {?}
         */
        _item => {
            this.resetSelected(_item);
        }));
    }
    /**
     * Method that adds selected state to elements, depending on children from deeper level
     * @param {?} list
     * @return {?}
     */
    propagateSelected(list) {
        list.nestedItems.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => this.resetSelected(item)));
        list.nestedItems.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => this.selected(item)));
    }
}
if (false) {
    /**
     * @hidden
     * The condensed state is modified by the parent and read by nested lists.
     * @type {?}
     */
    NestedListStateService.prototype.condensed;
    /**
     * Event, that is thrown always, when the link's selected state is changed, triggers
     * @type {?}
     */
    NestedListStateService.prototype.refresh$;
}
