// @fundamental-ngx/core@0.13.2/esm2015/lib/nested-list/nested-list-directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list-directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
export class NestedListHeaderDirective {
    constructor() {
        /**
         * @hidden
         */
        this.fdNestedListHeaderClass = true;
    }
}
NestedListHeaderDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fdNestedDirectivesHeader], [fd-nested-list-header]'
            },] }
];
NestedListHeaderDirective.propDecorators = {
    fdNestedListHeaderClass: [{ type: HostBinding, args: ['class.fd-nested-list__group-header',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListHeaderDirective.prototype.fdNestedListHeaderClass;
}
export class NestedListIconDirective extends AbstractFdNgxClass {
    /**
     * @hidden
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super(elementRef);
        this.elementRef = elementRef;
        /**
         * @hidden
         */
        this.fdNestedListIconClass = true;
    }
    /**
     * @hidden
     * @return {?}
     */
    _setProperties() {
        if (this.glyph) {
            this._addClassToElement('sap-icon--' + this.glyph);
        }
        this._addClassToElement('fd-nested-list__icon');
    }
}
NestedListIconDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fdNestedDirectivesIcon], [fd-nested-list-icon]'
            },] }
];
/** @nocollapse */
NestedListIconDirective.ctorParameters = () => [
    { type: ElementRef }
];
NestedListIconDirective.propDecorators = {
    fdNestedListIconClass: [{ type: HostBinding, args: ['class.fd-nested-list__icon',] }],
    glyph: [{ type: Input }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListIconDirective.prototype.fdNestedListIconClass;
    /**
     * The icon name to display. See the icon page for the list of icons
     * here: https://sap.github.io/fundamental-ngx/icon
     * @type {?}
     */
    NestedListIconDirective.prototype.glyph;
    /**
     * @type {?}
     * @private
     */
    NestedListIconDirective.prototype.elementRef;
}
export class NestedListTitleDirective {
    /**
     * @hidden
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        /**
         * @hidden
         */
        this.fdNestedListTitleClass = true;
    }
    /**
     * Returns element's InnerText
     * @return {?}
     */
    getInnerText() {
        return this.elementRef && this.elementRef.nativeElement.innerText;
    }
}
NestedListTitleDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fdNestedDirectivesTitle], [fd-nested-list-title]'
            },] }
];
/** @nocollapse */
NestedListTitleDirective.ctorParameters = () => [
    { type: ElementRef }
];
NestedListTitleDirective.propDecorators = {
    fdNestedListTitleClass: [{ type: HostBinding, args: ['class.fd-nested-list__title',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListTitleDirective.prototype.fdNestedListTitleClass;
    /**
     * @type {?}
     * @private
     */
    NestedListTitleDirective.prototype.elementRef;
}
