// @fundamental-ngx/core@0.13.2/esm2015/lib/product-switch/product-switch/product-switch.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/product-switch/product-switch/product-switch.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { PopoverComponent } from '../../popover/popover.component';
export class ProductSwitchComponent extends PopoverComponent {
}
ProductSwitchComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-product-switch',
                template: "<div class=\"fd-product-switch\">\n    <fd-popover [options]=\"{placement: 'bottom-end'}\"\n                [closeOnEscapeKey]=\"closeOnEscapeKey\"\n                [closeOnOutsideClick]=\"closeOnOutsideClick\"\n                [(isOpen)]=\"isOpen\"\n                (isOpenChange)=\"openChanged($event)\"\n                [disabled]=\"disabled\"\n                [triggers]=\"triggers\"\n                [isDropdown]=\"isDropdown\"\n                [focusTrapped]=\"false\">\n        <fd-popover-control>\n            <button fd-button class=\"fd-product-switch__control\" [glyph]=\"'grid'\" [options]=\"'light'\"></button>\n        </fd-popover-control>\n        <fd-popover-body>\n            <ng-content></ng-content>\n        </fd-popover-body>\n    </fd-popover>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            }] }
];
