// @fundamental-ngx/core@0.13.2/esm5/lib/button-group/button-grouped.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/button-group/button-grouped.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
/**
 * Directive to be applied to buttons that are members of a button group.
 *
 * ```html
 * <button fd-button-grouped>Button</button>
 * ```
 */
var ButtonGroupedDirective = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonGroupedDirective, _super);
    /** @hidden */
    function ButtonGroupedDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Defines if there will be added fd-button class. Enabled by default.
         */
        _this.fdButtonClass = true;
        /**
         * Whether the button should be in compact form.
         */
        _this.compact = false;
        /**
         * @hidden
         */
        _this.fdButtonGroupedClass = true;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ButtonGroupedDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-button--grouped');
        if (this.fdButtonClass) {
            this._addClassToElement('fd-button');
        }
        if (this.size) {
            this._addClassToElement('fd-button--' + this.size);
        }
        if (this.glyph) {
            this._addClassToElement('sap-icon--' + this.glyph);
        }
        if (this.state) {
            this._addClassToElement('is-' + this.state);
        }
    };
    ButtonGroupedDirective.decorators = [
        { type: Directive, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-button-grouped]'
                },] }
    ];
    /** @nocollapse */
    ButtonGroupedDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ButtonGroupedDirective.propDecorators = {
        size: [{ type: Input }],
        fdButtonClass: [{ type: Input }],
        glyph: [{ type: Input }],
        state: [{ type: Input }],
        compact: [{ type: Input }, { type: HostBinding, args: ['class.fd-button--compact',] }],
        fdButtonGroupedClass: [{ type: HostBinding, args: ['class.fd-button--grouped',] }]
    };
    return ButtonGroupedDirective;
}(AbstractFdNgxClass));
export { ButtonGroupedDirective };
if (false) {
    /**
     * @deprecated
     * Will be removed in 0.13.0
     * @type {?}
     */
    ButtonGroupedDirective.prototype.size;
    /**
     * Defines if there will be added fd-button class. Enabled by default.
     * @type {?}
     */
    ButtonGroupedDirective.prototype.fdButtonClass;
    /**
     * Glyph (icon) of the button.
     * @type {?}
     */
    ButtonGroupedDirective.prototype.glyph;
    /**
     * State of the button. Can be `selected` or `disabled`.
     * @type {?}
     */
    ButtonGroupedDirective.prototype.state;
    /**
     * Whether the button should be in compact form.
     * @type {?}
     */
    ButtonGroupedDirective.prototype.compact;
    /**
     * @hidden
     * @type {?}
     */
    ButtonGroupedDirective.prototype.fdButtonGroupedClass;
    /**
     * @type {?}
     * @private
     */
    ButtonGroupedDirective.prototype.elementRef;
}
