// @fundamental-ngx/core@0.13.2/esm2015/lib/notification/notification-footer/notification-footer.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-footer/notification-footer.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
export class NotificationFooterComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdNotificationFooterClass = true;
    }
}
NotificationFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-notification-footer',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            }] }
];
NotificationFooterComponent.propDecorators = {
    fdNotificationFooterClass: [{ type: HostBinding, args: ['class.fd-notification__footer',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationFooterComponent.prototype.fdNotificationFooterClass;
}
