// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification-group/notification-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-group/notification-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
var NotificationGroupComponent = /** @class */ (function () {
    function NotificationGroupComponent() {
        /**
         * @hidden
         */
        this.fdNotificationGroupClass = true;
    }
    NotificationGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification-group',
                    template: "<ng-content select=\"fd-notification\"></ng-content>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [""]
                }] }
    ];
    NotificationGroupComponent.propDecorators = {
        fdNotificationGroupClass: [{ type: HostBinding, args: ['class.fd-notification--group',] }]
    };
    return NotificationGroupComponent;
}());
export { NotificationGroupComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationGroupComponent.prototype.fdNotificationGroupClass;
}
