// @microsoft/rush-lib@5.62.4/lib/logic/operations/NullOperationRunner.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullOperationRunner = void 0;
/**
 * Implementation of `IOperationRunner` for operations that require no work, such as empty scripts,
 * or operations that are skipped by filters.
 */
class NullOperationRunner {
    constructor(name, result) {
        this.hadEmptyScript = true;
        // The operation may never be skipped; it doesn't do anything anyway
        this.isSkipAllowed = false;
        // The operation is a no-op, so skip writing an empty cache entry
        this.isCacheWriteAllowed = false;
        // Nothing will get logged, no point allowing warnings
        this.warningsAreAllowed = false;
        this.name = name;
        this._result = result;
    }
    async executeAsync(context) {
        return this._result;
    }
}
exports.NullOperationRunner = NullOperationRunner;
