// @microsoft/rush-lib@5.62.4/lib/logic/operations/Operation.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operation = void 0;
/**
 * The `Operation` class is a node in the dependency graph of work that needs to be scheduled by the
 * `OperationExecutionManager`. Each `Operation` has a `runner` member of type `IOperationRunner`, whose
 * implementation manages the actual process of running a single operation.
 */
class Operation {
    constructor(runner, initialStatus) {
        /**
         * A set of all dependencies which must be executed before this operation is complete.
         * When dependencies finish execution, they are removed from this list.
         */
        this.dependencies = new Set();
        /**
         * The inverse of dependencies, lists all projects which are directly dependent on this one.
         */
        this.dependents = new Set();
        this.runner = runner;
        this.status = initialStatus;
    }
    get name() {
        return this.runner.name;
    }
}
exports.Operation = Operation;
