// @fundamental-ngx/core@0.13.2/esm2015/lib/notification/notification-utils/notification-container.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-utils/notification-container.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
export class NotificationContainer {
}
NotificationContainer.decorators = [
    { type: Component, args: [{
                selector: 'fd-notification-container',
                template: ``,
                host: {
                    '[class.fd-notification-container]': 'true'
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [`
        .fd-notification-container {
            position: fixed;
            display: flex;
            flex-direction: column;
            z-index: 5000;
            align-items: center;
            top: 1rem;
            right: 1rem;
        }
    `]
            }] }
];
