// @fundamental-ngx/core@0.13.2/esm5/lib/image/image.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/image/image.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * The component that represents an image.
 *
 * ```html
 * <fd-image style="margin-right: 10px;" [size]="'l'" [circle]="true" [photo]="'https://placeimg.com/400/400/nature'"></fd-image>
 * ```
 */
var ImageComponent = /** @class */ (function () {
    function ImageComponent() {
        /**
         * The size of the image.
         * The predefined values for the size are *s*, *m*, and *l*.
         */
        this.size = 'm';
        /**
         * Whether to render a circle style for the image.
         */
        this.circle = false;
        /**
         * The image label.
         */
        this.label = 'Image label';
    }
    ImageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-image',
                    template: "<span\n  [ngClass]='(size ? \"fd-image--\" + size : \"\") + (circle ? \" fd-image--circle\" : \"\") '\n  [attr.aria-label]='label'\n  [ngStyle]= \"{'background-image': 'url(' + photo + ')'}\">\n</span>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-image+((--circle), (--s|--m|--l))\n*/[class*=fd-image]{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:inline-block;vertical-align:middle;background-repeat:no-repeat;background-size:cover;background-position:50%;border-radius:4px}[class*=fd-image]:after,[class*=fd-image]:before{box-sizing:inherit;font-size:inherit}.fd-image--s{width:24px;height:24px;max-width:24px;max-height:24px;min-width:24px;min-height:24px}.fd-image--m{width:36px;height:36px;max-width:36px;max-height:36px;min-width:36px;min-height:36px}.fd-image--l{width:48px;height:48px;max-width:48px;max-height:48px;min-width:48px;min-height:48px}.fd-image--circle{border-radius:50%}"]
                }] }
    ];
    ImageComponent.propDecorators = {
        size: [{ type: Input }],
        circle: [{ type: Input }],
        label: [{ type: Input }],
        photo: [{ type: Input }]
    };
    return ImageComponent;
}());
export { ImageComponent };
if (false) {
    /**
     * The size of the image.
     * The predefined values for the size are *s*, *m*, and *l*.
     * @type {?}
     */
    ImageComponent.prototype.size;
    /**
     * Whether to render a circle style for the image.
     * @type {?}
     */
    ImageComponent.prototype.circle;
    /**
     * The image label.
     * @type {?}
     */
    ImageComponent.prototype.label;
    /**
     * The image url.
     * @type {?}
     */
    ImageComponent.prototype.photo;
}
