// @fundamental-ngx/core@0.13.2/esm5/lib/menu/menu-item.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/menu/menu-item.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding } from '@angular/core';
/**
 * The directive that represents a menu item.
 */
var MenuItemDirective = /** @class */ (function () {
    /** @hidden */
    function MenuItemDirective(itemEl) {
        this.itemEl = itemEl;
        /**
         * @hidden
         */
        this.fdMenuItemClass = true;
    }
    /**
     * @return {?}
     */
    MenuItemDirective.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.focus();
    };
    /**
     * @return {?}
     */
    MenuItemDirective.prototype.click = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.click();
    };
    MenuItemDirective.decorators = [
        { type: Directive, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-menu-item]',
                },] }
    ];
    /** @nocollapse */
    MenuItemDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    MenuItemDirective.propDecorators = {
        fdMenuItemClass: [{ type: HostBinding, args: ['class.fd-menu__item',] }]
    };
    return MenuItemDirective;
}());
export { MenuItemDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MenuItemDirective.prototype.fdMenuItemClass;
    /** @type {?} */
    MenuItemDirective.prototype.itemEl;
}
