// @fundamental-ngx/core@0.13.2/esm5/lib/menu/menu-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/menu/menu-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * The component that represents a menu group.
 */
var MenuGroupComponent = /** @class */ (function () {
    function MenuGroupComponent() {
        /**
         * @hidden
         */
        this.fdHasDisplayBlockClass = true;
        /**
         * @hidden
         */
        this.fdMenuGroupClass = true;
    }
    MenuGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-menu-group',
                    template: "<ng-content select=\"h1.fd-menu-title\"></ng-content>\n<ng-content select=\"ul.fd-menu-list\"></ng-content>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    MenuGroupComponent.propDecorators = {
        fdHasDisplayBlockClass: [{ type: HostBinding, args: ['class.fd-has-display-block',] }],
        fdMenuGroupClass: [{ type: HostBinding, args: ['class.fd-menu__group',] }]
    };
    return MenuGroupComponent;
}());
export { MenuGroupComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MenuGroupComponent.prototype.fdHasDisplayBlockClass;
    /**
     * @hidden
     * @type {?}
     */
    MenuGroupComponent.prototype.fdMenuGroupClass;
}
