// @fundamental-ngx/core@0.13.2/esm2015/lib/input-group/input-group-number.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/input-group/input-group-number.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, forwardRef, Input, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * The component that represents an integer value input.
 * The value is increased or decreased using the spinner add-on.
 *
 * ```html
 * <fd-input-group-number [disabled]="false" [(ngModel)]="numberValue"></fd-input-group-number>
 * ```
 */
export class InputGroupNumberComponent {
    /**
     * @hidden
     * @param {?} changeDetectorRef
     */
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        /**
         * Aria label for the 'step up' button.
         */
        this.stepUpLabel = 'Step up';
        /**
         * Aria label for the 'step down' button.
         */
        this.stepDownLabel = 'Step down';
        /**
         * @hidden
         */
        this.onChange = (/**
         * @return {?}
         */
        () => { });
        /**
         * @hidden
         */
        this.onTouched = (/**
         * @return {?}
         */
        () => { });
    }
    /**
     * Get the value of the text input.
     * @return {?}
     */
    get inputText() {
        return this.inputTextValue;
    }
    /**
     * Set the value of the text input.
     * @param {?} value
     * @return {?}
     */
    set inputText(value) {
        this.inputTextValue = value;
        this.onChange(value);
        this.onTouched();
    }
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.inputTextValue = value;
        this.changeDetectorRef.markForCheck();
    }
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @hidden
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
        this.changeDetectorRef.markForCheck();
    }
    /**
     * @hidden
     * @return {?}
     */
    stepUpClicked() {
        this.inputTextValue++;
        this.onChange(this.inputTextValue);
        this.onTouched();
    }
    /**
     * @hidden
     * @return {?}
     */
    stepDownClicked() {
        this.inputTextValue--;
        this.onChange(this.inputTextValue);
        this.onTouched();
    }
}
InputGroupNumberComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-input-group-number',
                template: "<div class=\"fd-input-group fd-input-group--after\">\n    <input type=\"number\"\n           class=\"fd-input fd-input-group__input fd-input--no-number-spinner\"\n           [disabled]=\"disabled\"\n           [(ngModel)]=\"inputText\"\n           placeholder=\"{{placeholder}}\"/>\n    <span fd-input-group-addon [button]=\"true\" [placement]=\"'after'\">\n        <button fd-button\n                [disabled]=\"disabled\"\n                [glyph]=\"'slim-arrow-up'\"\n                [fdType]=\"'half'\"\n                [options]=\"'light'\"\n                class=\"fd-input-group__button\"\n                [attr.aria-label]=\"stepUpLabel\"\n                (click)=\"stepUpClicked()\">\n        </button>\n        <button fd-button\n                [disabled]=\"disabled\"\n                [glyph]=\"'slim-arrow-down'\"\n                [fdType]=\"'half'\"\n                [options]=\"'light'\"\n                class=\"fd-input-group__button\"\n                [attr.aria-label]=\"stepDownLabel\"\n                (click)=\"stepDownClicked()\">\n        </button>\n    </span>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => InputGroupNumberComponent)),
                        multi: true
                    }
                ],
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-input-group+(--inline)\n    .fd-input-group__addon+(--button)\n*/.fd-input-group{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:.0625rem solid #89919a;display:flex;border-width:var(--sapField_BorderWidth,.0625rem);border-color:var(--sapField_BorderColor,#89919a);border-radius:.125rem;border-radius:var(--sapField_BorderCornerRadius,.125rem);background-color:#fff;background-color:var(--sapField_Background,#fff);width:100%}.fd-input-group:after,.fd-input-group:before{box-sizing:inherit;font-size:inherit}.fd-input-group .is-hover,.fd-input-group:hover{background-color:#fff;background-color:var(--sapField_Hover_Background,#fff);border-color:#0854a0;border-color:var(--sapField_Hover_BorderColor,#0854a0)}.fd-input-group.is-valid{background-color:#fff;background-color:var(--sapField_SuccessBackground,#fff);border-color:#107e3e;border-color:var(--sapField_SuccessColor,#107e3e);border-width:.0625rem}.fd-input-group.is-valid .is-hover,.fd-input-group.is-valid:hover{background-color:#fff;background-color:var(--sapField_SuccessBackground,#fff);border-color:#107e3e;border-color:var(--sapField_SuccessColor,#107e3e)}.fd-input-group.is-invalid,.fd-input-group.is-invalid .is-hover,.fd-input-group.is-invalid:hover{background-color:#fff;background-color:var(--sapField_InvalidBackground,#fff);border-color:#b00;border-color:var(--sapField_InvalidColor,#b00)}.fd-input-group.is-alert,.fd-input-group.is-alert .is-hover,.fd-input-group.is-alert:hover,.fd-input-group.is-warning,.fd-input-group.is-warning .is-hover,.fd-input-group.is-warning:hover{background-color:#fff;background-color:var(--sapField_WarningBackground,#fff);border-color:#e9730c;border-color:var(--sapField_WarningColor,#e9730c)}.fd-input-group.is-information,.fd-input-group.is-information .is-hover,.fd-input-group.is-information:hover{background-color:#fff;background-color:var(--sapField_InformationBackground,#fff);border-color:#0a6ed1;border-color:var(--sapField_InformationColor,#0a6ed1)}.fd-input-group.is-focus,.fd-input-group:focus{outline-width:var(--sapContent_FocusWidth,.0625rem);outline-color:var(--sapContent_FocusColor,#000);outline-style:var(--sapContent_FocusStyle,dotted);box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;box-shadow:none;outline-offset:-.1875rem;outline:#000 dotted .0625rem;outline:var(--sapContent_FocusWidth,.0625rem) var(--sapContent_FocusStyle,dotted) var(--sapContent_FocusColor,#000)}.fd-input-group.is-alert,.fd-input-group.is-information,.fd-input-group.is-invalid,.fd-input-group.is-warning{border-width:.125rem}.fd-input-group.is-alert.is-focus,.fd-input-group.is-alert:focus,.fd-input-group.is-information.is-focus,.fd-input-group.is-information:focus,.fd-input-group.is-invalid.is-focus,.fd-input-group.is-invalid:focus,.fd-input-group.is-warning.is-focus,.fd-input-group.is-warning:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;box-shadow:none;outline-offset:-.25rem}.fd-input-group.is-disabled,.fd-input-group:disabled,.fd-input-group[aria-disabled=true]{pointer-events:none;opacity:.4}.fd-input-group.is-disabled::-webkit-input-placeholder,.fd-input-group:disabled::-webkit-input-placeholder,.fd-input-group[aria-disabled=true]::-webkit-input-placeholder{color:transparent}.fd-input-group.is-disabled::-moz-placeholder,.fd-input-group:disabled::-moz-placeholder,.fd-input-group[aria-disabled=true]::-moz-placeholder{color:transparent}.fd-input-group.is-disabled:-ms-input-placeholder,.fd-input-group:disabled:-ms-input-placeholder,.fd-input-group[aria-disabled=true]:-ms-input-placeholder{color:transparent}.fd-input-group.is-disabled::-ms-input-placeholder,.fd-input-group:disabled::-ms-input-placeholder,.fd-input-group[aria-disabled=true]::-ms-input-placeholder{color:transparent}.fd-input-group.is-disabled::placeholder,.fd-input-group:disabled::placeholder,.fd-input-group[aria-disabled=true]::placeholder{color:transparent}.fd-input-group.is-readonly,.fd-input-group[readonly]{background-color:hsla(0,0%,94.9%,.5);background-color:var(--sapField_ReadOnly_Background,hsla(0,0%,94.9%,.5));border-color:#89919a;border-color:var(--sapField_ReadOnly_BorderColor,#89919a)}.fd-input-group.is-readonly::-webkit-input-placeholder,.fd-input-group[readonly]::-webkit-input-placeholder{color:transparent}.fd-input-group.is-readonly::-moz-placeholder,.fd-input-group[readonly]::-moz-placeholder{color:transparent}.fd-input-group.is-readonly:-ms-input-placeholder,.fd-input-group[readonly]:-ms-input-placeholder{color:transparent}.fd-input-group.is-readonly::-ms-input-placeholder,.fd-input-group[readonly]::-ms-input-placeholder{color:transparent}.fd-input-group.is-readonly::placeholder,.fd-input-group[readonly]::placeholder{color:transparent}.fd-input-group>*{border-radius:.125rem;border-radius:var(--sapField_BorderCornerRadius,.125rem)}.fd-input-group>:first-child{border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group>:last-child{border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group>:not(:first-child):not(:last-child){border-radius:0}.fd-input-group>[dir=rtl],[dir=rtl] .fd-input-group>*{border-radius:.125rem;border-radius:var(--sapField_BorderCornerRadius,.125rem)}.fd-input-group>[dir=rtl]:first-child,[dir=rtl] .fd-input-group>:first-child{border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group>[dir=rtl]:last-child,[dir=rtl] .fd-input-group>:last-child{border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group .fd-input-group__input{background-color:transparent;border:none;flex:1;padding-right:.25rem;padding-left:.25rem}.fd-input-group .fd-input-group__input:nth-child(n+1){margin:0}.fd-input-group .fd-input-group__input.is-focus,.fd-input-group .fd-input-group__input:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;outline:0;box-shadow:none}.fd-input-group .fd-input-group__input.is-disabled,.fd-input-group .fd-input-group__input:disabled,.fd-input-group .fd-input-group__input[aria-disabled=true]{pointer-events:none;opacity:.4}.fd-input-group .fd-input-group__input:first-child{padding-left:.625rem;padding-right:.25rem}.fd-input-group .fd-input-group__input:first-child[dir=rtl],[dir=rtl] .fd-input-group .fd-input-group__input:first-child{padding-left:.25rem;padding-right:.625rem}.fd-input-group .fd-input-group__input:last-child{padding-left:.25rem;padding-right:.625rem}.fd-input-group .fd-input-group__input:last-child[dir=rtl],[dir=rtl] .fd-input-group .fd-input-group__input:last-child{padding-left:.625rem;padding-right:.25rem}@media (-ms-high-contrast:active),(-ms-high-contrast:none){.fd-input-group .fd-input-group__input{min-width:160px}}.fd-input-group--inline{display:inline-flex;vertical-align:bottom;width:auto}.fd-input-group__addon{font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 .625rem;margin:0;border:0;display:flex;justify-content:center;align-items:center;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;color:#6a6d70;color:var(--sapContent_NonInteractiveIconColor,#6a6d70);background-color:#fff;background-color:var(--sapField_Background,#fff);font-size:1rem;font-size:var(--sapFontLargeSize,1rem);min-width:2.25rem;min-height:2.25rem}.fd-input-group__addon:after,.fd-input-group__addon:before{box-sizing:inherit;font-size:inherit}.fd-input-group__addon span[class*=sap-icon--]:before{font-size:1rem;font-size:var(--sapFontLargeSize,1rem)}.fd-input-group__addon--compact{min-width:2rem;min-height:1.625rem}.fd-input-group__addon--readonly{opacity:.4}.fd-input-group__addon--button{padding:0;display:flex;flex-direction:column;border:none;border-radius:.0625rem}.fd-input-group__addon--button .fd-input-group__button{border-radius:0;height:100%;max-height:100%}.fd-input-group__addon--button .fd-input-group__button:nth-child(n+1){margin:0}.fd-input-group__addon--button>*{border-radius:0}.fd-input-group__addon--button:first-child{border-radius:.0625rem 0 0 .0625rem}.fd-input-group__addon--button:last-child{border-radius:0 .0625rem .0625rem 0}.fd-input-group__addon--button:not(:first-child):not(:last-child) .fd-input-group__button{border-radius:0}.fd-input-group__addon--button[dir=rtl]:first-child,[dir=rtl] .fd-input-group__addon--button:first-child{border-radius:0 .0625rem .0625rem 0;margin-right:0}.fd-input-group__addon--button[dir=rtl]:last-child,[dir=rtl] .fd-input-group__addon--button:last-child{border-radius:.0625rem 0 0 .0625rem;margin-left:0}.fd-input-group__button{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;border:none;width:100%}.fd-input-group .fd-textarea{resize:vertical}fd-input-group .fd-input-group__addon--button.fd-input-group__addon--textarea>*{max-height:100%;height:100%}fd-input-group .fd-input-group__addon--button .fd-input-group__button:hover{border-color:transparent}fd-input-group .fd-input-group__input{width:100%}"]
            }] }
];
/** @nocollapse */
InputGroupNumberComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
InputGroupNumberComponent.propDecorators = {
    disabled: [{ type: Input }],
    placeholder: [{ type: Input }],
    stepUpLabel: [{ type: Input }],
    stepDownLabel: [{ type: Input }]
};
if (false) {
    /**
     * Whether the input is disabled.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.disabled;
    /**
     * Placeholder for the input field.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.placeholder;
    /**
     * Aria label for the 'step up' button.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.stepUpLabel;
    /**
     * Aria label for the 'step down' button.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.stepDownLabel;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.inputTextValue;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.onChange;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    InputGroupNumberComponent.prototype.changeDetectorRef;
}
