// @fundamental-ngx/core@0.13.2/esm5/lib/alert/alert-utils/alert-container.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/alert/alert-utils/alert-container.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
import { alertContainerNgIf } from './alert-animations';
var AlertContainerComponent = /** @class */ (function () {
    function AlertContainerComponent() {
        /**
         * @hidden
         */
        this.fdAlertContainerClass = true;
    }
    AlertContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-alert-container',
                    template: "",
                    host: {
                        '[@alertContainerNgIf]': ''
                    },
                    animations: [
                        alertContainerNgIf
                    ],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        .fd-alert-container {\n            position: fixed;\n            display: flex;\n            flex-direction: column;\n            z-index: 5000;\n            align-items: center;\n            top: 0;\n            right: 50%;\n            left: 50%;\n        }\n    "]
                }] }
    ];
    AlertContainerComponent.propDecorators = {
        fdAlertContainerClass: [{ type: HostBinding, args: ['class.fd-alert-container',] }]
    };
    return AlertContainerComponent;
}());
export { AlertContainerComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    AlertContainerComponent.prototype.fdAlertContainerClass;
}
