// @fundamental-ngx/core@0.13.2/esm5/lib/alert/alert-utils/alert-animations.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/alert/alert-utils/alert-animations.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { animate, style, transition, trigger } from '@angular/animations';
/** @type {?} */
export var alertFadeNgIf = trigger('fadeAlertNgIf', [
    transition(':enter', [
        style({
            opacity: 0
        }),
        animate('250ms ease-in-out', style({
            opacity: 1
        }))
    ]),
    transition(':leave', [
        style({
            opacity: 1,
            marginTop: '*',
            paddingTop: '*',
            paddingBottom: '*',
            height: '*',
            overflow: 'hidden'
        }),
        animate('400ms ease-in-out', style({
            opacity: 0,
            marginTop: 0,
            paddingTop: 0,
            paddingBottom: 0,
            height: 0,
            overflow: 'hidden'
        }))
    ])
]);
/** @type {?} */
export var alertContainerNgIf = trigger('alertContainerNgIf', [
    transition(':leave', [
        style({ opacity: 1 }),
        animate('400ms ease-in-out', style({ opacity: 0 }))
    ])
]);
