// @fundamental-ngx/core@0.13.2/esm2015/lib/mega-menu/mega-menu-subitem.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mega-menu/mega-menu-subitem.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChild, Directive, EventEmitter, HostBinding, HostListener, Output } from '@angular/core';
import { MegaMenuSublinkDirective } from './mega-menu-sublink.directive';
/**
 *  Directive represents mega menu subitem, which can contain sublink.
 *  ```html
 * <li fd-mega-menu-subitem>
 *      <a fd-mega-menu-sublink>Sub Item 2</a>
 * </li>
 *  ```
 *
 */
export class MegaMenuSubitemDirective {
    constructor() {
        /**
         * @hidden
         */
        this.fdMegaMenuClass = true;
        /**
         *
         */
        this.keyDown = new EventEmitter();
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    handleKeyboardEvent(event) {
        this.keyDown.emit(event);
    }
    /**
     * @hidden
     * @return {?}
     */
    focus() {
        this.link.focus();
    }
    /**
     * @hidden
     * @return {?}
     */
    click() {
        this.link.click();
    }
}
MegaMenuSubitemDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: '[fd-mega-menu-subitem]'
            },] }
];
MegaMenuSubitemDirective.propDecorators = {
    fdMegaMenuClass: [{ type: HostBinding, args: ['class.fd-mega-menu__subitem',] }],
    link: [{ type: ContentChild, args: [MegaMenuSublinkDirective, { static: false },] }],
    keyDown: [{ type: Output }],
    handleKeyboardEvent: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuSubitemDirective.prototype.fdMegaMenuClass;
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuSubitemDirective.prototype.link;
    /**
     *
     * @type {?}
     */
    MegaMenuSubitemDirective.prototype.keyDown;
}
