// @fundamental-ngx/core@0.13.2/esm2015/lib/mega-menu/mega-menu-sublink.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mega-menu/mega-menu-sublink.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding } from '@angular/core';
/**
 *  Directive represents mega menu sub link.
 *  ```html
 * <a fd-mega-menu-sublink href="#">Link</a>
 *  ```
 *
 */
export class MegaMenuSublinkDirective {
    /**
     * @hidden
     * @param {?} itemEl
     */
    constructor(itemEl) {
        this.itemEl = itemEl;
        /**
         * @hidden
         */
        this.fdMegaMenuClass = true;
    }
    /**
     * @return {?}
     */
    focus() {
        this.itemEl.nativeElement.focus();
    }
    /**
     * @return {?}
     */
    click() {
        this.itemEl.nativeElement.click();
    }
}
MegaMenuSublinkDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: '[fd-mega-menu-sublink]',
                host: {
                    'tabindex': '0'
                }
            },] }
];
/** @nocollapse */
MegaMenuSublinkDirective.ctorParameters = () => [
    { type: ElementRef }
];
MegaMenuSublinkDirective.propDecorators = {
    fdMegaMenuClass: [{ type: HostBinding, args: ['class.fd-mega-menu__sublink',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuSublinkDirective.prototype.fdMegaMenuClass;
    /** @type {?} */
    MegaMenuSublinkDirective.prototype.itemEl;
}
