// @fundamental-ngx/core@0.13.2/esm5/lib/badge-label/label/label.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/badge-label/label/label.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
/**
 * Label component, used to indicate status, without any background or border
 * Colors, generally in combination with text, are used to easily highlight the state of an object.
 */
var LabelComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LabelComponent, _super);
    /** @hidden */
    function LabelComponent(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    LabelComponent.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-label');
        if (this.status) {
            this._addClassToElement('fd-label--' + this.status);
        }
    };
    LabelComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[fd-label]',
                    template: "<ng-content></ng-content>",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-label+(--success | --warning | --error)\n*/.fd-label{font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 8px;margin:0;border:0;font-size:14px;font-weight:400;color:#0a6ed1;color:var(--fd-color-action,#0a6ed1);line-height:24px;vertical-align:middle;display:inline-block}.fd-label:after,.fd-label:before{box-sizing:inherit;font-size:inherit}.fd-label--success{color:#0a7e3e;color:var(--fd-color-positive,#0a7e3e)}.fd-label--warning{color:#e9730c;color:var(--fd-color-alert,#e9730c)}.fd-label--error{color:#b00;color:var(--fd-color-negative,#b00)}"]
                }] }
    ];
    /** @nocollapse */
    LabelComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    LabelComponent.propDecorators = {
        status: [{ type: Input }]
    };
    return LabelComponent;
}(AbstractFdNgxClass));
export { LabelComponent };
if (false) {
    /**
     * Color coded status for the label. Options are 'success', 'warning', and 'error'. Leave empty for default label.
     * @type {?}
     */
    LabelComponent.prototype.status;
    /**
     * @type {?}
     * @private
     */
    LabelComponent.prototype.elementRef;
}
