// @fundamental-ngx/core@0.13.2/esm2015/lib/badge-label/badge-label/badge.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/badge-label/badge-label/badge.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Input, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
/**
 * Badge component, used to indicate status.
 * Colors, generally in combination with text, are used to easily highlight the state of an object.
 */
export class BadgeComponent extends AbstractFdNgxClass {
    /**
     * @hidden
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super(elementRef);
        this.elementRef = elementRef;
        /**
         * @hidden
         */
        this.fdBadgeClass = true;
    }
    /**
     * @hidden
     * @return {?}
     */
    _setProperties() {
        if (this.status) {
            this._addClassToElement('fd-badge--' + this.status);
        }
        if (this.modifier) {
            this._addClassToElement('fd-badge--' + this.modifier);
        }
    }
}
BadgeComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: '[fd-badge]',
                template: `<ng-content></ng-content>`,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-badge+(( --success | --warning | --error ), (--pill | --filled))\n*/.fd-badge{font-size:var(--sapFontSize,.875rem);color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:1px 8px;margin:0;border:1px solid #0a6ed1;font-size:12px;font-weight:400;border-color:var(--fd-color-action,#0a6ed1);background-color:#f5faff;background-color:var(--fd-color-background-6,#f5faff);line-height:16px;vertical-align:middle;display:inline-block;border-radius:4px}.fd-badge:after,.fd-badge:before{box-sizing:inherit;font-size:inherit}.fd-badge--pill{border-radius:12px}.fd-badge--filled{font-weight:700;color:#fff;color:var(--fd-color-text-5,#fff);background-color:#0a6ed1;background-color:var(--fd-color-action,#0a6ed1);border:none}.fd-badge--filled.fd-badge--success{background-color:#0a7e3e;background-color:var(--fd-color-positive,#0a7e3e)}.fd-badge--filled.fd-badge--warning{background-color:#e9730c;background-color:var(--fd-color-alert,#e9730c)}.fd-badge--filled.fd-badge--error{background-color:#b00;background-color:var(--fd-color-negative,#b00)}.fd-badge--success{background-color:#f1fdf6;background-color:var(--fd-color-background-3,#f1fdf6);border-color:#0a7e3e;border-color:var(--fd-color-positive,#0a7e3e)}.fd-badge--warning{background-color:#fef7f1;background-color:var(--fd-color-background-4,#fef7f1);border-color:#e9730c;border-color:var(--fd-color-alert,#e9730c)}.fd-badge--error{background-color:#ffebeb;background-color:var(--fd-color-background-5,#ffebeb);border-color:#b00;border-color:var(--fd-color-negative,#b00)}"]
            }] }
];
/** @nocollapse */
BadgeComponent.ctorParameters = () => [
    { type: ElementRef }
];
BadgeComponent.propDecorators = {
    status: [{ type: Input }],
    modifier: [{ type: Input }],
    fdBadgeClass: [{ type: HostBinding, args: ['class.fd-badge',] }]
};
if (false) {
    /**
     * Color coded status for the badge. Options are 'success', 'warning', and 'error'. Leave empty for default badge.
     * @type {?}
     */
    BadgeComponent.prototype.status;
    /**
     * Modifier for the badge. Options are 'pill' and 'filled'. Leave empty for normal.
     * @type {?}
     */
    BadgeComponent.prototype.modifier;
    /**
     * @hidden
     * @type {?}
     */
    BadgeComponent.prototype.fdBadgeClass;
    /**
     * @type {?}
     * @private
     */
    BadgeComponent.prototype.elementRef;
}
