// @fundamental-ngx/core@0.13.2/esm5/lib/date-picker/format/date-parser.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-picker/format/date-parser.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FdDate } from '../../calendar/models/fd-date';
import * as i0 from "@angular/core";
/**
 * @return {?}
 */
export function DATE_FORMAT_FACTORY() {
    return new DateFormatParserDefault();
}
/**
 * Abstract class which defines the behaviour of the date format and parser.
 * @abstract
 */
var DateFormatParser = /** @class */ (function () {
    function DateFormatParser() {
        /**
         * Delimiter for the range. This should not show up in the string representation of the dates.
         */
        this.rangeDelimiter = ' - ';
    }
    DateFormatParser.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                    useFactory: DATE_FORMAT_FACTORY
                },] }
    ];
    /** @nocollapse */ DateFormatParser.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: DATE_FORMAT_FACTORY, token: DateFormatParser, providedIn: "root" });
    return DateFormatParser;
}());
export { DateFormatParser };
if (false) {
    /**
     * Delimiter for the range. This should not show up in the string representation of the dates.
     * @type {?}
     */
    DateFormatParser.prototype.rangeDelimiter;
    /**
     * Should take in a string value and return a FdDate model object.
     * @abstract
     * @param {?} value String to concert to a FdDate model object.
     * @return {?}
     */
    DateFormatParser.prototype.parse = function (value) { };
    /**
     * Should take in a FdDate model object and return a string representation.
     * @abstract
     * @param {?} date FdDate to format to string value.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    DateFormatParser.prototype.format = function (date) { };
}
/**
 * Default implementation of the DateFormatParser service.
 */
var DateFormatParserDefault = /** @class */ (function (_super) {
    tslib_1.__extends(DateFormatParserDefault, _super);
    function DateFormatParserDefault() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Takes in a string value and return a FdDate model object.
     * @param value String to concert to a FdDate model object.
     */
    /**
     * Takes in a string value and return a FdDate model object.
     * @param {?} value String to concert to a FdDate model object.
     * @return {?}
     */
    DateFormatParserDefault.prototype.parse = /**
     * Takes in a string value and return a FdDate model object.
     * @param {?} value String to concert to a FdDate model object.
     * @return {?}
     */
    function (value) {
        if (value) {
            /** @type {?} */
            var date = new Date(value);
            return new FdDate(date.getFullYear(), date.getMonth() + 1, date.getDate());
        }
        else {
            return new FdDate(null, null, null);
        }
    };
    /**
     * Takes in a FdDate model object and return a string representation.
     * @param date FdDate to format to string value.
     * Return null, to keep default angular DatePipe as a formatter.
     */
    /**
     * Takes in a FdDate model object and return a string representation.
     * @param {?} date FdDate to format to string value.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    DateFormatParserDefault.prototype.format = /**
     * Takes in a FdDate model object and return a string representation.
     * @param {?} date FdDate to format to string value.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    function (date) {
        return null;
    };
    DateFormatParserDefault.decorators = [
        { type: Injectable }
    ];
    return DateFormatParserDefault;
}(DateFormatParser));
export { DateFormatParserDefault };
