// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification-header/notification-header.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-header/notification-header.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Input, Output, ViewEncapsulation } from '@angular/core';
var NotificationHeaderComponent = /** @class */ (function () {
    function NotificationHeaderComponent() {
        /**
         * @hidden
         */
        this.fdNotificationHeaderClass = true;
        /**
         * Whether user wants to hide close button at the top. It's shown by default.
         */
        this.closeButton = true;
        /**
         * Event thrown always, when the close button is clicked
         */
        this.closeButtonClick = new EventEmitter();
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    NotificationHeaderComponent.prototype.closeButtonClicked = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.closeButtonClick.emit();
    };
    NotificationHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification-header',
                    template: "<div [ngClass]=\"['fd-notification__indicator--' + type]\" *ngIf=\"type\"></div>\n<ng-content select=\"[fd-notification-title]\"></ng-content>\n<button fd-button [glyph]=\"'decline'\" [options]=\"'light'\" class=\"fd-notification__close\" (click)=\"closeButtonClicked()\"></button>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [""]
                }] }
    ];
    NotificationHeaderComponent.propDecorators = {
        fdNotificationHeaderClass: [{ type: HostBinding, args: ['class.fd-notification__header',] }],
        type: [{ type: Input }],
        closeButton: [{ type: Input }],
        closeButtonClick: [{ type: Output }]
    };
    return NotificationHeaderComponent;
}());
export { NotificationHeaderComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationHeaderComponent.prototype.fdNotificationHeaderClass;
    /**
     * type of Notification 'success' | 'warning' | 'information' | 'error'
     * @type {?}
     */
    NotificationHeaderComponent.prototype.type;
    /**
     * Whether user wants to hide close button at the top. It's shown by default.
     * @type {?}
     */
    NotificationHeaderComponent.prototype.closeButton;
    /**
     * Event thrown always, when the close button is clicked
     * @type {?}
     */
    NotificationHeaderComponent.prototype.closeButtonClick;
}
