// @fundamental-ngx/core@0.13.2/esm2015/lib/tabs/tab-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tabs/tab-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, EventEmitter, Input, Output, QueryList, ViewChildren, ViewEncapsulation } from '@angular/core';
import { TabPanelComponent } from './tab/tab-panel.component';
import { TabsService } from './tabs.service';
/**
 * Represents a list of tab-panels.
 */
export class TabListComponent {
    /**
     * @param {?} tabsService
     */
    constructor(tabsService) {
        this.tabsService = tabsService;
        /**
         * Index of the selected tab panel.
         */
        this.selectedIndex = 0;
        /**
         * Event emitted when the selected panel changes.
         */
        this.selectedIndexChange = new EventEmitter();
    }
    /**
     * @hidden
     * @return {?}
     */
    ngAfterContentInit() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.selectTab(this.selectedIndex);
        }));
        this._tabSelectSubscription = this.tabsService.tabSelected.subscribe((/**
         * @param {?} index
         * @return {?}
         */
        index => {
            if (index !== this.selectedIndex) {
                this.selectTab(index);
            }
        }));
        this._tabsSubscription = this.panelTabs.changes.subscribe((/**
         * @return {?}
         */
        () => {
            if (!this.isIndexInRange() || this.isTabContentEmpty()) {
                this.resetTabHook();
            }
        }));
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnDestroy() {
        this._tabsSubscription.unsubscribe();
        this._tabSelectSubscription.unsubscribe();
    }
    /**
     * @hidden
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.selectedIndex) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.selectTab(changes.selectedIndex.currentValue);
            }));
        }
    }
    /**
     * Function to select a new tab from an index.
     * @param {?} tabIndex Index of the tab to select.
     * @return {?}
     */
    selectTab(tabIndex) {
        if (this.isIndexInRange() && this.isTargetTabEnabled(tabIndex)) {
            this.panelTabs.forEach((/**
             * @param {?} tab
             * @param {?} index
             * @return {?}
             */
            (tab, index) => {
                tab.expanded = index === tabIndex;
            }));
            this.selectedIndex = tabIndex;
            this.selectedIndexChange.emit(tabIndex);
        }
    }
    /**
     * @hidden
     * @param {?} tabIndex
     * @return {?}
     */
    tabHeaderClickHandler(tabIndex) {
        if (this.selectedIndex !== tabIndex) {
            this.selectTab(tabIndex);
        }
    }
    /**
     * @hidden
     * @param {?} index
     * @param {?} event
     * @return {?}
     */
    tabHeaderKeyHandler(index, event) {
        this.tabsService.tabHeaderKeyHandler(index, event, this.tabLinks.map((/**
         * @param {?} tab
         * @return {?}
         */
        tab => tab.nativeElement)));
    }
    /**
     * @private
     * @return {?}
     */
    isIndexInRange() {
        return this.panelTabs && this.panelTabs.length > 0 && this.selectedIndex < this.panelTabs.length;
    }
    /**
     * @private
     * @param {?} index
     * @return {?}
     */
    isTargetTabEnabled(index) {
        return !this.panelTabs.toArray()[index].disabled;
    }
    /**
     * @private
     * @return {?}
     */
    isTabContentEmpty() {
        /** @type {?} */
        let result = true;
        this.panelTabs.forEach((/**
         * @param {?} tab
         * @return {?}
         */
        tab => {
            if (tab.expanded) {
                result = false;
            }
        }));
        return result;
    }
    /**
     * @private
     * @return {?}
     */
    resetTabHook() {
        this.selectedIndex = 0;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.selectTab(this.selectedIndex);
        }));
    }
}
TabListComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-tab-list',
                template: "<ul class=\"fd-tabs\"\n    role=\"tablist\">\n    <li fd-tab-item\n        *ngFor=\"let tab of panelTabs.toArray(); let i = index;\">\n        <a #tabLink\n           fd-tab-link\n           [disabled]=\"tab.disabled\"\n           [active]=\"tab.expanded\"\n           [attr.tabindex]=\"tab.disabled ? -1 : 0\"\n           [attr.aria-controls]=\"tab.id\"\n           [attr.aria-label]=\"tab.ariaLabel || null\"\n           [attr.aria-labelledby]=\"(!tab.ariaLabel && tab.ariaLabelledBy) ? tab.ariaLabelledBy : null\"\n           (keydown)=\"tabHeaderKeyHandler(i, $event)\"\n           (click)=\"tabHeaderClickHandler(i)\">\n\n            <ng-container *ngIf=\"tab.titleTemplate\">\n                <ng-container [fd-tab-load-title]=\"tab.titleTemplate\"></ng-container>\n            </ng-container>\n            <ng-container *ngIf=\"!tab.titleTemplate\">{{ tab.title }}</ng-container>\n        </a>\n    </li>\n</ul>\n<ng-content select=\"fd-tab\"></ng-content>\n<ng-content></ng-content>\n",
                host: {
                    class: 'fd-tabs-custom'
                },
                encapsulation: ViewEncapsulation.None,
                providers: [TabsService],
                styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-tabs+()\n    .fd-tabs__item?\n    .fd-tabs__link+((.is-selected|[aria-selected=true]),(.is-disabled|[aria-disabled=true]))\n.fd-tabs__panel+([aria-expanded]|.is-expanded)\n*/.fd-tabs{--fd-tabs-border-color:var(--fd-color-neutral-3, #d9d9d9);--fd-tabs-background-color:var(--fd-color-background-2, #fff);--fd-tabs-link-color:var(--fd-color-text-2, #51555a);--fd-tabs-link-border-color:var(--fd-color-action-1);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 12px;border:0;background-color:#fff;background-color:var(--fd-tabs-background-color,#fff);display:flex;flex-wrap:wrap;list-style:none;border-bottom-color:var(--fd-tabs-background-color,#fff);border-bottom:1px solid #fff}.fd-tabs:after,.fd-tabs:before{box-sizing:inherit;font-size:inherit}.fd-tabs__item{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 16px;margin:0;border:0}.fd-tabs__item:after,.fd-tabs__item:before{box-sizing:inherit;font-size:inherit}.fd-tabs__link{font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:16px 0;margin:0;border:0;display:block;position:relative;text-decoration:none;font-size:14px;line-height:1.42857;font-weight:400;cursor:pointer;color:#51555a;color:var(--fd-tabs-link-color,#51555a)}.fd-tabs__link:after,.fd-tabs__link:before{box-sizing:inherit;font-size:inherit}.fd-tabs__link:link{color:#51555a;color:var(--fd-tabs-link-color,#51555a)}.fd-tabs__link:after{transition:125ms ease-in;content:\"\";position:absolute;bottom:0;left:0;display:inline-block;height:3px;width:100%;background-color:transparent;border-radius:4px 4px 0 0}.fd-tabs__link .is-hover,.fd-tabs__link:hover{color:#0254a7;color:var(--fd-color-action-selected,#0254a7)}.fd-tabs__link.is-selected,.fd-tabs__link[aria-selected=true]{color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1)}.fd-tabs__link.is-selected:after,.fd-tabs__link[aria-selected=true]:after{background-color:#0a6ed1;background-color:var(--fd-tabs-link-border-color,#0a6ed1)}.fd-tabs__link.is-disabled,.fd-tabs__link:disabled,.fd-tabs__link[aria-disabled=true]{color:var(--fd-color-action-disabled,rgba(10,110,209,.4));cursor:not-allowed}.fd-tabs__panel{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-tabs__panel:after,.fd-tabs__panel:before{box-sizing:inherit;font-size:inherit}.fd-tabs__panel[aria-expanded=false]{display:none}.fd-tabs-custom,.fd-tabs__panel.is-expanded,.fd-tabs__panel[aria-expanded=true]{display:block}"]
            }] }
];
/** @nocollapse */
TabListComponent.ctorParameters = () => [
    { type: TabsService }
];
TabListComponent.propDecorators = {
    panelTabs: [{ type: ContentChildren, args: [TabPanelComponent,] }],
    tabLinks: [{ type: ViewChildren, args: ['tabLink',] }],
    selectedIndex: [{ type: Input }],
    selectedIndexChange: [{ type: Output }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    TabListComponent.prototype.panelTabs;
    /**
     * @hidden
     * @type {?}
     */
    TabListComponent.prototype.tabLinks;
    /**
     * Index of the selected tab panel.
     * @type {?}
     */
    TabListComponent.prototype.selectedIndex;
    /**
     * Event emitted when the selected panel changes.
     * @type {?}
     */
    TabListComponent.prototype.selectedIndexChange;
    /**
     * @type {?}
     * @private
     */
    TabListComponent.prototype._tabsSubscription;
    /**
     * @type {?}
     * @private
     */
    TabListComponent.prototype._tabSelectSubscription;
    /**
     * @type {?}
     * @private
     */
    TabListComponent.prototype.tabsService;
}
