// @fundamental-ngx/core@0.13.2/esm5/lib/form/form-input-message-group/form-input-message-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/form-input-message-group/form-input-message-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
var FormInputMessageGroupComponent = /** @class */ (function () {
    function FormInputMessageGroupComponent() {
        /*
             * To allow user to determine what event he wants to trigger the messages to show
             * Accepts any [HTML DOM Events](https://www.w3schools.com/jsref/dom_obj_event.asp).
             */
        this.triggers = ['click'];
        /*
             * Allows the user to decide if he wants to keep the error message after they click outside
             *  Whether the popover should close when a click is made outside its boundaries.
             */
        this.closeOnOutsideClick = true;
        /**
         * Whether the message is open. Can be used through two-way binding.
         */
        this.isOpen = false;
        /**
         * Event emitted when the state of the isOpen property changes.
         */
        this.isOpenChange = new EventEmitter();
    }
    /**
     * Function is called every time message changes isOpen attribute
     */
    /**
     * Function is called every time message changes isOpen attribute
     * @param {?} isOpen
     * @return {?}
     */
    FormInputMessageGroupComponent.prototype.openChanged = /**
     * Function is called every time message changes isOpen attribute
     * @param {?} isOpen
     * @return {?}
     */
    function (isOpen) {
        this.isOpenChange.emit(isOpen);
    };
    FormInputMessageGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-form-input-message-group',
                    template: "<fd-popover class=\"fd-form-input-message-group\"\n            [triggers]=\"triggers\"\n            [fillControlMode]=\"fillControlMode\"\n            [closeOnOutsideClick]=\"closeOnOutsideClick\"\n            [isOpen]=\"isOpen\"\n            (isOpenChange)=\"openChanged($event)\"\n            [addContainerClass]=\"'fd-popover-container-custom--message'\">\n    <fd-popover-control>\n        <ng-content select=\"[fd-form-control]\"></ng-content>\n    </fd-popover-control>\n    <fd-popover-body>\n        <ng-content select=\"fd-form-message\"></ng-content>\n    </fd-popover-body>\n</fd-popover>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n\n.fd-form-input-message-group\n*/.fd-form-input-message-group{display:flex;flex-grow:1;position:relative}.fd-popover-container-custom--message{flex-grow:1}.fd-popover-container-custom{border:none}"]
                }] }
    ];
    FormInputMessageGroupComponent.propDecorators = {
        triggers: [{ type: Input }],
        closeOnOutsideClick: [{ type: Input }],
        fillControlMode: [{ type: Input }],
        isOpen: [{ type: Input }],
        isOpenChange: [{ type: Output }]
    };
    return FormInputMessageGroupComponent;
}());
export { FormInputMessageGroupComponent };
if (false) {
    /** @type {?} */
    FormInputMessageGroupComponent.prototype.triggers;
    /** @type {?} */
    FormInputMessageGroupComponent.prototype.closeOnOutsideClick;
    /**
     * Preset options for the message body width.
     * * `at-least` will apply a minimum width to the body equivalent to the width of the control.
     * * `equal` will apply a width to the body equivalent to the width of the control.
     * * Leave blank for no effect.
     * @type {?}
     */
    FormInputMessageGroupComponent.prototype.fillControlMode;
    /**
     * Whether the message is open. Can be used through two-way binding.
     * @type {?}
     */
    FormInputMessageGroupComponent.prototype.isOpen;
    /**
     * Event emitted when the state of the isOpen property changes.
     * @type {?}
     */
    FormInputMessageGroupComponent.prototype.isOpenChange;
}
