// @fundamental-ngx/core@0.13.2/esm5/lib/form/form-label/form-label.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/form-label/form-label.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation } from '@angular/core';
/**
 * Label to be linked to a form control.
 *
 * ```html
 * <label fd-form-label for="input-id">Label Text</label>
 * <input fd-form-control type="text" id="input-id" />
 * ```
 */
var FormLabelComponent = /** @class */ (function () {
    function FormLabelComponent() {
        /**
         * @hidden
         */
        this.fdFormLabelClass = true;
        /**
         * Whether form is required
         */
        this.required = false;
        /**
         * Whether label is for checkbox
         */
        this.checkbox = false;
        /**
         * Whether label is for radio
         */
        this.radio = false;
        /**
         * Whether label is for inline-help
         */
        this.inlineHelp = false;
    }
    FormLabelComponent.decorators = [
        { type: Component, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:component-selector
                    selector: '[fd-form-label]',
                    template: "<ng-content></ng-content>",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-label\n*/.fd-form-label{line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 6px 0 0;margin:0 8px 0 0;border:0;white-space:nowrap;text-overflow:ellipsis;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;position:relative;max-width:100%;font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#6a6d70;color:var(--sapContent_LabelColor,#6a6d70);cursor:text}.fd-form-label:after,.fd-form-label:before{box-sizing:inherit;font-size:inherit}.fd-form-label[dir=rtl],[dir=rtl] .fd-form-label{padding-left:6px;padding-right:0;margin-left:8px;margin-right:0}.fd-form-label.is-disabled{opacity:.4}.fd-form-label--required:after,.fd-form-label[aria-required=true]:after{content:\"*\";font-size:.875rem;font-size:var(--sapFontSize,.875rem);font-weight:700;color:#ce3b3b;color:var(--sapField_RequiredColor,#ce3b3b);margin-left:.125rem;position:absolute;right:0;top:0}.fd-form-label--required[dir=rtl]:after,.fd-form-label[aria-required=true][dir=rtl]:after,[dir=rtl] .fd-form-label--required:after,[dir=rtl] .fd-form-label[aria-required=true]:after{margin-left:0;position:absolute;right:auto;left:0}.fd-form-label--inline-help{width:100%;display:flex;justify-content:space-between;align-items:center}.fd-form-label--inline-help .fd-inline-help{top:0}.fd-form-label--toggle{overflow:initial;margin-top:5px;margin-bottom:5px}.fd-form-label{overflow:visible}"]
                }] }
    ];
    FormLabelComponent.propDecorators = {
        fdFormLabelClass: [{ type: HostBinding, args: ['class.fd-form-label',] }],
        required: [{ type: Input }, { type: HostBinding, args: ['class.fd-form-label--required',] }],
        checkbox: [{ type: Input }, { type: HostBinding, args: ['class.fd-form-label--checkbox',] }],
        radio: [{ type: Input }, { type: HostBinding, args: ['class.fd-form-label--radio',] }],
        inlineHelp: [{ type: Input }, { type: HostBinding, args: ['class.fd-form-label--inline-help',] }]
    };
    return FormLabelComponent;
}());
export { FormLabelComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    FormLabelComponent.prototype.fdFormLabelClass;
    /**
     * Whether form is required
     * @type {?}
     */
    FormLabelComponent.prototype.required;
    /**
     * Whether label is for checkbox
     * @type {?}
     */
    FormLabelComponent.prototype.checkbox;
    /**
     * Whether label is for radio
     * @type {?}
     */
    FormLabelComponent.prototype.radio;
    /**
     * Whether label is for inline-help
     * @type {?}
     */
    FormLabelComponent.prototype.inlineHelp;
}
