// @fundamental-ngx/core@0.13.2/esm5/lib/input-group/input-group-directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/input-group/input-group-directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, Input, Renderer2 } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
var InputGroupInputDirective = /** @class */ (function (_super) {
    tslib_1.__extends(InputGroupInputDirective, _super);
    /** @hidden */
    function InputGroupInputDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        _this.compact = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupInputDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-input');
        this._addClassToElement('fd-input-group__input');
        if (this.compact) {
            this._addClassToElement('fd-input--compact');
        }
    };
    InputGroupInputDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-input-group-input]',
                },] }
    ];
    /** @nocollapse */
    InputGroupInputDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    InputGroupInputDirective.propDecorators = {
        compact: [{ type: Input }]
    };
    return InputGroupInputDirective;
}(AbstractFdNgxClass));
export { InputGroupInputDirective };
if (false) {
    /** @type {?} */
    InputGroupInputDirective.prototype.compact;
    /**
     * @type {?}
     * @private
     */
    InputGroupInputDirective.prototype.elementRef;
}
var InputGroupTextareaDirective = /** @class */ (function () {
    function InputGroupTextareaDirective() {
    }
    InputGroupTextareaDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-textarea-group-input]',
                },] }
    ];
    return InputGroupTextareaDirective;
}());
export { InputGroupTextareaDirective };
var InputGroupAddOnDirective = /** @class */ (function (_super) {
    tslib_1.__extends(InputGroupAddOnDirective, _super);
    /** @hidden */
    function InputGroupAddOnDirective(elementRef, renderer) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        /**
         * @hidden
         */
        _this.fdInputGroupAddonClass = true;
        /**
         * The placement of the add-on. Options include *before* and *after*
         */
        _this.placement = 'after';
        /**
         * Whether to apply compact mode to the AddOn.
         */
        _this.compact = false;
        /**
         * Whether the icon add-on or the text add-on is a button.
         */
        _this.button = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupAddOnDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-input-group__addon');
        if (this.button) {
            this._addClassToElement('fd-input-group__addon--button');
        }
        if (this.type) {
            this._addClassToElement('fd-input-group__addon--' + this.type);
        }
        if (this.state) {
            this._addClassToElement('is-' + this.state);
        }
        if (this.compact) {
            this._addClassToElement('fd-input-group__addon--compact');
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupAddOnDirective.prototype.ngAfterContentInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        /**
         * Add fd-input-group__button to button child element
         * @type {?}
         */
        var button = this.elementRef.nativeElement.querySelector('button');
        if (button) {
            this.renderer.addClass(button, 'fd-input-group__button');
        }
    };
    InputGroupAddOnDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-input-group-addon]'
                },] }
    ];
    /** @nocollapse */
    InputGroupAddOnDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    InputGroupAddOnDirective.propDecorators = {
        fdInputGroupAddonClass: [{ type: HostBinding, args: ['class.fd-input-group__addon',] }],
        placement: [{ type: Input }],
        compact: [{ type: Input }],
        type: [{ type: Input }],
        state: [{ type: Input }],
        button: [{ type: Input }]
    };
    return InputGroupAddOnDirective;
}(AbstractFdNgxClass));
export { InputGroupAddOnDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.fdInputGroupAddonClass;
    /**
     * The placement of the add-on. Options include *before* and *after*
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.placement;
    /**
     * Whether to apply compact mode to the AddOn.
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.compact;
    /**
     * The placement of the add-on. Options include *before* and *after*
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.type;
    /**
     *  The state of the form control - applies css classes.
     *  Can be `valid`, `invalid`, `warning`, `information` or blank for default.
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.state;
    /**
     * Whether the icon add-on or the text add-on is a button.
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.button;
    /**
     * @type {?}
     * @private
     */
    InputGroupAddOnDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    InputGroupAddOnDirective.prototype.renderer;
}
