// @fundamental-ngx/core@0.13.2/esm5/lib/tabs/tab-nav/tab-nav.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tabs/tab-nav/tab-nav.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChildren, EventEmitter, Output, QueryList, Renderer2, ViewEncapsulation } from '@angular/core';
import { TabLinkDirective } from '../tab-link/tab-link.directive';
import { TabItemDirective } from '../tab-item/tab-item.directive';
import { TabsService } from '../tabs.service';
/**
 * Tab Nav for only navigation mode when you want for example use router-outlet
 *
 * ```html
 * <nav fd-tab-nav>
 *  <div fd-tab-item>
 *      <a fd-tab-link
 *      [active]="true">
 *          Link
 *      </a>
 *  </div>
 *  <div fd-tab-item>
 *      <a fd-tab-link
 *      [active]="false">
 *          Link
 *      </a>
 *  </div>
 *  <a fd-tab-link
 *  [active]="false">
 *      Link
 *  </a>
 * </nav>
 * ```
 */
var TabNavComponent = /** @class */ (function () {
    /** @hidden */
    function TabNavComponent(renderer, tabsService) {
        this.renderer = renderer;
        this.tabsService = tabsService;
        /**
         * Event Thrown every time something is clicked
         */
        this.onKeyDown = new EventEmitter();
    }
    Object.defineProperty(TabNavComponent.prototype, "tabLinks", {
        /** Function that gives possibility to get all the link directives, with and without nav__item wrapper */
        get: /**
         * Function that gives possibility to get all the link directives, with and without nav__item wrapper
         * @return {?}
         */
        function () {
            /** @type {?} */
            var tabLinks = [];
            if (this.links) {
                tabLinks = tabLinks.concat(this.links.map((/**
                 * @param {?} link
                 * @return {?}
                 */
                function (link) { return link; })));
            }
            if (this.items) {
                tabLinks = tabLinks.concat(this.items.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return !!item.linkItem; })).map((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.linkItem; })));
            }
            return tabLinks;
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    TabNavComponent.prototype.ngAfterContentInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        this._tabSelectSubscription = this.tabsService.tabSelected.subscribe((/**
         * @param {?} index
         * @return {?}
         */
        function (index) {
            _this.selectTab(index);
        }));
        this.tabLinks.forEach((/**
         * @param {?} linkElement
         * @param {?} index
         * @return {?}
         */
        function (linkElement, index) {
            _this.renderer.listen(linkElement.elementRef.nativeElement, 'keydown', (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                _this.tabsService.tabHeaderKeyHandler(index, event, _this.tabLinks.map((/**
                 * @param {?} link
                 * @return {?}
                 */
                function (link) { return link.elementRef.nativeElement; })));
            }));
        }));
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    TabNavComponent.prototype.ngOnDestroy = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._tabSelectSubscription.unsubscribe();
    };
    /**
     * Function to select a new tab from an index.
     * @param tabIndex Index of the tab to select.
     */
    /**
     * Function to select a new tab from an index.
     * @param {?} tabIndex Index of the tab to select.
     * @return {?}
     */
    TabNavComponent.prototype.selectTab = /**
     * Function to select a new tab from an index.
     * @param {?} tabIndex Index of the tab to select.
     * @return {?}
     */
    function (tabIndex) {
        this.tabLinks[tabIndex].elementRef.nativeElement.click();
    };
    TabNavComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[fd-tab-nav]',
                    host: {
                        'class': 'fd-tabs',
                        'role': 'tablist'
                    },
                    template: "<ng-content></ng-content>",
                    providers: [TabsService],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-tabs+()\n    .fd-tabs__item?\n    .fd-tabs__link+((.is-selected|[aria-selected=true]),(.is-disabled|[aria-disabled=true]))\n.fd-tabs__panel+([aria-expanded]|.is-expanded)\n*/.fd-tabs{--fd-tabs-border-color:var(--fd-color-neutral-3, #d9d9d9);--fd-tabs-background-color:var(--fd-color-background-2, #fff);--fd-tabs-link-color:var(--fd-color-text-2, #51555a);--fd-tabs-link-border-color:var(--fd-color-action-1);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 12px;border:0;background-color:#fff;background-color:var(--fd-tabs-background-color,#fff);display:flex;flex-wrap:wrap;list-style:none;border-bottom-color:var(--fd-tabs-background-color,#fff);border-bottom:1px solid #fff}.fd-tabs:after,.fd-tabs:before{box-sizing:inherit;font-size:inherit}.fd-tabs__item{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 16px;margin:0;border:0}.fd-tabs__item:after,.fd-tabs__item:before{box-sizing:inherit;font-size:inherit}.fd-tabs__link{font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:16px 0;margin:0;border:0;display:block;position:relative;text-decoration:none;font-size:14px;line-height:1.42857;font-weight:400;cursor:pointer;color:#51555a;color:var(--fd-tabs-link-color,#51555a)}.fd-tabs__link:after,.fd-tabs__link:before{box-sizing:inherit;font-size:inherit}.fd-tabs__link:link{color:#51555a;color:var(--fd-tabs-link-color,#51555a)}.fd-tabs__link:after{transition:125ms ease-in;content:\"\";position:absolute;bottom:0;left:0;display:inline-block;height:3px;width:100%;background-color:transparent;border-radius:4px 4px 0 0}.fd-tabs__link .is-hover,.fd-tabs__link:hover{color:#0254a7;color:var(--fd-color-action-selected,#0254a7)}.fd-tabs__link.is-selected,.fd-tabs__link[aria-selected=true]{color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1)}.fd-tabs__link.is-selected:after,.fd-tabs__link[aria-selected=true]:after{background-color:#0a6ed1;background-color:var(--fd-tabs-link-border-color,#0a6ed1)}.fd-tabs__link.is-disabled,.fd-tabs__link:disabled,.fd-tabs__link[aria-disabled=true]{color:var(--fd-color-action-disabled,rgba(10,110,209,.4));cursor:not-allowed}.fd-tabs__panel{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-tabs__panel:after,.fd-tabs__panel:before{box-sizing:inherit;font-size:inherit}.fd-tabs__panel[aria-expanded=false]{display:none}.fd-tabs__panel.is-expanded,.fd-tabs__panel[aria-expanded=true]{display:block}"]
                }] }
    ];
    /** @nocollapse */
    TabNavComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: TabsService }
    ]; };
    TabNavComponent.propDecorators = {
        links: [{ type: ContentChildren, args: [TabLinkDirective,] }],
        items: [{ type: ContentChildren, args: [TabItemDirective,] }],
        onKeyDown: [{ type: Output }]
    };
    return TabNavComponent;
}());
export { TabNavComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    TabNavComponent.prototype.links;
    /**
     * @hidden
     * @type {?}
     */
    TabNavComponent.prototype.items;
    /**
     * @hidden
     * @type {?}
     * @private
     */
    TabNavComponent.prototype._tabSelectSubscription;
    /**
     * Event Thrown every time something is clicked
     * @type {?}
     */
    TabNavComponent.prototype.onKeyDown;
    /**
     * @type {?}
     * @private
     */
    TabNavComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    TabNavComponent.prototype.tabsService;
}
