// @fundamental-ngx/core@0.13.2/esm5/lib/file-input/directives/file-select.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/file-input/directives/file-select.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener, HostBinding } from '@angular/core';
/**
 * Directive tool to facilitate interacting with a native file input element.
 */
var FileSelectDirective = /** @class */ (function () {
    function FileSelectDirective() {
        /**
         * Whether the input should accept multiple file selections.
         */
        this.multiple = true;
        /**
         * Event emitted when files are selected.
         */
        this.onFileSelect = new EventEmitter();
    }
    Object.defineProperty(FileSelectDirective.prototype, "multipleBinding", {
        /** @hidden */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this.multiple ? '' : undefined;
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    FileSelectDirective.prototype.onChange = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.target instanceof HTMLInputElement) {
            /** @type {?} */
            var elRef = ((/** @type {?} */ (event.target)));
            /** @type {?} */
            var files = elRef.files;
            /** @type {?} */
            var fileArray = Array.from(files);
            if (files.length) {
                this.onFileSelect.emit(fileArray);
            }
        }
    };
    FileSelectDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdFileSelect]',
                },] }
    ];
    FileSelectDirective.propDecorators = {
        multiple: [{ type: Input }],
        onFileSelect: [{ type: Output }],
        multipleBinding: [{ type: HostBinding, args: ['attr.multiple',] }],
        onChange: [{ type: HostListener, args: ['change', ['$event'],] }]
    };
    return FileSelectDirective;
}());
export { FileSelectDirective };
if (false) {
    /**
     * Whether the input should accept multiple file selections.
     * @type {?}
     * @private
     */
    FileSelectDirective.prototype.multiple;
    /**
     * Event emitted when files are selected.
     * @type {?}
     */
    FileSelectDirective.prototype.onFileSelect;
}
