// @fundamental-ngx/core@0.13.2/esm5/lib/mega-menu/mega-menu-link/mega-menu-link.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mega-menu/mega-menu-link/mega-menu-link.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
/**
 *  Directive represents mega menu link.
 *  ```html
 * <a fd-mega-menu-link href="#">Link</a>
 *  ```
 *
 */
var MegaMenuLinkDirective = /** @class */ (function () {
    /** @hidden */
    function MegaMenuLinkDirective(itemEl) {
        this.itemEl = itemEl;
        /**
         * @hidden
         */
        this.fdMegaMenuClass = true;
        /**
         * @hidden
         */
        this.hasChild = false;
        /**
         * @hidden
         */
        this.isExpanded = false;
    }
    /**
     * @return {?}
     */
    MegaMenuLinkDirective.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.focus();
    };
    /**
     * @return {?}
     */
    MegaMenuLinkDirective.prototype.click = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.click();
    };
    MegaMenuLinkDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-mega-menu-link]',
                    host: {
                        'tabindex': '0'
                    }
                },] }
    ];
    /** @nocollapse */
    MegaMenuLinkDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    MegaMenuLinkDirective.propDecorators = {
        fdMegaMenuClass: [{ type: HostBinding, args: ['class.fd-mega-menu__link',] }],
        hasChild: [{ type: Input }, { type: HostBinding, args: ['class.has-child',] }, { type: HostBinding, args: ['attr.aria-haspopup',] }],
        isExpanded: [{ type: Input }, { type: HostBinding, args: ['attr.aria-expanded',] }]
    };
    return MegaMenuLinkDirective;
}());
export { MegaMenuLinkDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuLinkDirective.prototype.fdMegaMenuClass;
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuLinkDirective.prototype.hasChild;
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuLinkDirective.prototype.isExpanded;
    /** @type {?} */
    MegaMenuLinkDirective.prototype.itemEl;
}
