// @fundamental-ngx/core@0.13.2/esm5/lib/radio/radio-button/radio-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/radio/radio-button/radio-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, forwardRef, Input, ViewChild, ChangeDetectionStrategy } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { applyCssClass } from '../../utils/public_api';
/** @type {?} */
var uniqueId = 0;
var RadioButtonComponent = /** @class */ (function () {
    /** @hidden */
    function RadioButtonComponent(changeDetectionRef) {
        this.changeDetectionRef = changeDetectionRef;
        this._compact = false;
        this._state = 'default';
        this._disabled = false;
        /**
         * Set uniqueId to a radio button
         */
        this.id = "radio-id-" + uniqueId++;
        // ControlValueAccessor implementation
        /**
         * @hidden
         */
        this.onChange = (/**
         * @return {?}
         */
        function () { });
        /**
         * @hidden
         */
        this.onTouched = (/**
         * @return {?}
         */
        function () { });
    }
    Object.defineProperty(RadioButtonComponent.prototype, "compact", {
        /** Whether to apply compact mode to the radio button.
         * Value: true or false
         * By default field is set to false
         */
        set: /**
         * Whether to apply compact mode to the radio button.
         * Value: true or false
         * By default field is set to false
         * @param {?} isCompact
         * @return {?}
         */
        function (isCompact) {
            this._compact = isCompact;
            this.buildComponentCssClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioButtonComponent.prototype, "state", {
        /** The field to set state of radio button using:
         * 'valid' | 'invalid' | 'warning' | 'default' | 'information'
         * by default value is set to 'default'
         */
        set: /**
         * The field to set state of radio button using:
         * 'valid' | 'invalid' | 'warning' | 'default' | 'information'
         * by default value is set to 'default'
         * @param {?} newState
         * @return {?}
         */
        function (newState) {
            this._state = newState;
            this.buildComponentCssClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioButtonComponent.prototype, "disabled", {
        /** @hidden */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /** The field is used to tell if radio button should be disabled
         * Value: true or false
         * by default disabled state is set to false
         */
        set: /**
         * The field is used to tell if radio button should be disabled
         * Value: true or false
         * by default disabled state is set to false
         * @param {?} isDisabled
         * @return {?}
         */
        function (isDisabled) {
            this._disabled = isDisabled;
            this.buildComponentCssClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioButtonComponent.prototype, "selectedValue", {
        /** The field should be only used with reactive forms
         * Its purpose is to pass a current selected value from froumGroup
         * The field is mandatory when working with reactive forms
         */
        set: /**
         * The field should be only used with reactive forms
         * Its purpose is to pass a current selected value from froumGroup
         * The field is mandatory when working with reactive forms
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this.actualValue = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioButtonComponent.prototype, "name", {
        /** @hidden */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this._name;
        },
        /** The name of the radio button
         * The field is mandatory
         */
        set: /**
         * The name of the radio button
         * The field is mandatory
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._name = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadioButtonComponent.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.value === undefined) {
                return false;
            }
            return this.actualValue === this.value;
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden */
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    RadioButtonComponent.prototype.registerOnChange = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    RadioButtonComponent.prototype.registerOnTouched = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} isDisabled
     * @return {?}
     */
    RadioButtonComponent.prototype.setDisabledState = /**
     * @hidden
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
        this.changeDetectionRef.detectChanges();
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    RadioButtonComponent.prototype.writeValue = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.valueChange(value);
    };
    // End implementation
    /** @hidden */
    // End implementation
    /**
     * @hidden
     * @return {?}
     */
    RadioButtonComponent.prototype.labelClicked = 
    // End implementation
    /**
     * @hidden
     * @return {?}
     */
    function () {
        this.valueChange(this.value);
        this._setFocusOnNativeElement();
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    RadioButtonComponent.prototype.valueChange = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.actualValue = value;
        this._setNativeElementCheckedState();
        this.changeDetectionRef.detectChanges();
        this.onChange(value);
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    RadioButtonComponent.prototype.ngAfterViewInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.buildComponentCssClass();
        this._checkMandatoryFields();
    };
    /** This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with @applyCssClass decorator
     */
    /**
     * This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with \@applyCssClass decorator
     * @return {?}
     */
    RadioButtonComponent.prototype.buildComponentCssClass = /**
     * This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with \@applyCssClass decorator
     * @return {?}
     */
    function () {
        return [
            'fd-radio',
            this._compact ? 'fd-radio--compact' : '',
            this._state !== 'default' ? "is-" + this._state : ''
        ].join(' ');
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    RadioButtonComponent.prototype.elementRef = /**
     * @hidden
     * @return {?}
     */
    function () {
        return this.inputElement;
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @return {?}
     */
    RadioButtonComponent.prototype._checkMandatoryFields = /**
     * @hidden
     * @private
     * @return {?}
     */
    function () {
        if (!this.name) {
            throw 'name field is required';
        }
        if (!this.value) {
            throw 'value field is required';
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @return {?}
     */
    RadioButtonComponent.prototype._setFocusOnNativeElement = /**
     * @hidden
     * @private
     * @return {?}
     */
    function () {
        if (this.inputElement) {
            this.inputElement.nativeElement.focus();
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @return {?}
     */
    RadioButtonComponent.prototype._setNativeElementCheckedState = /**
     * @hidden
     * @private
     * @return {?}
     */
    function () {
        if (this.inputElement) {
            this.inputElement.nativeElement.checked = this.checked;
        }
    };
    RadioButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-radio-button',
                    template: "<input\n    type=\"radio\"\n    class=\"fd-radio\"\n    [disabled]=\"disabled\"\n    [id]=\"id\"\n    [name]=\"name\"\n    [ngModel]=\"actualValue\"\n    (ngModelChange)=\"valueChange($event)\"\n    #inputElement\n    [value]=\"value\"\n/>\n<label class=\"fd-radio__label\" [for]=\"id\" (click)=\"labelClicked()\">\n    <ng-content></ng-content>\n</label>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return RadioButtonComponent; })),
                            multi: true
                        }
                    ],
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-radio\n*/.fd-radio{position:absolute;opacity:.0001;line-height:1.42857;color:var(--sapTextColor,#32363a);-webkit-font-smoothing:antialiased;-webkit-box-sizing:border-box;padding:0;margin:0;border:.0625rem solid #89919a;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-size:.875rem;font-size:var(--sapFontSize,.875rem);font-weight:400;color:#32363a;color:var(--sapField_TextColor,#32363a);background-color:#fff;background-color:var(--sapField_Background,#fff);box-sizing:border-box;border-width:var(--sapField_BorderWidth,.0625rem);border-color:var(--sapField_BorderColor,#89919a);border-radius:.125rem;border-radius:var(--sapField_BorderCornerRadius,.125rem);outline:0;appearance:none;-webkit-appearance:none;-moz-appearance:none}.fd-radio:after,.fd-radio:before{box-sizing:inherit;font-size:inherit}.fd-radio::-webkit-input-placeholder{font-style:italic;color:#74777a;color:var(--sapField_PlaceholderTextColor,#74777a)}.fd-radio::-moz-placeholder{font-style:italic;color:#74777a;color:var(--sapField_PlaceholderTextColor,#74777a)}.fd-radio:-ms-input-placeholder{font-style:italic;color:#74777a;color:var(--sapField_PlaceholderTextColor,#74777a)}.fd-radio::-ms-input-placeholder{font-style:italic;color:#74777a;color:var(--sapField_PlaceholderTextColor,#74777a)}.fd-radio::placeholder{font-style:italic;color:#74777a;color:var(--sapField_PlaceholderTextColor,#74777a)}.fd-radio::-moz-selection{color:#fff;color:var(--sapContent_ContrastTextColor,#fff);background-color:#0854a0;background-color:var(--sapSelectedColor,#0854a0)}.fd-radio::selection{color:#fff;color:var(--sapContent_ContrastTextColor,#fff);background-color:#0854a0;background-color:var(--sapSelectedColor,#0854a0)}.fd-radio .is-hover,.fd-radio:hover{background-color:#fff;background-color:var(--sapField_Hover_Background,#fff);border-color:#0854a0;border-color:var(--sapField_Hover_BorderColor,#0854a0)}.fd-radio.is-valid{background-color:#fff;background-color:var(--sapField_SuccessBackground,#fff);border-color:#107e3e;border-color:var(--sapField_SuccessColor,#107e3e);border-width:.0625rem}.fd-radio.is-valid .is-hover,.fd-radio.is-valid:hover{background-color:#fff;background-color:var(--sapField_SuccessBackground,#fff);border-color:#107e3e;border-color:var(--sapField_SuccessColor,#107e3e)}.fd-radio.is-invalid,.fd-radio.is-invalid .is-hover,.fd-radio.is-invalid:hover{background-color:#fff;background-color:var(--sapField_InvalidBackground,#fff);border-color:#b00;border-color:var(--sapField_InvalidColor,#b00)}.fd-radio.is-alert,.fd-radio.is-alert .is-hover,.fd-radio.is-alert:hover,.fd-radio.is-warning,.fd-radio.is-warning .is-hover,.fd-radio.is-warning:hover{background-color:#fff;background-color:var(--sapField_WarningBackground,#fff);border-color:#e9730c;border-color:var(--sapField_WarningColor,#e9730c)}.fd-radio.is-information,.fd-radio.is-information .is-hover,.fd-radio.is-information:hover{background-color:#fff;background-color:var(--sapField_InformationBackground,#fff);border-color:#0a6ed1;border-color:var(--sapField_InformationColor,#0a6ed1)}.fd-radio.is-focus,.fd-radio:focus{box-shadow:none;outline-offset:-.1875rem;outline-width:var(--sapContent_FocusWidth,.0625rem);outline-color:var(--sapContent_FocusColor,#000);outline-style:var(--sapContent_FocusStyle,dotted);outline:#000 dotted .0625rem;box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5}.fd-radio.is-alert,.fd-radio.is-information,.fd-radio.is-invalid,.fd-radio.is-warning{border-width:.125rem}.fd-radio.is-alert.is-focus,.fd-radio.is-alert:focus,.fd-radio.is-information.is-focus,.fd-radio.is-information:focus,.fd-radio.is-invalid.is-focus,.fd-radio.is-invalid:focus,.fd-radio.is-warning.is-focus,.fd-radio.is-warning:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;box-shadow:none;outline-offset:-.25rem}.fd-radio.is-disabled,.fd-radio:disabled,.fd-radio[aria-disabled=true]{pointer-events:none;opacity:.4;opacity:.0001}.fd-radio.is-disabled::-webkit-input-placeholder,.fd-radio:disabled::-webkit-input-placeholder,.fd-radio[aria-disabled=true]::-webkit-input-placeholder{color:transparent}.fd-radio.is-disabled::-moz-placeholder,.fd-radio:disabled::-moz-placeholder,.fd-radio[aria-disabled=true]::-moz-placeholder{color:transparent}.fd-radio.is-disabled:-ms-input-placeholder,.fd-radio:disabled:-ms-input-placeholder,.fd-radio[aria-disabled=true]:-ms-input-placeholder{color:transparent}.fd-radio.is-disabled::-ms-input-placeholder,.fd-radio:disabled::-ms-input-placeholder,.fd-radio[aria-disabled=true]::-ms-input-placeholder{color:transparent}.fd-radio.is-disabled::placeholder,.fd-radio:disabled::placeholder,.fd-radio[aria-disabled=true]::placeholder{color:transparent}.fd-radio.is-readonly,.fd-radio[readonly]{background-color:hsla(0,0%,94.9%,.5);background-color:var(--sapField_ReadOnly_Background,hsla(0,0%,94.9%,.5));border-color:#89919a;border-color:var(--sapField_ReadOnly_BorderColor,#89919a)}.fd-radio.is-readonly::-webkit-input-placeholder,.fd-radio[readonly]::-webkit-input-placeholder{color:transparent}.fd-radio.is-readonly::-moz-placeholder,.fd-radio[readonly]::-moz-placeholder{color:transparent}.fd-radio.is-readonly:-ms-input-placeholder,.fd-radio[readonly]:-ms-input-placeholder{color:transparent}.fd-radio.is-readonly::-ms-input-placeholder,.fd-radio[readonly]::-ms-input-placeholder{color:transparent}.fd-radio.is-readonly::placeholder,.fd-radio[readonly]::placeholder{color:transparent}.fd-radio__label{line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 8px 0 0;border:0;white-space:nowrap;text-overflow:ellipsis;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;position:relative;max-width:100%;font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#6a6d70;color:var(--sapContent_LabelColor,#6a6d70);cursor:pointer;display:flex;align-items:center;overflow:visible}.fd-radio__label:after,.fd-radio__label:before{box-sizing:inherit;font-size:inherit}.fd-radio__label[dir=rtl],[dir=rtl] .fd-radio__label{padding-left:6px;padding-right:0;margin-left:8px;margin-right:0}.fd-radio__label:before{content:\"\";height:1.375rem;width:1.375rem;font-size:.6875rem;display:inline-flex;justify-content:center;align-items:center;font-family:SAP-icons;background-color:#fff;background-color:var(--sapField_Background,#fff);margin:.6875rem;color:#0854a0;color:var(--sapSelectedColor,#0854a0);border:.0625rem solid #89919a;border:var(--sapField_BorderWidth,.0625rem) solid var(--sapField_BorderColor,#89919a);border-radius:50%}.fd-radio__label .is-hover:before,.fd-radio__label:hover:before{background-color:#fff;background-color:var(--sapField_Hover_Background,#fff);border-color:#0854a0;border-color:var(--sapField_Hover_BorderColor,#0854a0)}.fd-radio:checked+.fd-radio__label:before{content:\"\uE255\"}.fd-radio.is-alert+.fd-radio__label:before,.fd-radio.is-warning+.fd-radio__label:before{background-color:#fff;background-color:var(--sapField_WarningBackground,#fff);border:.125rem solid #e9730c;border:.125rem solid var(--sapField_WarningColor,#e9730c);color:#32363a;color:var(--sapField_TextColor,#32363a)}.fd-radio.is-alert+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-alert+.fd-radio__label .is-hover:disabled,.fd-radio.is-alert+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-alert+.fd-radio__label:hover.is-disabled,.fd-radio.is-alert+.fd-radio__label:hover:disabled,.fd-radio.is-alert+.fd-radio__label:hover[aria-disabled=true],.fd-radio.is-warning+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-warning+.fd-radio__label .is-hover:disabled,.fd-radio.is-warning+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-warning+.fd-radio__label:hover.is-disabled,.fd-radio.is-warning+.fd-radio__label:hover:disabled,.fd-radio.is-warning+.fd-radio__label:hover[aria-disabled=true]{border-color:#e9730c;border-color:var(--sapField_WarningColor,#e9730c)}.fd-radio.is-invalid+.fd-radio__label:before{background-color:#fff;background-color:var(--sapField_InvalidBackground,#fff);border:.125rem solid #b00;border:.125rem solid var(--sapField_InvalidColor,#b00);color:#b00;color:var(--sapField_InvalidColor,#b00)}.fd-radio.is-invalid+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-invalid+.fd-radio__label .is-hover:disabled,.fd-radio.is-invalid+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-invalid+.fd-radio__label:hover.is-disabled,.fd-radio.is-invalid+.fd-radio__label:hover:disabled,.fd-radio.is-invalid+.fd-radio__label:hover[aria-disabled=true]{border-color:#b00;border-color:var(--sapField_InvalidColor,#b00)}.fd-radio.is-valid+.fd-radio__label:before{background-color:#fff;background-color:var(--sapField_SuccessBackground,#fff);border:.0625rem solid #107e3e;border:var(--sapField_BorderWidth,.0625rem) solid var(--sapField_SuccessColor,#107e3e);color:#107e3e;color:var(--sapField_SuccessColor,#107e3e)}.fd-radio.is-valid+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-valid+.fd-radio__label .is-hover:disabled,.fd-radio.is-valid+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-valid+.fd-radio__label:hover.is-disabled,.fd-radio.is-valid+.fd-radio__label:hover:disabled,.fd-radio.is-valid+.fd-radio__label:hover[aria-disabled=true]{border-color:#107e3e;border-color:var(--sapField_SuccessColor,#107e3e)}.fd-radio.is-information+.fd-radio__label:before{background-color:#fff;background-color:var(--sapField_InformationBackground,#fff);border:.125rem solid #0a6ed1;border:.125rem solid var(--sapField_InformationColor,#0a6ed1);color:#0a6ed1;color:var(--sapField_InformationColor,#0a6ed1)}.fd-radio.is-information+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-information+.fd-radio__label .is-hover:disabled,.fd-radio.is-information+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-information+.fd-radio__label:hover.is-disabled,.fd-radio.is-information+.fd-radio__label:hover:disabled,.fd-radio.is-information+.fd-radio__label:hover[aria-disabled=true]{border-color:#0a6ed1;border-color:var(--sapField_InformationColor,#0a6ed1)}.fd-radio.is-readonly+.fd-radio__label:before,.fd-radio[readonly]+.fd-radio__label:before{background-color:var(--sapField_ReadOnly_Background,hsla(0,0%,94.9%,.5));border:.0625rem solid #89919a;border:var(--sapField_BorderWidth,.0625rem) solid var(--sapField_ReadOnly_BorderColor,#89919a);color:#6a6d70;color:var(--sapContent_NonInteractiveIconColor,#6a6d70)}.fd-radio.is-readonly+.fd-radio__label .is-hover.is-disabled,.fd-radio.is-readonly+.fd-radio__label .is-hover:disabled,.fd-radio.is-readonly+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio.is-readonly+.fd-radio__label:hover.is-disabled,.fd-radio.is-readonly+.fd-radio__label:hover:disabled,.fd-radio.is-readonly+.fd-radio__label:hover[aria-disabled=true],.fd-radio[readonly]+.fd-radio__label .is-hover.is-disabled,.fd-radio[readonly]+.fd-radio__label .is-hover:disabled,.fd-radio[readonly]+.fd-radio__label .is-hover[aria-disabled=true],.fd-radio[readonly]+.fd-radio__label:hover.is-disabled,.fd-radio[readonly]+.fd-radio__label:hover:disabled,.fd-radio[readonly]+.fd-radio__label:hover[aria-disabled=true]{border-color:#89919a;border-color:var(--sapField_ReadOnly_BorderColor,#89919a)}.fd-radio.is-focus+.fd-radio__label:after,.fd-radio:focus+.fd-radio__label:after{content:\"\";position:absolute;border:.0625rem dotted #000;border:var(--sapContent_FocusWidth,.0625rem) var(--sapContent_FocusStyle,dotted) var(--sapContent_FocusColor,#000);top:.5625rem;left:.5625rem;bottom:.5625rem;right:-.125rem}.fd-radio.is-focus+.fd-radio__label[dir=rtl]:after,.fd-radio:focus+.fd-radio__label[dir=rtl]:after,[dir=rtl] .fd-radio.is-focus+.fd-radio__label:after,[dir=rtl] .fd-radio:focus+.fd-radio__label:after{left:-.125rem;right:.5625rem}.fd-radio.is-disabled+.fd-radio__label,.fd-radio:disabled+.fd-radio__label,.fd-radio[aria-disabled=true]+.fd-radio__label{pointer-events:none;opacity:.4}.fd-radio--compact+.fd-radio__label:before{font-size:.4375rem;height:1rem;width:1rem;margin:.5rem}.fd-radio--compact.is-focus,.fd-radio--compact:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5}.fd-radio--compact.is-focus+.fd-radio__label:after,.fd-radio--compact:focus+.fd-radio__label:after{content:\"\";position:absolute;border:.0625rem dotted #000;border:var(--sapContent_FocusWidth,.0625rem) var(--sapContent_FocusStyle,dotted) var(--sapContent_FocusColor,#000);top:.375rem;left:.375rem;bottom:.375rem;right:-.125rem}.fd-radio--compact.is-focus+.fd-radio__label[dir=rtl]:after,.fd-radio--compact:focus+.fd-radio__label[dir=rtl]:after,[dir=rtl] .fd-radio--compact.is-focus+.fd-radio__label:after,[dir=rtl] .fd-radio--compact:focus+.fd-radio__label:after{left:-.125rem;right:.375rem}"]
                }] }
    ];
    /** @nocollapse */
    RadioButtonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    RadioButtonComponent.propDecorators = {
        inputElement: [{ type: ViewChild, args: ['inputElement', { static: false },] }],
        compact: [{ type: Input }],
        state: [{ type: Input }],
        disabled: [{ type: Input }],
        selectedValue: [{ type: Input }],
        name: [{ type: Input }],
        id: [{ type: Input }],
        value: [{ type: Input }]
    };
    tslib_1.__decorate([
        applyCssClass,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", String)
    ], RadioButtonComponent.prototype, "buildComponentCssClass", null);
    return RadioButtonComponent;
}());
export { RadioButtonComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    RadioButtonComponent.prototype.class;
    /**
     * @hidden
     * @type {?}
     */
    RadioButtonComponent.prototype.inputElement;
    /**
     * @hidden
     * @type {?}
     */
    RadioButtonComponent.prototype.actualValue;
    /**
     * @type {?}
     * @private
     */
    RadioButtonComponent.prototype._compact;
    /**
     * @type {?}
     * @private
     */
    RadioButtonComponent.prototype._state;
    /**
     * @type {?}
     * @private
     */
    RadioButtonComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    RadioButtonComponent.prototype._name;
    /**
     * Set uniqueId to a radio button
     * @type {?}
     */
    RadioButtonComponent.prototype.id;
    /**
     * Value field stores information about holding value by radio button
     * The field is mandatory
     * @type {?}
     */
    RadioButtonComponent.prototype.value;
    /**
     * @hidden
     * @type {?}
     */
    RadioButtonComponent.prototype.onChange;
    /**
     * @hidden
     * @type {?}
     */
    RadioButtonComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    RadioButtonComponent.prototype.changeDetectionRef;
}
