// @fundamental-ngx/core@0.13.2/esm2015/lib/datetime-picker/models/fd-datetime.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datetime-picker/models/fd-datetime.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Default date with time model used by the fundamental components.
 */
import { FdDate } from '../../calendar/models/fd-date';
export class FdDatetime {
    /**
     * Static function to get the current date in FdDateTime form.
     * @return {?}
     */
    static getToday() {
        /** @type {?} */
        const date = new Date();
        /** @type {?} */
        const time = { hour: date.getHours(), minute: date.getMinutes(), second: date.getSeconds() };
        return new FdDatetime(FdDate.getToday(), time);
    }
    /**
     * Constructor to build a FdDateTime object from a FdDate and TimeObject.
     * @param {?} date the FdDate object.
     * @param {?} time the TimeObject object.
     */
    constructor(date, time) {
        this.date = date;
        this.time = time;
    }
    /**
     * Get Luxon date object converted to string from FdDate.
     * @return {?}
     */
    toLocaleDateString() {
        if (this.toDate() && this.isTimeValid() && this.isDateValid()) {
            return this.toDate().toLocaleString();
        }
        else {
            return null;
        }
    }
    /**
     * Method that checks validity of time on FdDateTime object.
     * @return {?}
     */
    isTimeValid() {
        if (!this.time ||
            this.hour === undefined ||
            this.minute === undefined ||
            this.second === undefined ||
            isNaN(this.hour) ||
            isNaN(this.minute) ||
            isNaN(this.second)) {
            return false;
        }
        if (this.hour > 23 || this.hour < 0) {
            return false;
        }
        if (this.minute > 59 || this.minute < 0) {
            return false;
        }
        if (this.second > 59 || this.second < 0) {
            return false;
        }
        return true;
    }
    /**
     * Method that checks validity of date on FdDateTime object.
     * @return {?}
     */
    isDateValid() {
        return this.date && this.date.isDateValid();
    }
    /**
     * @return {?}
     */
    get year() {
        if (this.date) {
            return this.date.year;
        }
    }
    /**
     * @return {?}
     */
    get month() {
        if (this.date) {
            return this.date.month;
        }
    }
    /**
     * @return {?}
     */
    get day() {
        if (this.date) {
            return this.date.day;
        }
    }
    /**
     * @return {?}
     */
    get hour() {
        if (this.time) {
            return this.time.hour;
        }
    }
    /**
     * @return {?}
     */
    get minute() {
        if (this.time) {
            return this.time.minute;
        }
    }
    /**
     * @return {?}
     */
    get second() {
        if (this.time) {
            return this.time.second;
        }
    }
    /**
     * Get native date object from FdDate.
     * @return {?}
     */
    toDate() {
        return new Date(this.year, this.month - 1, this.day, this.hour, this.minute, this.second);
    }
}
if (false) {
    /** @type {?} */
    FdDatetime.prototype.date;
    /** @type {?} */
    FdDatetime.prototype.time;
}
