// @fundamental-ngx/core@0.13.2/esm5/lib/datetime-picker/format/datetime-parser.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datetime-picker/format/datetime-parser.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FdDatetime } from '../models/fd-datetime';
import { FdDate } from '../../calendar/models/fd-date';
import * as i0 from "@angular/core";
/**
 * @return {?}
 */
export function DATE_TIME_FORMAT_FACTORY() {
    return new DateTimeFormatParserDefault();
}
/**
 * Abstract class which defines the behaviour of the datetime format and parser.
 * @abstract
 */
var DateTimeFormatParser = /** @class */ (function () {
    function DateTimeFormatParser() {
    }
    DateTimeFormatParser.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                    useFactory: DATE_TIME_FORMAT_FACTORY
                },] }
    ];
    /** @nocollapse */ DateTimeFormatParser.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: DATE_TIME_FORMAT_FACTORY, token: DateTimeFormatParser, providedIn: "root" });
    return DateTimeFormatParser;
}());
export { DateTimeFormatParser };
if (false) {
    /**
     * Should take in a string value and return a FdDatetime model object.
     * @abstract
     * @param {?} value String to concert to a FdDatetime model object.
     * @return {?}
     */
    DateTimeFormatParser.prototype.parse = function (value) { };
    /**
     * Should take in a FdDatetime model object and return a string representation.
     * @abstract
     * @param {?} date FdDatetime object to concert to a date string.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    DateTimeFormatParser.prototype.format = function (date) { };
}
/**
 * Default implementation of the DateFormatParser service.
 */
var DateTimeFormatParserDefault = /** @class */ (function (_super) {
    tslib_1.__extends(DateTimeFormatParserDefault, _super);
    function DateTimeFormatParserDefault() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Takes in a string representation of a date and returns a FdDatetime object.
     * @param value String to convert to a FdDatetime model object.
     */
    /**
     * Takes in a string representation of a date and returns a FdDatetime object.
     * @param {?} value String to convert to a FdDatetime model object.
     * @return {?}
     */
    DateTimeFormatParserDefault.prototype.parse = /**
     * Takes in a string representation of a date and returns a FdDatetime object.
     * @param {?} value String to convert to a FdDatetime model object.
     * @return {?}
     */
    function (value) {
        if (!value) {
            return FdDatetime.getToday();
        }
        else {
            /** @type {?} */
            var date = new Date(value);
            return new FdDatetime(new FdDate(date.getFullYear(), date.getMonth() + 1, date.getDate()), {
                hour: date.getHours(),
                minute: date.getMinutes(),
                second: date.getSeconds()
            });
        }
    };
    /**
     * Takes in a FdDatetime object and returns the string representation.
     * @param date FdDatetime model object to convert to a string.
     * Return null, to keep default angular DatePipe as a formatter.
     */
    /**
     * Takes in a FdDatetime object and returns the string representation.
     * @param {?} date FdDatetime model object to convert to a string.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    DateTimeFormatParserDefault.prototype.format = /**
     * Takes in a FdDatetime object and returns the string representation.
     * @param {?} date FdDatetime model object to convert to a string.
     * Return null, to keep default angular DatePipe as a formatter.
     * @return {?}
     */
    function (date) {
        return null;
    };
    DateTimeFormatParserDefault.decorators = [
        { type: Injectable }
    ];
    return DateTimeFormatParserDefault;
}(DateTimeFormatParser));
export { DateTimeFormatParserDefault };
