// @fundamental-ngx/core@0.13.2/esm5/lib/nested-list/nested-list/nested-list.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list/nested-list.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ContentChildren, Directive, ElementRef, forwardRef, HostBinding, Input, QueryList } from '@angular/core';
import { NestedItemDirective } from '../nested-item/nested-item.directive';
import { NestedListStateService } from '../nested-list-state.service';
var NestedListDirective = /** @class */ (function () {
    /** @hidden */
    function NestedListDirective(nestedListStateService, elementRef) {
        this.nestedListStateService = nestedListStateService;
        this.elementRef = elementRef;
        /**
         * @hidden
         */
        this.fdNestedListItemClass = true;
        /**
         * In case the user wants to no use icons for items in this list
         */
        this.textOnly = false;
        /**
         * In case the user wants put compact mode in this list
         */
        this.compact = false;
        /**
         * @hidden
         */
        this.hidden = false;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    NestedListDirective.prototype.ngAfterContentInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        /** @type {?} */
        var nestedLevel = this.getNestedLevel();
        /** If there is condensed mode, maximum 2nd level class of nest can be added */
        if (this.nestedListStateService.condensed) {
            nestedLevel = Math.min.apply(Math, tslib_1.__spread([nestedLevel, 2]));
        }
        this.handleNestedLevel(nestedLevel);
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @param {?} level
     * @return {?}
     */
    NestedListDirective.prototype.handleNestedLevel = /**
     * @hidden
     * @private
     * @param {?} level
     * @return {?}
     */
    function (level) {
        /** Adding class with the nested level */
        this.elementRef.nativeElement.classList.add('level-' + level);
    };
    /**
     * @hidden
     * Method, that checks how deep is the list element
     */
    /**
     * @hidden
     * Method, that checks how deep is the list element
     * @private
     * @return {?}
     */
    NestedListDirective.prototype.getNestedLevel = /**
     * @hidden
     * Method, that checks how deep is the list element
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var element = this.elementRef.nativeElement;
        /** @type {?} */
        var parentElements = [];
        /** Method that gathers all of the parentNode elements of current NestedListDirective element */
        while (element.parentNode) {
            parentElements.unshift(element);
            element = element.parentNode;
        }
        /**
         * Filter only elements, that has `fd-nested-list` directive attribute
         * @type {?}
         */
        var filteredParentElements = parentElements.filter((/**
         * @param {?} _element
         * @return {?}
         */
        function (_element) { return _element.hasAttribute('fd-nested-list'); }));
        return filteredParentElements.length;
    };
    NestedListDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdNestedList], [fd-nested-list]'
                },] }
    ];
    /** @nocollapse */
    NestedListDirective.ctorParameters = function () { return [
        { type: NestedListStateService },
        { type: ElementRef }
    ]; };
    NestedListDirective.propDecorators = {
        fdNestedListItemClass: [{ type: HostBinding, args: ['class.fd-nested-list',] }],
        textOnly: [{ type: Input }, { type: HostBinding, args: ['class.fd-nested-list--text-only',] }],
        compact: [{ type: Input }, { type: HostBinding, args: ['class.fd-nested-list--compact',] }],
        nestedItems: [{ type: ContentChildren, args: [forwardRef((/**
                     * @return {?}
                     */
                    function () { return NestedItemDirective; })),] }],
        hidden: [{ type: HostBinding, args: ['attr.aria-hidden',] }]
    };
    return NestedListDirective;
}());
export { NestedListDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListDirective.prototype.fdNestedListItemClass;
    /**
     * In case the user wants to no use icons for items in this list
     * @type {?}
     */
    NestedListDirective.prototype.textOnly;
    /**
     * In case the user wants put compact mode in this list
     * @type {?}
     */
    NestedListDirective.prototype.compact;
    /**
     * @hidden
     * This variable is mostly to keep track of this list's children. There is not usage of it inside this directive,
     * but it's used by services and NestedItemDirective by itself,
     * @type {?}
     */
    NestedListDirective.prototype.nestedItems;
    /**
     * @hidden
     * @type {?}
     */
    NestedListDirective.prototype.hidden;
    /**
     * @type {?}
     * @private
     */
    NestedListDirective.prototype.nestedListStateService;
    /**
     * @type {?}
     * @private
     */
    NestedListDirective.prototype.elementRef;
}
