// @fundamental-ngx/core@0.13.2/esm2015/lib/mega-menu/mega-menu-list/mega-menu-list.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mega-menu/mega-menu-list/mega-menu-list.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, Directive, HostBinding, QueryList } from '@angular/core';
import { MegaMenuItemComponent } from '../mega-menu-item/mega-menu-item.component';
import { merge, Subject } from 'rxjs';
import { startWith, takeUntil } from 'rxjs/operators';
import { MenuKeyboardService } from '../../menu/menu-keyboard.service';
/**
 *  Directive represents mega menu list, which contains items.
 *  ```html
 *  <ul fd-mega-menu-list>
 *      <fd-mega-menu-item>
 *          <a fd-mega-menu-link>Item 0</a>
 *          <li fd-mega-menu-subitem>
 *              <a fd-mega-menu-sublink>Sub Item 1</a>
 *          </li>
 *          <li fd-mega-menu-subitem>
 *              <a fd-mega-menu-sublink>Sub Item 2</a>
 *          </li>
 *          <li fd-mega-menu-subitem>
 *              <a fd-mega-menu-sublink>Sub Item 3</a>
 *          </li>
 *      </fd-mega-menu-item>
 *  </ul>
 *  ```
 *
 */
export class MegaMenuListDirective {
    /**
     * @hidden
     * @param {?} menuKeyboardService
     */
    constructor(menuKeyboardService) {
        this.menuKeyboardService = menuKeyboardService;
        /**
         * @hidden
         */
        this.fdMegaMenuClass = true;
        /**
         * An RxJS Subject that will kill the data stream upon component’s destruction (for unsubscribing)
         */
        this.onDestroy$ = new Subject();
        /**
         * An RxJS Subject that will kill the data stream upon queryList changes (for unsubscribing)
         */
        this.onRefresh$ = new Subject();
    }
    /**
     * @hidden
     * @return {?}
     */
    ngAfterContentInit() {
        this.items.changes
            .pipe(takeUntil(this.onDestroy$), startWith(5))
            .subscribe((/**
         * @return {?}
         */
        () => this.refreshSubscription()));
    }
    /**
     * Method that provides handles keydown events from menu item list
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    handleListKeyDown(event, index) {
        this.menuKeyboardService.keyDownHandler(event, index, this.items.toArray());
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnDestroy() {
        this.onDestroy$.next();
        this.onDestroy$.complete();
    }
    /**
     * Whether any querylist detects any changes
     * @private
     * @return {?}
     */
    refreshSubscription() {
        /** Finish all of the streams, form before */
        this.onRefresh$.next();
        /**
         * Merge refresh/destroy observables
         * @type {?}
         */
        const refreshObs = merge(this.onRefresh$, this.onDestroy$);
        /** New subscription streams */
        this.items.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => item.keyDown
            .pipe(takeUntil(refreshObs))
            .subscribe((/**
         * @param {?} keyboardEvent
         * @return {?}
         */
        (keyboardEvent) => this.handleListKeyDown(keyboardEvent, index)))));
    }
}
MegaMenuListDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: '[fd-mega-menu-list]'
            },] }
];
/** @nocollapse */
MegaMenuListDirective.ctorParameters = () => [
    { type: MenuKeyboardService }
];
MegaMenuListDirective.propDecorators = {
    fdMegaMenuClass: [{ type: HostBinding, args: ['class.fd-mega-menu__list',] }],
    items: [{ type: ContentChildren, args: [MegaMenuItemComponent,] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuListDirective.prototype.fdMegaMenuClass;
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuListDirective.prototype.items;
    /**
     * An RxJS Subject that will kill the data stream upon component’s destruction (for unsubscribing)
     * @type {?}
     * @private
     */
    MegaMenuListDirective.prototype.onDestroy$;
    /**
     * An RxJS Subject that will kill the data stream upon queryList changes (for unsubscribing)
     * @type {?}
     * @private
     */
    MegaMenuListDirective.prototype.onRefresh$;
    /**
     * @type {?}
     * @private
     */
    MegaMenuListDirective.prototype.menuKeyboardService;
}
