// @fundamental-ngx/core@0.13.2/esm5/lib/panel/panel-header/panel-header.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/panel/panel-header/panel-header.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Header of the panel. Contains a head and actions.
 *
 * ```html
 * <fd-panel>
 *     <fd-panel-header>
 *         <fd-panel-head>
 *             <fd-panel-title>Title!</fd-panel-title>
 *         </fd-panel-head>
 *     </fd-panel-header>
 * </fd-panel>
 * ```
 */
var PanelHeaderComponent = /** @class */ (function () {
    function PanelHeaderComponent() {
        /**
         * @hidden
         */
        this.fdPanelHeaderClass = true;
    }
    PanelHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-panel-header',
                    template: "<ng-content select=\"fd-panel-head\"></ng-content>\n<ng-content select=\"fd-panel-actions\"></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PanelHeaderComponent.propDecorators = {
        fdPanelHeaderClass: [{ type: HostBinding, args: ['class.fd-panel__header',] }]
    };
    return PanelHeaderComponent;
}());
export { PanelHeaderComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PanelHeaderComponent.prototype.fdPanelHeaderClass;
}
