// @fundamental-ngx/core@0.13.2/esm5/lib/mega-menu/mega-menu-sublink.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mega-menu/mega-menu-sublink.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding } from '@angular/core';
/**
 *  Directive represents mega menu sub link.
 *  ```html
 * <a fd-mega-menu-sublink href="#">Link</a>
 *  ```
 *
 */
var MegaMenuSublinkDirective = /** @class */ (function () {
    /** @hidden */
    function MegaMenuSublinkDirective(itemEl) {
        this.itemEl = itemEl;
        /**
         * @hidden
         */
        this.fdMegaMenuClass = true;
    }
    /**
     * @return {?}
     */
    MegaMenuSublinkDirective.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.focus();
    };
    /**
     * @return {?}
     */
    MegaMenuSublinkDirective.prototype.click = /**
     * @return {?}
     */
    function () {
        this.itemEl.nativeElement.click();
    };
    MegaMenuSublinkDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-mega-menu-sublink]',
                    host: {
                        'tabindex': '0'
                    }
                },] }
    ];
    /** @nocollapse */
    MegaMenuSublinkDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    MegaMenuSublinkDirective.propDecorators = {
        fdMegaMenuClass: [{ type: HostBinding, args: ['class.fd-mega-menu__sublink',] }]
    };
    return MegaMenuSublinkDirective;
}());
export { MegaMenuSublinkDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MegaMenuSublinkDirective.prototype.fdMegaMenuClass;
    /** @type {?} */
    MegaMenuSublinkDirective.prototype.itemEl;
}
