// @fundamental-ngx/core@0.13.2/esm5/lib/nested-list/nested-list-state.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list-state.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
var NestedListStateService = /** @class */ (function () {
    function NestedListStateService() {
        /**
         * @hidden
         * The condensed state is modified by the parent and read by nested lists.
         */
        this.condensed = false;
        /**
         * Event, that is thrown always, when the link's selected state is changed, triggers
         */
        this.refresh$ = new Subject();
    }
    /**
     * @hidden
     * Recursive method, that adds selected state to parents, when any of the children is selected
     */
    /**
     * @hidden
     * Recursive method, that adds selected state to parents, when any of the children is selected
     * @private
     * @param {?} item
     * @return {?}
     */
    NestedListStateService.prototype.selected = /**
     * @hidden
     * Recursive method, that adds selected state to parents, when any of the children is selected
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var _this = this;
        item.allChildrenItems.forEach((/**
         * @param {?} _item
         * @return {?}
         */
        function (_item) { return _item.linkItem.controlSelected = _item.linkItem.selected; }));
        /** @type {?} */
        var hasChildSelected = !!item.allChildrenItems.find((/**
         * @param {?} _item
         * @return {?}
         */
        function (_item) { return (_item.linkItem && _item.linkItem.selected || _this.selected(_item)); }));
        if (hasChildSelected) {
            item.linkItem.controlSelected = hasChildSelected;
        }
        return hasChildSelected;
    };
    /**
     * @hidden
     * Recursive method, that removes all of controlSelected properties.
     */
    /**
     * @hidden
     * Recursive method, that removes all of controlSelected properties.
     * @private
     * @param {?} item
     * @return {?}
     */
    NestedListStateService.prototype.resetSelected = /**
     * @hidden
     * Recursive method, that removes all of controlSelected properties.
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var _this = this;
        if (item.linkItem) {
            item.linkItem.controlSelected = false;
        }
        item.allChildrenItems.forEach((/**
         * @param {?} _item
         * @return {?}
         */
        function (_item) {
            _this.resetSelected(_item);
        }));
    };
    /** Method that adds selected state to elements, depending on children from deeper level */
    /**
     * Method that adds selected state to elements, depending on children from deeper level
     * @param {?} list
     * @return {?}
     */
    NestedListStateService.prototype.propagateSelected = /**
     * Method that adds selected state to elements, depending on children from deeper level
     * @param {?} list
     * @return {?}
     */
    function (list) {
        var _this = this;
        list.nestedItems.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this.resetSelected(item); }));
        list.nestedItems.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this.selected(item); }));
    };
    return NestedListStateService;
}());
export { NestedListStateService };
if (false) {
    /**
     * @hidden
     * The condensed state is modified by the parent and read by nested lists.
     * @type {?}
     */
    NestedListStateService.prototype.condensed;
    /**
     * Event, that is thrown always, when the link's selected state is changed, triggers
     * @type {?}
     */
    NestedListStateService.prototype.refresh$;
}
