// @fundamental-ngx/core@0.13.2/esm5/lib/nested-list/nested-list-directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list-directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
var NestedListHeaderDirective = /** @class */ (function () {
    function NestedListHeaderDirective() {
        /**
         * @hidden
         */
        this.fdNestedListHeaderClass = true;
    }
    NestedListHeaderDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdNestedDirectivesHeader], [fd-nested-list-header]'
                },] }
    ];
    NestedListHeaderDirective.propDecorators = {
        fdNestedListHeaderClass: [{ type: HostBinding, args: ['class.fd-nested-list__group-header',] }]
    };
    return NestedListHeaderDirective;
}());
export { NestedListHeaderDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListHeaderDirective.prototype.fdNestedListHeaderClass;
}
var NestedListIconDirective = /** @class */ (function (_super) {
    tslib_1.__extends(NestedListIconDirective, _super);
    /** @hidden */
    function NestedListIconDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * @hidden
         */
        _this.fdNestedListIconClass = true;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    NestedListIconDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        if (this.glyph) {
            this._addClassToElement('sap-icon--' + this.glyph);
        }
        this._addClassToElement('fd-nested-list__icon');
    };
    NestedListIconDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdNestedDirectivesIcon], [fd-nested-list-icon]'
                },] }
    ];
    /** @nocollapse */
    NestedListIconDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NestedListIconDirective.propDecorators = {
        fdNestedListIconClass: [{ type: HostBinding, args: ['class.fd-nested-list__icon',] }],
        glyph: [{ type: Input }]
    };
    return NestedListIconDirective;
}(AbstractFdNgxClass));
export { NestedListIconDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListIconDirective.prototype.fdNestedListIconClass;
    /**
     * The icon name to display. See the icon page for the list of icons
     * here: https://sap.github.io/fundamental-ngx/icon
     * @type {?}
     */
    NestedListIconDirective.prototype.glyph;
    /**
     * @type {?}
     * @private
     */
    NestedListIconDirective.prototype.elementRef;
}
var NestedListTitleDirective = /** @class */ (function () {
    /** @hidden */
    function NestedListTitleDirective(elementRef) {
        this.elementRef = elementRef;
        /**
         * @hidden
         */
        this.fdNestedListTitleClass = true;
    }
    /** Returns element's InnerText */
    /**
     * Returns element's InnerText
     * @return {?}
     */
    NestedListTitleDirective.prototype.getInnerText = /**
     * Returns element's InnerText
     * @return {?}
     */
    function () {
        return this.elementRef && this.elementRef.nativeElement.innerText;
    };
    NestedListTitleDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdNestedDirectivesTitle], [fd-nested-list-title]'
                },] }
    ];
    /** @nocollapse */
    NestedListTitleDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NestedListTitleDirective.propDecorators = {
        fdNestedListTitleClass: [{ type: HostBinding, args: ['class.fd-nested-list__title',] }]
    };
    return NestedListTitleDirective;
}());
export { NestedListTitleDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListTitleDirective.prototype.fdNestedListTitleClass;
    /**
     * @type {?}
     * @private
     */
    NestedListTitleDirective.prototype.elementRef;
}
