// @fundamental-ngx/core@0.13.2/esm5/lib/form/form-control/form-control.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/form-control/form-control.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, ElementRef, Directive } from '@angular/core';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
/**
 * Directive intended for use on form controls.
 *
 * ```html
 * <input type="text" fd-form-control />
 * ```
 */
var FormControlDirective = /** @class */ (function (_super) {
    tslib_1.__extends(FormControlDirective, _super);
    /** @hidden */
    function FormControlDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Whether form is in compact mode
         */
        _this.compact = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    FormControlDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        if (this.state) {
            this._addClassToElement('is-' + this.state);
        }
        switch (this.type) {
            case 'radio': {
                this._addControlClass('fd-radio');
                break;
            }
            default: {
                if (this.getElementTag() === 'input') {
                    this._addControlClass('fd-input');
                }
                else if (this.getElementTag() === 'textarea') {
                    this._addControlClass('fd-textarea');
                }
                else if (this.getElementTag() === 'select') {
                    this._addControlClass('fd-form-select');
                }
                break;
            }
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @return {?}
     */
    FormControlDirective.prototype.getElementTag = /**
     * @hidden
     * @private
     * @return {?}
     */
    function () {
        if (this.elementRef && this.elementRef.nativeElement) {
            return this.elementRef.nativeElement.tagName.toLocaleLowerCase();
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @private
     * @param {?} className
     * @return {?}
     */
    FormControlDirective.prototype._addControlClass = /**
     * @hidden
     * @private
     * @param {?} className
     * @return {?}
     */
    function (className) {
        this._addClassToElement(className);
        if (this.compact) {
            this._addClassToElement(className + '--compact');
        }
    };
    FormControlDirective.decorators = [
        { type: Directive, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-form-control]',
                },] }
    ];
    /** @nocollapse */
    FormControlDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FormControlDirective.propDecorators = {
        state: [{ type: Input }],
        compact: [{ type: Input }],
        type: [{ type: Input }]
    };
    return FormControlDirective;
}(AbstractFdNgxClass));
export { FormControlDirective };
if (false) {
    /**
     *  The state of the form control - applies css classes.
     *  Can be `valid`, `invalid`, `warning`, `information` or blank for default.
     * @type {?}
     */
    FormControlDirective.prototype.state;
    /**
     * Whether form is in compact mode
     * @type {?}
     */
    FormControlDirective.prototype.compact;
    /** @type {?} */
    FormControlDirective.prototype.type;
    /**
     * @type {?}
     * @private
     */
    FormControlDirective.prototype.elementRef;
}
