// @fundamental-ngx/core@0.13.2/esm5/lib/tile/product-tile.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tile/product-tile.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
/**
 * Product tile is used to display product information.
 * ```html
 * <fd-product-tile>
 *     <fd-product-tile-media [photo]="'https://placeimg.com/600/600/nature'">
 *     </fd-product-tile-media>
 *     <fd-product-tile-content>
 *         <h2 fd-product-tile-title>Default Product Tile</h2>
 *     </fd-product-tile-content>
 * </fd-product-tile>
 * ```
 */
var ProductTileComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ProductTileComponent, _super);
    /** @hidden */
    function ProductTileComponent(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Whether the product tile is disabled.
         */
        _this.disabled = false;
        /**
         * Whether the product tile is rendered as a button.
         */
        _this.isButton = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ProductTileComponent.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-product-tile');
        if (this.disabled) {
            this._addClassToElement('is-disabled');
        }
    };
    ProductTileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-product-tile',
                    host: {
                        '[attr.role]': '(this.isButton === true ? "button" : "")',
                        '[class.fd-product-tile-custom]': 'true'
                    },
                    template: "<ng-content select=\"[fd-product-tile-media]\"></ng-content>\n<ng-content select=\"[fd-product-tile-content]\"></ng-content>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-product-tile+()\n  .fd-product-tile__media+()\n  .fd-product-tile__title+()\n  .fd-product-tile__text+()\n*/.fd-product-tile{--fd-product-tile-color:var(--fd-color-text-3, #6a6d70);--fd-product-tile-background-color:var(--fd-color-background-2, #fff);--fd-product-tile-title-color:var(--fd-color-text-1, #32363a);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#6a6d70;color:var(--fd-product-tile-color,#6a6d70);background-color:#fff;background-color:var(--fd-product-tile-background-color,#fff)}.fd-product-tile:after,.fd-product-tile:before{box-sizing:inherit;font-size:inherit}.fd-product-tile--button,.fd-product-tile[role=button]{cursor:pointer;transition:box-shadow 125ms ease-in}.fd-product-tile--button:hover,.fd-product-tile[role=button]:hover{box-shadow:0 0 5px 0 rgba(106,109,112,.4);position:relative;z-index:1}.fd-product-tile.is-disabled,.fd-product-tile[aria-disabled=true]{opacity:.4}.fd-product-tile__text{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-product-tile__text:after,.fd-product-tile__text:before{box-sizing:inherit;font-size:inherit}.fd-product-tile__media{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;width:100%;min-height:210px;background-repeat:no-repeat;background-position:50%;background-size:cover}.fd-product-tile__media:after,.fd-product-tile__media:before{box-sizing:inherit;font-size:inherit}.fd-product-tile__content{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:24px;margin:0;border:0}.fd-product-tile__content:after,.fd-product-tile__content:before{box-sizing:inherit;font-size:inherit}.fd-product-tile__title{padding:0;margin:0;font-size:16px;line-height:1.25;font-weight:400;color:#32363a;color:var(--fd-product-tile-title-color,#32363a)}.fd-product-tile-custom{display:block}"]
                }] }
    ];
    /** @nocollapse */
    ProductTileComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ProductTileComponent.propDecorators = {
        disabled: [{ type: Input }],
        isButton: [{ type: Input }]
    };
    return ProductTileComponent;
}(AbstractFdNgxClass));
export { ProductTileComponent };
if (false) {
    /**
     * Whether the product tile is disabled.
     * @type {?}
     */
    ProductTileComponent.prototype.disabled;
    /**
     * Whether the product tile is rendered as a button.
     * @type {?}
     */
    ProductTileComponent.prototype.isButton;
    /**
     * @type {?}
     * @private
     */
    ProductTileComponent.prototype.elementRef;
}
