// @fundamental-ngx/core@0.13.2/esm5/lib/tile/tile.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tile/tile.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
/**
 * Tile is used to display information in a simple container format.
 * ```html
 * <fd-tile>
 *     <fd-tile-content>
 *         <h2 fd-tile-title>Tile Tile</h2>
 *         <p>Tile Description</p>
 *     </fd-tile-content>
 * </fd-tile>
 * ```
 */
var TileComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TileComponent, _super);
    /** @hidden */
    function TileComponent(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Whether the tile is disabled.
         */
        _this.disabled = false;
        /**
         * Whether the tile is rendered as a button.
         */
        _this.isButton = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    TileComponent.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-tile');
        if (this.disabled) {
            this._addClassToElement('is-disabled');
        }
        if (this.rowSpan) {
            this._addClassToElement('fd-has-grid-row-span-' + this.rowSpan);
        }
        if (this.columnSpan) {
            this._addClassToElement('fd-has-grid-column-span-' + this.columnSpan);
        }
        if (this.colorAccent) {
            this._addClassToElement('fd-has-background-color-accent-' + this.colorAccent);
        }
    };
    TileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-tile',
                    host: {
                        '[attr.role]': '(this.isButton === true ? \'button\' : \'\')'
                    },
                    template: "<ng-content select=\"[fd-tile-media]\"></ng-content>\n<ng-content select=\"[fd-tile-content]\"></ng-content>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-tile+()\n        .fd-tile__media+()\n        .fd-tile__content+()\n        .fd-tile__title+()\n        .fd-tile__actions+()\n*/.fd-tile{color:var(--sapTile_TextColor,#6a6d70);background-color:#fff;background-color:var(--sapTile_Background,#fff);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:flex}.fd-tile:after,.fd-tile:before{box-sizing:inherit;font-size:inherit}.fd-tile__text{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-tile__text:after,.fd-tile__text:before{box-sizing:inherit;font-size:inherit}.fd-tile--button,.fd-tile[role=button]{cursor:pointer;transition:box-shadow 125ms ease-in}.fd-tile--button:hover,.fd-tile[role=button]:hover{box-shadow:0 0 5px 0 rgba(transparent,.4);box-shadow:0 0 5px 0 rgba(var(--sapTile_BorderColor,transparent),.4);position:relative;z-index:1}.fd-tile.is-disabled{opacity:.4}.fd-tile__media{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:10px 0 10px 16px;margin:0;border:0}.fd-tile__media:after,.fd-tile__media:before{box-sizing:inherit;font-size:inherit}.fd-tile__media[dir=rtl],[dir=rtl] .fd-tile__media{padding-right:16px;padding-left:0}.fd-tile__content{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:10px 16px 10px 12px;margin:0;border:0;flex:1;-ms-grid-row-align:center;align-self:center}.fd-tile__content:after,.fd-tile__content:before{box-sizing:inherit;font-size:inherit}.fd-tile__content[dir=rtl],[dir=rtl] .fd-tile__content{padding-right:12px;padding-left:16px}.fd-tile__content:first-child{padding-left:16px}.fd-tile__content:first-child[dir=rtl],[dir=rtl] .fd-tile__content:first-child{padding-right:16px}.fd-tile__title{font-size:14px;line-height:1.42857;font-weight:400;color:#32363a;color:var(--sapTile_TitleTextColor,#32363a);padding:0;margin:0}.fd-tile__actions{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:12px 16px 12px 12px;margin:0;border:0}.fd-tile__actions:after,.fd-tile__actions:before{box-sizing:inherit;font-size:inherit}"]
                }] }
    ];
    /** @nocollapse */
    TileComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    TileComponent.propDecorators = {
        disabled: [{ type: Input }],
        isButton: [{ type: Input }],
        rowSpan: [{ type: Input }],
        columnSpan: [{ type: Input }],
        colorAccent: [{ type: Input }]
    };
    return TileComponent;
}(AbstractFdNgxClass));
export { TileComponent };
if (false) {
    /**
     * Whether the tile is disabled.
     * @type {?}
     */
    TileComponent.prototype.disabled;
    /**
     * Whether the tile is rendered as a button.
     * @type {?}
     */
    TileComponent.prototype.isButton;
    /**
     * Specifies the number of rows a tile should span.
     * @type {?}
     */
    TileComponent.prototype.rowSpan;
    /**
     * Specifies the number of columns a tile should span.
     * @type {?}
     */
    TileComponent.prototype.columnSpan;
    /**
     * A number specifying the background color of the tile.
     * @type {?}
     */
    TileComponent.prototype.colorAccent;
    /**
     * @type {?}
     * @private
     */
    TileComponent.prototype.elementRef;
}
