// @fundamental-ngx/core@0.13.2/esm2015/lib/nested-list/nested-list-popover/nested-list-popover.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-list-popover/nested-list-popover.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, HostBinding, ViewChild, ViewEncapsulation } from '@angular/core';
import { NestedLinkDirective } from '../nested-link/nested-link.directive';
import { NestedListKeyboardService } from '../nested-list-keyboard.service';
import { PopoverComponent } from '../../popover/popover.component';
export class NestedListPopoverComponent {
    /**
     * @hidden
     * @param {?} keyboardNestService
     */
    constructor(keyboardNestService) {
        this.keyboardNestService = keyboardNestService;
        /**
         * @hidden
         */
        this.popoverClass = true;
        /**
         * @hidden
         */
        this.open = false;
        this.keyboardNestService.refresh$.subscribe((/**
         * @return {?}
         */
        () => {
            /** Update popover position, on list of hidden items change */
            if (this.popoverComponent) {
                this.popoverComponent.updatePopover();
            }
        }));
    }
    /**
     * Method called, when open state is changed, from popover component (escape key, outside click).
     * @param {?} open
     * @return {?}
     */
    handleOpenChange(open) {
        this.open = open;
        if (this.parentItemElement) {
            if (open) {
                this.parentItemElement.triggerOpen();
            }
            else {
                this.parentItemElement.triggerClose();
            }
        }
    }
}
NestedListPopoverComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-nested-list-popover',
                template: "<fd-popover [placement]=\"'right-start'\" [triggers]=\"[]\" [noArrow]=\"false\" [isOpen]=\"open\" (isOpenChange)=\"handleOpenChange($event)\">\n    <fd-popover-control>\n        <ng-content select=\"[fd-nested-list-link]\"></ng-content>\n    </fd-popover-control>\n    <fd-popover-body class=\"fd-nested-list--text-only\">\n        <a fd-nested-list-link class=\"fd-nested-list__popover-header\">\n            <span fd-nested-list-title>{{linkDirective?.getTitle()}}</span>\n        </a>\n        <ng-content select=\"[fd-nested-list]\"></ng-content>\n        <ng-content select=\"fd-prepared-nested-list\"></ng-content>\n    </fd-popover-body>\n</fd-popover>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".fd-side-nav--condensed .fd-popover-container-custom .fd-nested-list__title{display:inline-block}.fd-side-nav--condensed .fd-nested-list__popover-header{box-shadow:0 0 .25rem 0 rgba(0,0,0,.15),inset 0 -.0625rem 0 0 #d9d9d9}"]
            }] }
];
/** @nocollapse */
NestedListPopoverComponent.ctorParameters = () => [
    { type: NestedListKeyboardService }
];
NestedListPopoverComponent.propDecorators = {
    popoverComponent: [{ type: ViewChild, args: [PopoverComponent, { static: false },] }],
    popoverClass: [{ type: HostBinding, args: ['class.fd-nested-list__popover',] }],
    linkDirective: [{ type: ContentChild, args: [NestedLinkDirective, { static: false },] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedListPopoverComponent.prototype.popoverComponent;
    /**
     * @hidden
     * @type {?}
     */
    NestedListPopoverComponent.prototype.popoverClass;
    /**
     * @hidden
     * @type {?}
     */
    NestedListPopoverComponent.prototype.linkDirective;
    /**
     * @hidden
     * Reference to parent item, to propagate open and close change from popover.
     * @type {?}
     */
    NestedListPopoverComponent.prototype.parentItemElement;
    /**
     * @hidden
     * @type {?}
     */
    NestedListPopoverComponent.prototype.open;
    /**
     * @type {?}
     * @private
     */
    NestedListPopoverComponent.prototype.keyboardNestService;
}
