// @fundamental-ngx/core@0.13.2/esm5/lib/tree/tree.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tree/tree.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, QueryList, ViewChildren, ViewEncapsulation } from '@angular/core';
import { TreeChildComponent } from './tree-child.component';
var TreeComponent = /** @class */ (function () {
    function TreeComponent() {
        this.editRowClicked = new EventEmitter();
        this.deleteRowClicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TreeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.hideAll = false;
    };
    /**
     * @return {?}
     */
    TreeComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.treeData && this.treeData.length) {
            this.treeData.forEach((/**
             * @param {?} row
             * @return {?}
             */
            function (row) {
                _this.getChildDepth(row, 0);
                _this.handleEmptyTrailingCells(row); // handle empty cells for parents
            }));
        }
    };
    /**
     * @return {?}
     */
    TreeComponent.prototype.toggleDisplayAll = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.hideAll = !this.hideAll;
        this.treeChildren.forEach((/**
         * @param {?} child
         * @return {?}
         */
        function (child) {
            child.toggleDisplayChildren(_this.hideAll);
        }));
    };
    /**
     * @param {?} row
     * @param {?} depth
     * @return {?}
     */
    TreeComponent.prototype.getChildDepth = /**
     * @param {?} row
     * @param {?} depth
     * @return {?}
     */
    function (row, depth) {
        var _this = this;
        if (depth > 0) {
            row.sublevelClass = 'fd-tree__group--sublevel-' + depth;
        }
        if (row.children) {
            row.children.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                _this.getChildDepth(child, depth + 1);
                _this.handleEmptyTrailingCells(child); // handle empty cells for children
            }));
        }
    };
    /**
     * @param {?} row
     * @return {?}
     */
    TreeComponent.prototype.handleEmptyTrailingCells = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        if (row &&
            row.rowData &&
            row.rowData.length &&
            typeof row.rowData[0] !== 'object' &&
            this.headers &&
            this.headers.length) {
            while (row.rowData.length < this.headers.length) {
                row.rowData.push('');
            }
        }
        else if (row &&
            row.rowData &&
            row.rowData.length &&
            typeof row.rowData[0] === 'object' &&
            this.headers &&
            this.headers.length) {
            while (row.rowData.length < this.headers.length) {
                row.rowData.push({
                    displayText: ''
                });
            }
        }
    };
    /**
     * @param {?} row
     * @return {?}
     */
    TreeComponent.prototype.editClicked = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this.editRowClicked.emit(row);
    };
    /**
     * @param {?} row
     * @return {?}
     */
    TreeComponent.prototype.deleteClicked = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this.deleteRowClicked.emit(row);
    };
    TreeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-tree',
                    template: "<div class=\"fd-tree fd-tree--header\"\n     *ngIf=\"headers.length > 0\">\n  <div class=\"fd-tree__row fd-tree__row--header\">\n    <div class=\"fd-tree__col fd-tree__col--control\">\n      <button (click)=\"toggleDisplayAll()\"\n              class=\"fd-tree__control\"\n              aria-label=\"Expand all\"\n              [ngClass]=\"{'is-pressed' : !hideAll}\"\n              [attr.aria-pressed]=\"!hideAll\"></button>\n      {{headers[0]}}\n    </div>\n    <div class=\"fd-tree__col\"\n         *ngFor=\"let header of headers | slice:1\">\n      {{header}}\n    </div>\n    <div class=\"fd-tree__col fd-tree__col--actions\"></div>\n  </div>\n</div>\n<ul class=\"fd-tree\"\n    id=\"ENusD653\"\n    role=\"tree\">\n  <fd-tree-child (editClicked)=\"editClicked($event)\"\n                 (deleteClicked)=\"deleteClicked($event)\"\n                 [displayTreeActions]=\"displayTreeActions\"\n                 *ngFor=\"let row of treeData\"\n                 [row]=\"row\"></fd-tree-child>\n</ul>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-tree+(--header)\n    .fd-tree__group+(--sublevel-1...-6, ([aria-hidden] | .is-hidden))\n    .fd-tree__item\n        .fd-tree__row+(--header, ([aria-selected] | .is-selected))\n            .fd-tree__col+(--control, --actions)\n                .fd-tree__control+([aria-pressed] | .is-pressed)\n*/.fd-tree{--fd-tree-cell-spacing:var(--fd-width-gutter);--fd-tree-border-color:var(--fd-color-neutral-2);--fd-tree-background-color:var(--fd-color-background-2);--fd-tree-link-color:var(--fd-color-action-1);--fd-tree-header-color:var(--fd-color-text-3);--fd-tree-header-background-color:var(--fd-color-neutral-2);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 12px;border:0;position:relative;width:100%;max-width:100%;background-color:#fff;background-color:var(--fd-tree-background-color,#fff);border-bottom-color:var(--fd-tree-border-color,#eeeeef);border-bottom:0 solid #eeeeef}.fd-tree:after,.fd-tree:before{box-sizing:inherit;font-size:inherit}.fd-tree:last-child{margin-right:0;margin-bottom:0}.fd-tree--header{border-bottom:0;border-top-color:var(--fd-tree-border-color,#eeeeef);margin-bottom:0;background-color:#eeeeef;background-color:var(--fd-tree-header-background-color,#eeeeef);border-top:0 solid #eeeeef}.fd-tree__group{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;transition:opacity 125ms linear;max-height:auto;opacity:1}.fd-tree__group:after,.fd-tree__group:before{box-sizing:inherit;font-size:inherit}.fd-tree__group.is-hidden{max-height:0;opacity:0;overflow:hidden}.fd-tree__item{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;border-top-color:var(--fd-tree-border-color,#eeeeef);list-style:none;border-top:0 solid #eeeeef}.fd-tree__item:after,.fd-tree__item:before{box-sizing:inherit;font-size:inherit}.fd-tree__row{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 16px;margin:0;border:0;padding-left:var(--fd-tree-cell-spacing,16px);padding-right:var(--fd-tree-cell-spacing,16px);display:flex;align-items:center;position:relative;transition:background-color 125ms ease-in}.fd-tree__row:after,.fd-tree__row:before{box-sizing:inherit;font-size:inherit}.fd-tree__row .is-hover,.fd-tree__row:hover{background-color:#fafafa;background-color:var(--fd-color-background-hover,#fafafa)}.fd-tree__row--header{color:#6a6d70;color:var(--fd-tree-header-color,#6a6d70);background-color:#eeeeef;background-color:var(--fd-tree-header-background-color,#eeeeef);text-transform:uppercase}.fd-tree__row--header .is-hover,.fd-tree__row--header:hover{background-color:initial}.fd-tree__row.is-selected,.fd-tree__row[aria-selected=true]{background-color:rgba(10,110,209,.07);background-color:var(--fd-color-background-selected,rgba(10,110,209,.07))}.fd-tree__col{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:12px 0;margin:0;border:0;flex:1;padding-left:30px}.fd-tree__col:after,.fd-tree__col:before{box-sizing:inherit;font-size:inherit}.fd-tree__col[dir=rtl],[dir=rtl] .fd-tree__col{padding-left:0;padding-right:30px}.fd-tree__col--control{flex:auto;width:20%}.fd-tree__group--sublevel-1 .fd-tree__col--control{padding-left:60px}.fd-tree__group--sublevel-1 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-1 .fd-tree__col--control{padding-left:0;padding-right:60px}.fd-tree__group--sublevel-2 .fd-tree__col--control{padding-left:90px}.fd-tree__group--sublevel-2 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-2 .fd-tree__col--control{padding-left:0;padding-right:90px}.fd-tree__group--sublevel-3 .fd-tree__col--control{padding-left:120px}.fd-tree__group--sublevel-3 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-3 .fd-tree__col--control{padding-left:0;padding-right:120px}.fd-tree__group--sublevel-4 .fd-tree__col--control{padding-left:150px}.fd-tree__group--sublevel-4 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-4 .fd-tree__col--control{padding-left:0;padding-right:150px}.fd-tree__group--sublevel-5 .fd-tree__col--control{padding-left:180px}.fd-tree__group--sublevel-5 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-5 .fd-tree__col--control{padding-left:0;padding-right:180px}.fd-tree__group--sublevel-6 .fd-tree__col--control{padding-left:210px}.fd-tree__group--sublevel-6 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-6 .fd-tree__col--control{padding-left:0;padding-right:210px}.fd-tree__group--sublevel-7 .fd-tree__col--control{padding-left:240px}.fd-tree__group--sublevel-7 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-7 .fd-tree__col--control{padding-left:0;padding-right:240px}.fd-tree__group--sublevel-8 .fd-tree__col--control{padding-left:270px}.fd-tree__group--sublevel-8 .fd-tree__col--control[dir=rtl],[dir=rtl] .fd-tree__group--sublevel-8 .fd-tree__col--control{padding-left:0;padding-right:270px}.fd-tree__col--actions{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;text-align:right}.fd-tree__col--actions:after,.fd-tree__col--actions:before{box-sizing:inherit;font-size:inherit}.fd-tree__col--actions[dir=rtl],[dir=rtl] .fd-tree__col--actions{text-align:left}.fd-tree__control{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;display:inline-block;margin:-8px 12px 0 -36px;padding:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;border:0;text-decoration:none;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;white-space:nowrap;background-color:transparent;border-radius:.25rem;border-radius:var(--sapButton_BorderCornerRadius,.25rem);position:absolute;top:calc(50% - 9px);width:34px;height:34px;transform:rotate(180deg);vertical-align:middle;transition:transform 125ms linear;color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1)}.fd-tree__control:after,.fd-tree__control:before{box-sizing:inherit;font-size:inherit}.fd-tree__control::-moz-focus-inner{border:0}.fd-tree__control:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE1EB\";font-size:14px;line-height:1.42857}.fd-tree__control.is-focus,.fd-tree__control:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5}.fd-tree__control[dir=rtl],[dir=rtl] .fd-tree__control{margin-right:-40px;transform:rotate(0)}.fd-tree__control.is-pressed{transform:rotate(270deg)}"]
                }] }
    ];
    TreeComponent.propDecorators = {
        headers: [{ type: Input }],
        treeData: [{ type: Input }],
        hideAll: [{ type: Input }],
        displayTreeActions: [{ type: Input }],
        editRowClicked: [{ type: Output }],
        deleteRowClicked: [{ type: Output }],
        treeChildren: [{ type: ViewChildren, args: [TreeChildComponent,] }]
    };
    return TreeComponent;
}());
export { TreeComponent };
if (false) {
    /** @type {?} */
    TreeComponent.prototype.headers;
    /** @type {?} */
    TreeComponent.prototype.treeData;
    /** @type {?} */
    TreeComponent.prototype.hideAll;
    /** @type {?} */
    TreeComponent.prototype.displayTreeActions;
    /** @type {?} */
    TreeComponent.prototype.editRowClicked;
    /** @type {?} */
    TreeComponent.prototype.deleteRowClicked;
    /** @type {?} */
    TreeComponent.prototype.treeChildren;
}
