// @fundamental-ngx/core@0.13.2/esm5/lib/tree/tree-child.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tree/tree-child.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
var TreeChildComponent = /** @class */ (function () {
    function TreeChildComponent() {
        this.editClicked = new EventEmitter();
        this.deleteClicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TreeChildComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.hideChildren = false;
    };
    /**
     * @param {?=} hideAll
     * @return {?}
     */
    TreeChildComponent.prototype.toggleDisplayChildren = /**
     * @param {?=} hideAll
     * @return {?}
     */
    function (hideAll) {
        if (hideAll !== undefined) {
            this.hideChildren = hideAll;
        }
        else {
            this.hideChildren = !this.hideChildren;
        }
    };
    /**
     * @param {?=} variable
     * @return {?}
     */
    TreeChildComponent.prototype.typeOf = /**
     * @param {?=} variable
     * @return {?}
     */
    function (variable) {
        /** @type {?} */
        var retVal;
        if (typeof variable === 'string') {
            retVal = 'string';
        }
        else if (typeof variable === 'object') {
            retVal = 'object';
        }
        return retVal;
    };
    /**
     * @param {?=} row
     * @return {?}
     */
    TreeChildComponent.prototype.editTreeItem = /**
     * @param {?=} row
     * @return {?}
     */
    function (row) {
        if (row) {
            this.editClicked.emit(row);
        }
    };
    /**
     * @param {?=} row
     * @return {?}
     */
    TreeChildComponent.prototype.deleteTreeItem = /**
     * @param {?=} row
     * @return {?}
     */
    function (row) {
        if (row) {
            this.deleteClicked.emit(row);
        }
    };
    TreeChildComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-tree-child',
                    template: "<li #treeChild\n    class=\"fd-tree__item\"\n    role=\"treeitem\">\n  <div class=\"fd-tree__row\">\n    <div *ngFor=\"let cell of row.rowData; let i = index\"\n         [attr.data-index]=\"i\"\n         class=\"fd-tree__col\"\n         [ngClass]=\"{'fd-tree__col--control': i === 0}\">\n      <button (click)=\"toggleDisplayChildren()\"\n              *ngIf=\"row.children && i === 0\"\n              class=\"fd-tree__control\"\n              aria-label=\"Expand\"\n              [ngClass]=\"{'is-pressed' : !hideChildren}\"\n              [attr.aria-pressed]=\"!hideChildren\"></button>\n      <ng-container *ngIf=\"typeOf(cell) === 'string'\">\n        {{cell}}\n      </ng-container>\n      <ng-container *ngIf=\"typeOf(cell) === 'object'\">\n        <ng-container *ngIf=\"cell.linkUrl\">\n          <ng-container *ngIf=\"cell.displayText\">\n            <!-- link with display text -->\n            <a [attr.href]=\"cell.linkUrl\"\n               class=\"fd-has-font-weight-semi\">{{cell.displayText}}</a>\n          </ng-container>\n          <ng-container *ngIf=\"!cell.displayText\">\n            <!-- link without display text -->\n            <a [attr.href]=\"cell.linkUrl\"\n               class=\"fd-has-font-weight-semi\">{{cell.linkUrl}}</a>\n          </ng-container>\n        </ng-container>\n        <ng-container *ngIf=\"!cell.linkUrl\">\n          {{cell.displayText}}\n        </ng-container>\n      </ng-container>\n    </div>\n    <div class=\"fd-tree__col fd-tree__col--actions\">\n      <ng-container *ngIf=\"displayTreeActions\">\n        <fd-popover>\n          <fd-popover-control>\n            <button fd-button\n                    [options]=\"'light'\"\n                    [glyph]=\"'vertical-grip'\"></button>\n          </fd-popover-control>\n          <fd-popover-body>\n            <fd-menu>\n              <ul fd-menu-list>\n                <li fd-menu-item (click)=\"editTreeItem(row)\">Edit</li>\n                <li fd-menu-item (click)=\"deleteTreeItem(row)\">Delete</li>\n              </ul>\n            </fd-menu>\n          </fd-popover-body>\n        </fd-popover>\n      </ng-container>\n    </div>\n  </div>\n  <ul *ngIf=\"row.children && row.children.length > 0\"\n      [ngClass]=\"{'is-hidden': hideChildren}\"\n      class=\"fd-tree__group\"\n      role=\"group\">\n    <fd-tree-child *ngFor=\"let child of row.children\"\n                   [displayTreeActions]=\"displayTreeActions\"\n                   [row]=\"child\"\n                   [ngClass]=\"child.sublevelClass\">\n    </fd-tree-child>\n  </ul>\n</li>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    TreeChildComponent.propDecorators = {
        row: [{ type: Input }],
        hideChildren: [{ type: Input }],
        displayTreeActions: [{ type: Input }],
        editClicked: [{ type: Output }],
        deleteClicked: [{ type: Output }]
    };
    return TreeChildComponent;
}());
export { TreeChildComponent };
if (false) {
    /** @type {?} */
    TreeChildComponent.prototype.row;
    /** @type {?} */
    TreeChildComponent.prototype.hideChildren;
    /** @type {?} */
    TreeChildComponent.prototype.displayTreeActions;
    /** @type {?} */
    TreeChildComponent.prototype.editClicked;
    /** @type {?} */
    TreeChildComponent.prototype.deleteClicked;
}
