// @fundamental-ngx/core@0.13.2/esm2015/lib/split-button/split-button-utils/split-button.directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/split-button/split-button-utils/split-button.directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
/**
 * Directive used to identify the template which will populate the main action button.
 * Used to achieve complex buttons that require more than a string.
 * ```html
 *    <fd-button-split>
 *        <ng-template fd-button-split-action-title>
 *            <p>Paragraph 1</p>
 *            <p>Paragraph 2</p>
 *        </ng-template>
 *        <div fd-button-split-menu>
 *            <fd-menu>
 *                <ul fd-menu-list>
 *                    <li fd-menu-item>
 *                       <a [routerLink]="'/'">option</a>
 *                    </li>
 *                    <li fd-menu-item>
 *                       <a [routerLink]="'/'">option2</a>
 *                    </li>
 *                </ul>
 *            </fd-menu>
 *        </div>
 *    </fd-button-split>
 * </fd-button-split>
 * ```
 */
export class SplitButtonActionTitle {
}
SplitButtonActionTitle.decorators = [
    { type: Directive, args: [{
                // TODO to be discussed
                // tslint:disable-next-line:directive-selector
                selector: '[fd-split-button-action-title]'
            },] }
];
/**
 *   Directive used to specify menu list of items for dropdown.
 * ```html
 *    <fd-button-split>
 *        Action Button
 *        <div fd-button-split-menu>
 *            <fd-menu>
 *                <ul fd-menu-list>
 *                    <li fd-menu-item>
 *                       <a [routerLink]="'/'">option</a>
 *                    </li>
 *                    <li fd-menu-item>
 *                       <a [routerLink]="'/'">option2</a>
 *                    </li>
 *                </ul>
 *            </fd-menu>
 *        </div>
 *    </fd-button-split>
 * </fd-button-split>
 * ```
 */
export class SplitButtonMenuDirective {
}
SplitButtonMenuDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: '[fd-split-button-menu]',
            },] }
];
/**
 * Not for external use. Portal to render the complex title template.
 */
export class SplitButtonLoadActionTitle {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        this.viewRef = viewRef;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.viewRef.clear();
        this.contentRef = this.viewRef.createEmbeddedView(this.content);
    }
}
SplitButtonLoadActionTitle.decorators = [
    { type: Directive, args: [{
                // TODO to be discussed
                // tslint:disable-next-line:directive-selector
                selector: '[fd-split-button-load-action-title]'
            },] }
];
/** @nocollapse */
SplitButtonLoadActionTitle.ctorParameters = () => [
    { type: ViewContainerRef }
];
SplitButtonLoadActionTitle.propDecorators = {
    content: [{ type: Input, args: ['fd-split-button-load-action-title',] }]
};
if (false) {
    /** @type {?} */
    SplitButtonLoadActionTitle.prototype.content;
    /**
     * @type {?}
     * @private
     */
    SplitButtonLoadActionTitle.prototype.contentRef;
    /**
     * @type {?}
     * @private
     */
    SplitButtonLoadActionTitle.prototype.viewRef;
}
