// @fundamental-ngx/core@0.13.2/esm2015/lib/nested-list/prepared-nested-list/prepared-nested-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/prepared-nested-list/prepared-nested-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, forwardRef, Input, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { NestedListDirective } from '../nested-list/nested-list.directive';
import { NestedListStateService } from '../nested-list-state.service';
import { NestedItemDirective } from '../nested-item/nested-item.directive';
/**
 * Component for internal usage, allows to generate the nested list from defined object.
 */
export class PreparedNestedListComponent {
    /**
     * @hidden
     * @param {?} changeDetRef
     * @param {?} stateService
     */
    constructor(changeDetRef, stateService) {
        this.changeDetRef = changeDetRef;
        this.stateService = stateService;
        /**
         * @hidden
         * For internal usage.
         * Defines if the component is first
         */
        this.first = true;
        /**
         * Defines if list should be displayed in condensed mode
         */
        this.condensed = false;
    }
    /**
     * In prepared nested list, nested items should be taken as reference of View, not Content.
     * There is direct reference to these directives here.
     * @return {?}
     */
    get nestedListDirective() {
        return Object.assign(this._nestedListDirective, { nestedItems: this.nestedListItems });
    }
    /**
     * @hidden
     * @return {?}
     */
    ngAfterViewInit() {
        this.stateService.propagateSelected(this.nestedListDirective);
        this.changeDetRef.markForCheck();
        this.changeDetRef.detectChanges();
    }
}
PreparedNestedListComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-prepared-nested-list',
                template: "<ul fd-nested-list [textOnly]=\"list?.textOnly\">\n    <ng-container *ngFor=\"let item of list?.items\">\n        <li *ngIf=\"item.headerTitle\" fd-nested-list-header>\n            {{item.headerTitle}}\n        </li>\n        <li *ngIf=\"!item.headerTitle\"\n            fd-nested-list-item\n            [(expanded)]=\"item.expanded\">\n\n            <fd-nested-list-popover *ngIf=\"condensed && first && item.list\">\n                <a fd-nested-list-link *ngIf=\"item.link\"\n                   [onClickCallback]=\"item.link.callback\"\n                   [selected]=\"item.link.selected\"\n                   [attr.href]=\"item.link.href ? item.link.href : null\"\n                   [routerLink]=\"item.link.routerLink ? item.link.routerLink : []\">\n                    <span fd-nested-list-icon *ngIf=\"item.link.icon\" [glyph]=\"item.link.icon\"></span>\n                    <span fd-nested-list-title>{{item.link.title}}</span>\n                </a>\n                <fd-prepared-nested-list [first]=\"false\" *ngIf=\"item.list\" [list]=\"item.list\"></fd-prepared-nested-list>\n            </fd-nested-list-popover>\n\n            <ng-container *ngIf=\"!condensed || !first || !item.list\">\n                <a fd-nested-list-link *ngIf=\"item.link\"\n                   [onClickCallback]=\"item.link.callback\"\n                   [selected]=\"item.link.selected\"\n                   [attr.href]=\"item.link.href ? item.link.href : null\"\n                   [routerLink]=\"item.link.routerLink ? item.link.routerLink : []\">\n                    <span fd-nested-list-icon *ngIf=\"item.link.icon\" [glyph]=\"item.link.icon\"></span>\n                    <span fd-nested-list-title>{{item.link.title}}</span>\n                </a>\n                <fd-prepared-nested-list [first]=\"false\" *ngIf=\"item.list\" [list]=\"item.list\"></fd-prepared-nested-list>\n            </ng-container>\n\n        </li>\n    </ng-container>\n</ul>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
PreparedNestedListComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: NestedListStateService }
];
PreparedNestedListComponent.propDecorators = {
    first: [{ type: Input }],
    condensed: [{ type: Input }],
    list: [{ type: Input }],
    _nestedListDirective: [{ type: ViewChild, args: [forwardRef((/**
                 * @return {?}
                 */
                () => NestedListDirective)), { static: false },] }],
    nestedListItems: [{ type: ViewChildren, args: [forwardRef((/**
                 * @return {?}
                 */
                () => NestedItemDirective)),] }]
};
if (false) {
    /**
     * @hidden
     * For internal usage.
     * Defines if the component is first
     * @type {?}
     */
    PreparedNestedListComponent.prototype.first;
    /**
     * Defines if list should be displayed in condensed mode
     * @type {?}
     */
    PreparedNestedListComponent.prototype.condensed;
    /**
     * List configuration
     * @type {?}
     */
    PreparedNestedListComponent.prototype.list;
    /**
     * @hidden
     * @type {?}
     */
    PreparedNestedListComponent.prototype._nestedListDirective;
    /**
     * @hidden
     * @type {?}
     */
    PreparedNestedListComponent.prototype.nestedListItems;
    /**
     * @type {?}
     * @private
     */
    PreparedNestedListComponent.prototype.changeDetRef;
    /**
     * @type {?}
     * @private
     */
    PreparedNestedListComponent.prototype.stateService;
}
