// @fundamental-ngx/core@0.13.2/esm2015/lib/nested-list/nested-link/nested-link.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/nested-list/nested-link/nested-link.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, ContentChild, Directive, ElementRef, EventEmitter, HostBinding, Input, Output, Renderer2 } from '@angular/core';
import { NestedListTitleDirective } from '../nested-list-directives';
import { NestedListStateService } from '../nested-list-state.service';
export class NestedLinkDirective {
    /**
     * @hidden
     * @param {?} renderer
     * @param {?} elementRef
     * @param {?} nestedListStateService
     * @param {?} changeDetRef
     */
    constructor(renderer, elementRef, nestedListStateService, changeDetRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.nestedListStateService = nestedListStateService;
        this.changeDetRef = changeDetRef;
        /**
         * @hidden
         */
        this.fdNestedListItemClass = true;
        /**
         * Event that is thrown, when any keyboard event is dispatched on this element
         */
        this.keyboardTriggered = new EventEmitter();
        /**
         * Event that is thrown, when this element is clicked
         */
        this.clicked = new EventEmitter();
        /**
         *
         */
        this.controlSelected = this._selected;
        /**
         * @hidden
         * Attribute controlled by the parent `NestedItemDirective`
         */
        this.expanded = false;
        /**
         * @hidden
         * Attribute controlled by the parent `NestedItemDirective`
         */
        this.hasChildren = false;
    }
    /**
     * Whether this element is selected, the `selected` state is propagated to all of parent elements
     * @param {?} selected
     * @return {?}
     */
    set selected(selected) {
        if (this._selected !== selected) {
            this._selected = selected;
            this.controlSelected = selected;
            this.nestedListStateService.refresh$.next();
        }
    }
    /**
     * @return {?}
     */
    get selected() {
        return this._selected;
    }
    /**
     * Set focus on the element.
     * @return {?}
     */
    focus() {
        this.elementRef.nativeElement.focus();
    }
    /**
     * Dispatches the click event on the element
     * @return {?}
     */
    click() {
        this.elementRef.nativeElement.click();
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnInit() {
        /** Add event listeners on the element */
        /** Keyboard */
        this.renderer.listen(this.elementRef.nativeElement, 'keydown', (/**
         * @param {?} event
         * @return {?}
         */
        (event) => this.keyboardTriggered.emit(event)));
        /** Mouse Click */
        this.renderer.listen(this.elementRef.nativeElement, 'click', (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (this.onClickCallback) {
                this.onClickCallback();
            }
            this.clicked.emit(event);
        }));
    }
    /**
     * Returns the title value of the title directive
     * @return {?}
     */
    getTitle() {
        return this.title && this.title.getInnerText();
    }
}
NestedLinkDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fdNestedLink], [fd-nested-list-link]',
                host: {
                    'tabindex': '0',
                }
            },] }
];
/** @nocollapse */
NestedLinkDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NestedListStateService },
    { type: ChangeDetectorRef }
];
NestedLinkDirective.propDecorators = {
    fdNestedListItemClass: [{ type: HostBinding, args: ['class.fd-nested-list__link',] }],
    title: [{ type: ContentChild, args: [NestedListTitleDirective, { static: false },] }],
    keyboardTriggered: [{ type: Output }],
    clicked: [{ type: Output }],
    selected: [{ type: Input }],
    onClickCallback: [{ type: Input }],
    controlSelected: [{ type: HostBinding, args: ['class.is-selected',] }],
    expanded: [{ type: HostBinding, args: ['class.is-expanded',] }, { type: HostBinding, args: ['attr.aria-expanded',] }],
    hasChildren: [{ type: HostBinding, args: ['class.has-child',] }, { type: HostBinding, args: ['attr.aria-haspopup',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NestedLinkDirective.prototype.fdNestedListItemClass;
    /**
     * @hidden
     *  Reference to title element, it is used, to get title for condensed mode.
     * @type {?}
     */
    NestedLinkDirective.prototype.title;
    /**
     * Event that is thrown, when any keyboard event is dispatched on this element
     * @type {?}
     */
    NestedLinkDirective.prototype.keyboardTriggered;
    /**
     * Event that is thrown, when this element is clicked
     * @type {?}
     */
    NestedLinkDirective.prototype.clicked;
    /** @type {?} */
    NestedLinkDirective.prototype._selected;
    /**
     * Function that is called on click event dispatch on this element.
     * @type {?}
     */
    NestedLinkDirective.prototype.onClickCallback;
    /**
     *
     * @type {?}
     */
    NestedLinkDirective.prototype.controlSelected;
    /**
     * @hidden
     * Attribute controlled by the parent `NestedItemDirective`
     * @type {?}
     */
    NestedLinkDirective.prototype.expanded;
    /**
     * @hidden
     * Attribute controlled by the parent `NestedItemDirective`
     * @type {?}
     */
    NestedLinkDirective.prototype.hasChildren;
    /**
     * @type {?}
     * @private
     */
    NestedLinkDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    NestedLinkDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    NestedLinkDirective.prototype.nestedListStateService;
    /** @type {?} */
    NestedLinkDirective.prototype.changeDetRef;
}
