// @fundamental-ngx/core@0.13.2/esm5/lib/popover/popover-dropdown/popover-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/popover/popover-dropdown/popover-dropdown.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, isDevMode, ViewEncapsulation } from '@angular/core';
/**
 * A component used to enforce a certain layout for the popover. With additional styling
 * ```html
 * <fd-popover>
 *     <fd-dropdown>Dropdown</fd-dropdown>
 *     <fd-popover-body>Popover Body</fd-popover-body>
 * </fd-popover>
 * ```
 */
var PopoverDropdownComponent = /** @class */ (function () {
    function PopoverDropdownComponent() {
        /**
         * Whether the popover should have an arrow.
         */
        this.noArrow = true;
        /**
         * Whether the popover is disabled.
         */
        this.disabled = false;
        /**
         * Whether the dropdown is in compact format.
         */
        this.compact = false;
        /**
         * Whether the dropdown is in a toolbar.
         */
        this.toolbar = false;
        /**
         * Whether the dropdown is opened.
         */
        this.isOpen = false;
        if (isDevMode()) {
            console.warn('Popover Dropdown has been deprecated, it will be removed soon. Replace it by popover with menu button');
        }
    }
    PopoverDropdownComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-dropdown-control',
                    host: {
                        class: 'fd-dropdown',
                    },
                    template: "<div class=\"fd-dropdown\">\n    <button\n        fd-button\n        class=\"fd-dropdown__control\"\n        [fdType]=\"btnType\"\n        [glyph]=\"glyph\"\n        [compact]=\"compact\"\n        [disabled]=\"disabled\"\n        [attr.aria-expanded]=\"this.disabled ? false : isOpen\"\n        [attr.aria-disabled]=\"this.disabled\"\n        aria-haspopup=\"true\"\n    >\n        <ng-content></ng-content>\n    </button>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PopoverDropdownComponent.ctorParameters = function () { return []; };
    PopoverDropdownComponent.propDecorators = {
        noArrow: [{ type: Input }],
        disabled: [{ type: Input }],
        glyph: [{ type: Input }],
        btnType: [{ type: Input }],
        compact: [{ type: Input }],
        toolbar: [{ type: Input }],
        isOpen: [{ type: Input }]
    };
    return PopoverDropdownComponent;
}());
export { PopoverDropdownComponent };
if (false) {
    /**
     * Whether the popover should have an arrow.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.noArrow;
    /**
     * Whether the popover is disabled.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.disabled;
    /**
     * The glyph to display.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.glyph;
    /**
     * The btnType to display.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.btnType;
    /**
     * Whether the dropdown is in compact format.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.compact;
    /**
     * Whether the dropdown is in a toolbar.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.toolbar;
    /**
     * Whether the dropdown is opened.
     * @type {?}
     */
    PopoverDropdownComponent.prototype.isOpen;
}
