// @fundamental-ngx/core@0.13.2/esm5/lib/tabs/tab-utils/tab-directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tabs/tab-utils/tab-directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
/**
 * Directive used to identify the template which will populate the tab header.
 * Used to achieve complex headers that require more than a string.
 *
 * ```html
 * <fd-tab>
 *      <ng-template fd-tab-title>
 *          <fd-icon [glyph]="'delete'"></fd-icon>
 *          <span>Tab Label</span>
 *      </ng-template>
 * </fd-tab>
 * ```
 */
var TabTitleDirective = /** @class */ (function () {
    function TabTitleDirective() {
    }
    TabTitleDirective.decorators = [
        { type: Directive, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-tab-title]'
                },] }
    ];
    return TabTitleDirective;
}());
export { TabTitleDirective };
/**
 * Not for external use. Portal to render the complex title template.
 */
var TabLoadTitleDirective = /** @class */ (function () {
    function TabLoadTitleDirective(viewRef) {
        this.viewRef = viewRef;
    }
    /**
     * @return {?}
     */
    TabLoadTitleDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.viewRef.clear();
        this.contentRef = this.viewRef.createEmbeddedView(this.content);
    };
    TabLoadTitleDirective.decorators = [
        { type: Directive, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-tab-load-title]'
                },] }
    ];
    /** @nocollapse */
    TabLoadTitleDirective.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    TabLoadTitleDirective.propDecorators = {
        content: [{ type: Input, args: ['fd-tab-load-title',] }]
    };
    return TabLoadTitleDirective;
}());
export { TabLoadTitleDirective };
if (false) {
    /** @type {?} */
    TabLoadTitleDirective.prototype.content;
    /**
     * @type {?}
     * @private
     */
    TabLoadTitleDirective.prototype.contentRef;
    /**
     * @type {?}
     * @private
     */
    TabLoadTitleDirective.prototype.viewRef;
}
