// @fundamental-ngx/core@0.13.2/esm5/lib/infinite-scroll/infinite-scroll.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/infinite-scroll/infinite-scroll.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { fromEvent } from 'rxjs';
/**
 * Tool directive used to achieve the infinite scroll mechanism.
 */
var InfiniteScrollDirective = /** @class */ (function () {
    /** @hidden */
    function InfiniteScrollDirective(element) {
        this.element = element;
        /**
         * Scroll percentage at which the onScrollAction event is fired.
         */
        this.scrollPercent = 75;
        /**
         * Event emitted when the scrollPercent threshold is met.
         */
        this.onScrollAction = new EventEmitter();
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InfiniteScrollDirective.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        this.scrollEvent = fromEvent(this.element.nativeElement, 'scroll');
        this.subscription = this.scrollEvent.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if ((e.target.scrollTop + e.target.offsetHeight) / e.target.scrollHeight > _this.scrollPercent / 100) {
                _this.onScrollAction.emit(null);
            }
        }));
    };
    /**
     * @return {?}
     */
    InfiniteScrollDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    InfiniteScrollDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fdInfiniteScroll]'
                },] }
    ];
    /** @nocollapse */
    InfiniteScrollDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    InfiniteScrollDirective.propDecorators = {
        scrollPercent: [{ type: Input }],
        onScrollAction: [{ type: Output }]
    };
    return InfiniteScrollDirective;
}());
export { InfiniteScrollDirective };
if (false) {
    /**
     * Scroll percentage at which the onScrollAction event is fired.
     * @type {?}
     */
    InfiniteScrollDirective.prototype.scrollPercent;
    /**
     * Event emitted when the scrollPercent threshold is met.
     * @type {?}
     */
    InfiniteScrollDirective.prototype.onScrollAction;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollDirective.prototype.scrollEvent;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollDirective.prototype.element;
}
