// @fundamental-ngx/core@0.13.2/esm5/lib/alert/alert.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/alert/alert.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IconModule } from '../icon/icon.module';
import { AlertComponent } from './alert.component';
import { AlertService } from './alert-service/alert.service';
import { AlertContainerComponent } from './alert-utils/alert-container.component';
import { DynamicComponentService } from '../utils/dynamic-component/dynamic-component.service';
import { ButtonModule } from '../button/button.module';
var AlertModule = /** @class */ (function () {
    function AlertModule() {
    }
    AlertModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [AlertComponent, AlertContainerComponent],
                    imports: [CommonModule, IconModule, ButtonModule],
                    exports: [AlertComponent, AlertContainerComponent],
                    entryComponents: [AlertContainerComponent, AlertComponent],
                    providers: [AlertService, DynamicComponentService]
                },] }
    ];
    return AlertModule;
}());
export { AlertModule };
