// @fundamental-ngx/core@0.13.2/esm5/lib/alert/alert.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/alert/alert.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ElementRef, ChangeDetectorRef, ViewChild, ComponentFactoryResolver, Type, ViewContainerRef, TemplateRef, Optional, Output, EventEmitter, ViewEncapsulation, HostListener, NgZone, ChangeDetectionStrategy } from '@angular/core';
import { alertFadeNgIf } from './alert-utils/alert-animations';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
import { AlertRef } from './alert-utils/alert-ref';
/** @type {?} */
var alertUniqueId = 0;
/**
 * The component that represents an alert. It can be only be used inline.
 * If the AlertService is used, this component is auto-generated.
 */
var AlertComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AlertComponent, _super);
    /** @hidden */
    function AlertComponent(elRef, cdRef, componentFactoryResolver, ngZone, alertRef) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        _this.cdRef = cdRef;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.ngZone = ngZone;
        _this.alertRef = alertRef;
        /**
         * Whether the alert is dismissible.
         */
        _this.dismissible = true;
        /**
         * Id for the alert component. If omitted, a unique one is generated.
         */
        _this.id = 'fd-alert-' + alertUniqueId++;
        /**
         * Duration of time *in milliseconds* that the alert will be visible. Set to -1 for indefinite.
         */
        _this.duration = 10000;
        /**
         * Whether the alert should stay open if the mouse is hovering over it.
         */
        _this.mousePersist = false;
        /**
         * Id of the element that labels the alert.
         */
        _this.ariaLabelledBy = null;
        /**
         * Aria label for the alert component element.
         */
        _this.ariaLabel = null;
        /**
         * Aria label for the dismiss button.
         */
        _this.dismissLabel = 'Dismiss';
        /**
         * Event fired when the alert is dismissed.
         */
        _this.onDismiss = new EventEmitter();
        /**
         * @hidden
         */
        _this.mouseInAlert = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    AlertComponent.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        if (this.alertRef) {
            this.open();
        }
        this._setProperties();
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    AlertComponent.prototype.ngAfterViewInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        if (this.childComponentType) {
            if (this.childComponentType instanceof Type) {
                this.loadFromComponent(this.childComponentType);
            }
            else if (this.childComponentType instanceof TemplateRef) {
                this.loadFromTemplate(this.childComponentType);
            }
            else {
                this.loadFromString(this.childComponentType);
            }
            this.cdRef.detectChanges();
        }
    };
    /**
     * Dismisses the alert. If the alert was generated via the AlertService, it is removed from the DOM.
     * Otherwise, it sets the display value to none. Fires the onDismiss event.
     *
     * @param manualDismiss Set to true to skip the dismiss animation.
     * @param reason Data to pass back to the calling component. Only usable if alert is opened using the Service.
     *
     */
    /**
     * Dismisses the alert. If the alert was generated via the AlertService, it is removed from the DOM.
     * Otherwise, it sets the display value to none. Fires the onDismiss event.
     *
     * @param {?=} reason Data to pass back to the calling component. Only usable if alert is opened using the Service.
     *
     * @param {?=} manualDismiss Set to true to skip the dismiss animation.
     * @return {?}
     */
    AlertComponent.prototype.dismiss = /**
     * Dismisses the alert. If the alert was generated via the AlertService, it is removed from the DOM.
     * Otherwise, it sets the display value to none. Fires the onDismiss event.
     *
     * @param {?=} reason Data to pass back to the calling component. Only usable if alert is opened using the Service.
     *
     * @param {?=} manualDismiss Set to true to skip the dismiss animation.
     * @return {?}
     */
    function (reason, manualDismiss) {
        if (manualDismiss === void 0) { manualDismiss = false; }
        if (manualDismiss) {
            this.elRef.nativeElement.classList.add('fd-has-display-none');
            this.elRef.nativeElement.classList.remove('fd-has-display-block');
        }
        if (this.alertRef) {
            this.alertRef.dismiss(reason);
        }
        else {
            this.elRef.nativeElement.classList.add('fd-has-display-none');
            this.elRef.nativeElement.classList.remove('fd-has-display-block');
        }
        this.onDismiss.emit();
    };
    /**
     * Opens the alert.
     */
    /**
     * Opens the alert.
     * @return {?}
     */
    AlertComponent.prototype.open = /**
     * Opens the alert.
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.alertRef) {
            if (this.elRef.nativeElement.style.display === 'block') {
                return;
            }
            this.elRef.nativeElement.classList.remove('fd-has-display-none');
            this.elRef.nativeElement.classList.add('fd-has-display-block');
        }
        if (this.duration >= 0) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    if (_this.mousePersist) {
                        /** @type {?} */
                        var wait_1 = (/**
                         * @return {?}
                         */
                        function () {
                            if (_this.mouseInAlert === true) {
                                setTimeout(wait_1, 500);
                            }
                            else {
                                _this.ngZone.run((/**
                                 * @return {?}
                                 */
                                function () { return _this.dismiss(); }));
                            }
                        });
                        wait_1();
                    }
                    else {
                        _this.ngZone.run((/**
                         * @return {?}
                         */
                        function () { return _this.dismiss(); }));
                    }
                }), _this.duration);
            }));
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    AlertComponent.prototype.handleAlertMouseEvent = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.type === 'mouseenter') {
            this.mouseInAlert = true;
        }
        else if (event.type === 'mouseleave') {
            this.mouseInAlert = false;
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    AlertComponent.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-alert');
        if (this.type) {
            this._addClassToElement('fd-alert--' + this.type);
        }
        if (this.dismissible) {
            this._addClassToElement('fd-alert--dismissible');
        }
    };
    /**
     * @private
     * @param {?} template
     * @return {?}
     */
    AlertComponent.prototype.loadFromTemplate = /**
     * @private
     * @param {?} template
     * @return {?}
     */
    function (template) {
        /** @type {?} */
        var context = {
            $implicit: this.alertRef
        };
        this.componentRef = this.containerRef.createEmbeddedView(template, context);
    };
    /**
     * @private
     * @param {?} componentType
     * @return {?}
     */
    AlertComponent.prototype.loadFromComponent = /**
     * @private
     * @param {?} componentType
     * @return {?}
     */
    function (componentType) {
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(componentType);
        this.containerRef.clear();
        this.componentRef = this.containerRef.createComponent(componentFactory);
    };
    /**
     * @private
     * @param {?} contentString
     * @return {?}
     */
    AlertComponent.prototype.loadFromString = /**
     * @private
     * @param {?} contentString
     * @return {?}
     */
    function (contentString) {
        this.containerRef.clear();
        this.message = contentString;
    };
    AlertComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-alert',
                    template: "<button class=\"fd-alert__close\"\n        fd-button\n        [compact]=\"true\"\n        [options]=\"'light'\"\n        *ngIf=\"dismissible\"\n        (click)=\"dismiss(undefined, true)\"\n        [attr.aria-controls]=\"id\"\n        [attr.aria-label]=\"dismissLabel\">\n</button>\n<p class=\"fd-alert__text\">\n    <ng-container #container>{{message}}</ng-container>\n    <ng-content></ng-content>\n</p>\n",
                    host: {
                        '[attr.aria-labelledby]': 'ariaLabelledBy',
                        '[attr.aria-label]': 'ariaLabel',
                        '[style.width]': 'width',
                        '[style.min-width]': 'minWidth',
                        'role': 'alert',
                        '[attr.id]': 'id',
                        '[@fadeAlertNgIf]': ''
                    },
                    animations: [
                        alertFadeNgIf
                    ],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-overlay\n*/.fd-overlay{--fd-overlay-background-color:rgba(74, 80, 92, 0.3);position:fixed;display:flex;top:0;align-items:center;justify-content:center;z-index:1000}.fd-overlay.is-hidden{display:none}.fd-overlay--modal{background-color:var(--fd-overlay-background-color,rgba(74,80,92,.3));height:100vh;width:100vw}.fd-overlay--alert{position:fixed;left:50%;transform:translate(-50%);margin-top:28px}/*!\n.fd-alert+(--information, --success, --warning, --error, --no-icon)\n    .fd-alert__close\n*/.fd-alert{line-height:1.42857;font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:.4375rem 1rem;margin:0;border:1px solid #6a6d70;font-size:.875rem;font-size:var(--sapFontSize,.875rem);position:relative;color:#32363a;color:var(--sapTextColor,#32363a);background-color:#f4f4f4;background-color:var(--sapNeutralBackground,#f4f4f4);border-color:var(--sapNeutralBorderColor,#6a6d70);border-radius:.25rem;align-items:center;min-height:2rem}.fd-alert:after,.fd-alert:before{box-sizing:inherit;font-size:inherit}.fd-alert .sap-icon--message-error,.fd-alert .sap-icon--message-information,.fd-alert .sap-icon--message-success,.fd-alert .sap-icon--message-warning{display:none}.fd-alert__close{display:inline-block;margin:0;padding:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;outline:0;text-decoration:none;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;vertical-align:middle;white-space:nowrap;background-color:transparent;border-radius:.25rem;border-radius:var(--sapButton_BorderCornerRadius,.25rem);position:absolute;top:.0625rem;right:.0625rem}.fd-alert__close::-moz-focus-inner{border:0}.fd-alert__close:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE03E\";font-size:16px;line-height:1}.fd-alert__close.is-focus,.fd-alert__close:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5}.fd-alert__close[dir=rtl],[dir=rtl] .fd-alert__close{left:.125rem;right:auto}.fd-alert--dismissible{padding-right:2.5rem}.fd-alert--dismissible[dir=rtl],[dir=rtl] .fd-alert--dismissible{padding-right:1rem;padding-left:2.5rem}.fd-alert--error,.fd-alert--information,.fd-alert--success,.fd-alert--warning{padding-left:2.5rem}.fd-alert--error:after,.fd-alert--error:before,.fd-alert--information:after,.fd-alert--information:before,.fd-alert--success:after,.fd-alert--success:before,.fd-alert--warning:after,.fd-alert--warning:before{position:absolute;width:2.5rem;top:.4375rem;left:0}.fd-alert--error[dir=rtl],.fd-alert--information[dir=rtl],.fd-alert--success[dir=rtl],.fd-alert--warning[dir=rtl],[dir=rtl] .fd-alert--error,[dir=rtl] .fd-alert--information,[dir=rtl] .fd-alert--success,[dir=rtl] .fd-alert--warning{padding-right:2.5rem;padding-left:1rem}.fd-alert--error[dir=rtl]:after,.fd-alert--error[dir=rtl]:before,.fd-alert--information[dir=rtl]:after,.fd-alert--information[dir=rtl]:before,.fd-alert--success[dir=rtl]:after,.fd-alert--success[dir=rtl]:before,.fd-alert--warning[dir=rtl]:after,.fd-alert--warning[dir=rtl]:before,[dir=rtl] .fd-alert--error:after,[dir=rtl] .fd-alert--error:before,[dir=rtl] .fd-alert--information:after,[dir=rtl] .fd-alert--information:before,[dir=rtl] .fd-alert--success:after,[dir=rtl] .fd-alert--success:before,[dir=rtl] .fd-alert--warning:after,[dir=rtl] .fd-alert--warning:before{display:none}.fd-alert--error.fd-alert--dismissible,.fd-alert--information.fd-alert--dismissible,.fd-alert--success.fd-alert--dismissible,.fd-alert--warning.fd-alert--dismissible{padding-right:2.5rem}.fd-alert--error.fd-alert--dismissible[dir=rtl],.fd-alert--information.fd-alert--dismissible[dir=rtl],.fd-alert--success.fd-alert--dismissible[dir=rtl],.fd-alert--warning.fd-alert--dismissible[dir=rtl],[dir=rtl] .fd-alert--error.fd-alert--dismissible,[dir=rtl] .fd-alert--information.fd-alert--dismissible,[dir=rtl] .fd-alert--success.fd-alert--dismissible,[dir=rtl] .fd-alert--warning.fd-alert--dismissible{padding-left:2.5rem}.fd-alert--error[dir=rtl]:after,.fd-alert--information[dir=rtl]:after,.fd-alert--success[dir=rtl]:after,.fd-alert--warning[dir=rtl]:after,[dir=rtl] .fd-alert--error:after,[dir=rtl] .fd-alert--information:after,[dir=rtl] .fd-alert--success:after,[dir=rtl] .fd-alert--warning:after{text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;position:absolute;right:0}.fd-alert--warning{background-color:#fef7f1;background-color:var(--sapWarningBackground,#fef7f1);border-color:#e9730c;border-color:var(--sapWarningBorderColor,#e9730c)}.fd-alert--warning:after,.fd-alert--warning:before{color:#e9730c;color:var(--sapWarningBorderColor,#e9730c)}.fd-alert--warning:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE201\";font-size:16px;line-height:1}.fd-alert--warning[dir=rtl]:after,[dir=rtl] .fd-alert--warning:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE201\";font-size:16px;line-height:1}.fd-alert--error{background-color:#ffebeb;background-color:var(--sapErrorBackground,#ffebeb);border-color:#b00;border-color:var(--sapErrorBorderColor,#b00)}.fd-alert--error:after,.fd-alert--error:before{color:#b00;color:var(--sapErrorBorderColor,#b00)}.fd-alert--error:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE0B1\";font-size:16px;line-height:1}.fd-alert--error[dir=rtl]:after,[dir=rtl] .fd-alert--error:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE0B1\";font-size:16px;line-height:1}.fd-alert--success{background-color:#f1fdf6;background-color:var(--sapSuccessBackground,#f1fdf6);border-color:#107e3e;border-color:var(--sapSuccessBorderColor,#107e3e)}.fd-alert--success:after,.fd-alert--success:before{color:#107e3e;color:var(--sapSuccessBorderColor,#107e3e)}.fd-alert--success:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE203\";font-size:16px;line-height:1}.fd-alert--success[dir=rtl]:after,[dir=rtl] .fd-alert--success:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE203\";font-size:16px;line-height:1}.fd-alert--information{background-color:#f5faff;background-color:var(--sapInformationBackground,#f5faff);border-color:#0a6ed1;border-color:var(--sapInformationBorderColor,#0a6ed1)}.fd-alert--information:after,.fd-alert--information:before{color:#0a6ed1;color:var(--sapInformationBorderColor,#0a6ed1)}.fd-alert--information:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE202\";font-size:16px;line-height:1}.fd-alert--information[dir=rtl]:after,[dir=rtl] .fd-alert--information:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE202\";font-size:16px;line-height:1}.fd-alert--no-icon{padding-left:1rem}.fd-alert--no-icon:before{display:none}.fd-alert--no-icon[dir=rtl],[dir=rtl] .fd-alert--no-icon{padding-right:1rem}.fd-alert--no-icon[dir=rtl]:after,[dir=rtl] .fd-alert--no-icon:after{display:none}.fd-alert__text{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;line-height:1rem}.fd-alert__text:after,.fd-alert__text:before{box-sizing:inherit;font-size:inherit}.fd-alert,.fd-has-display-block{display:block}.fd-has-display-none{display:none}.fd-alert__close{border:1px solid transparent}"]
                }] }
    ];
    /** @nocollapse */
    AlertComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: ComponentFactoryResolver },
        { type: NgZone },
        { type: AlertRef, decorators: [{ type: Optional }] }
    ]; };
    AlertComponent.propDecorators = {
        containerRef: [{ type: ViewChild, args: ['container', { read: ViewContainerRef, static: false },] }],
        dismissible: [{ type: Input }],
        type: [{ type: Input }],
        id: [{ type: Input }],
        duration: [{ type: Input }],
        mousePersist: [{ type: Input }],
        ariaLabelledBy: [{ type: Input }],
        ariaLabel: [{ type: Input }],
        dismissLabel: [{ type: Input }],
        width: [{ type: Input }],
        minWidth: [{ type: Input }],
        message: [{ type: Input }],
        onDismiss: [{ type: Output }],
        handleAlertMouseEvent: [{ type: HostListener, args: ['mouseenter', ['$event'],] }, { type: HostListener, args: ['mouseleave', ['$event'],] }]
    };
    return AlertComponent;
}(AbstractFdNgxClass));
export { AlertComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    AlertComponent.prototype.containerRef;
    /**
     * Whether the alert is dismissible.
     * @type {?}
     */
    AlertComponent.prototype.dismissible;
    /**
     * The type of the alert. Can be one of *warning*, *success*, *information*, *error* or null.
     * @type {?}
     */
    AlertComponent.prototype.type;
    /**
     * Id for the alert component. If omitted, a unique one is generated.
     * @type {?}
     */
    AlertComponent.prototype.id;
    /**
     * Duration of time *in milliseconds* that the alert will be visible. Set to -1 for indefinite.
     * @type {?}
     */
    AlertComponent.prototype.duration;
    /**
     * Whether the alert should stay open if the mouse is hovering over it.
     * @type {?}
     */
    AlertComponent.prototype.mousePersist;
    /**
     * Id of the element that labels the alert.
     * @type {?}
     */
    AlertComponent.prototype.ariaLabelledBy;
    /**
     * Aria label for the alert component element.
     * @type {?}
     */
    AlertComponent.prototype.ariaLabel;
    /**
     * Aria label for the dismiss button.
     * @type {?}
     */
    AlertComponent.prototype.dismissLabel;
    /**
     * Width of the alert.
     * @type {?}
     */
    AlertComponent.prototype.width;
    /**
     * Minimum width of the alert.
     * @type {?}
     */
    AlertComponent.prototype.minWidth;
    /**
     * Alternative way of passing in a message to the alert.
     * @type {?}
     */
    AlertComponent.prototype.message;
    /**
     * Event fired when the alert is dismissed.
     * @type {?}
     */
    AlertComponent.prototype.onDismiss;
    /**
     * @hidden
     * @type {?}
     */
    AlertComponent.prototype.mouseInAlert;
    /**
     * @hidden
     * @type {?}
     */
    AlertComponent.prototype.componentRef;
    /**
     * @hidden
     * @type {?}
     */
    AlertComponent.prototype.childComponentType;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.cdRef;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.alertRef;
}
