// @fundamental-ngx/core@0.13.2/esm5/lib/shellbar/shellbar-action/shellbar-action.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shellbar/shellbar-action/shellbar-action.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * The component that represents a shellbar action.
 * ```html
 *  <fd-shellbar-action *ngFor="let action of actions"
 *                      [glyph]="action.glyph"
 *                      [callback]="action.callback"
 *                      [label]="action.label"
 *                      [notificationCount]="action.notificationCount"
 *                      [notificationLabel]="action.notificationLabel">
 *  </fd-shellbar-action>
 * ```
 */
var ShellbarActionComponent = /** @class */ (function () {
    function ShellbarActionComponent() {
    }
    ShellbarActionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-shellbar-action',
                    template: "<div class=\"fd-shellbar__action fd-shellbar__action--desktop\">\n    <button fd-button class=\"fd-shellbar__button\" (click)=\"callback ? callback($event) : ''\"\n            [glyph]=\"glyph\">\n        <span *ngIf=\"notificationCount\" class=\"fd-counter fd-counter--notification\" [attr.aria-label]=\"notificationLabel\">{{notificationCount}}</span>\n    </button>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    ShellbarActionComponent.propDecorators = {
        glyph: [{ type: Input }],
        callback: [{ type: Input }],
        label: [{ type: Input }],
        notificationLabel: [{ type: Input }],
        notificationCount: [{ type: Input }]
    };
    return ShellbarActionComponent;
}());
export { ShellbarActionComponent };
if (false) {
    /**
     * The glyph (icon) name
     * @type {?}
     */
    ShellbarActionComponent.prototype.glyph;
    /**
     * Callback that hanldles the response to clicks on any of the actions.
     * @type {?}
     */
    ShellbarActionComponent.prototype.callback;
    /**
     * The action label.
     * @type {?}
     */
    ShellbarActionComponent.prototype.label;
    /**
     * The notification label.
     * @type {?}
     */
    ShellbarActionComponent.prototype.notificationLabel;
    /**
     * Represents the number of notifications.
     * @type {?}
     */
    ShellbarActionComponent.prototype.notificationCount;
}
