// @fundamental-ngx/core@0.13.2/esm5/lib/panel/panel-footer/panel-footer.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/panel/panel-footer/panel-footer.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Panel footer can be utilized for pagination, secondary actions, add more data, etc.
 *
 * ```html
 * <fd-panel>
 *     <fd-panel-footer>
 *         Some text can go here!
 *     </fd-panel-footer>
 * </fd-panel>
 * ```
 */
var PanelFooterComponent = /** @class */ (function () {
    function PanelFooterComponent() {
        /**
         * @hidden
         */
        this.fdPanelFooterClass = true;
    }
    PanelFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-panel-footer',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PanelFooterComponent.propDecorators = {
        fdPanelFooterClass: [{ type: HostBinding, args: ['class.fd-panel__footer',] }]
    };
    return PanelFooterComponent;
}());
export { PanelFooterComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PanelFooterComponent.prototype.fdPanelFooterClass;
}
