// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification/notification.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification/notification.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ComponentFactoryResolver, ElementRef, HostListener, Input, Optional, TemplateRef, Type, ViewChild, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { NotificationRef } from '../notification-utils/notification-ref';
import { DefaultNotificationComponent } from '../notification-utils/default-notification/default-notification.component';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
var NotificationComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NotificationComponent, _super);
    function NotificationComponent(elRef, componentFactoryResolver, cdRef, notificationRef) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.cdRef = cdRef;
        _this.notificationRef = notificationRef;
        _this.escKeyCloseable = true;
        _this.focusTrapped = true;
        _this.ariaLabelledBy = null;
        _this.ariaLabel = null;
        _this.ariaDescribedBy = null;
        _this.backdropClickCloseable = true;
        _this.hasBackdrop = true;
        _this.notificationPanelClass = '';
        return _this;
    }
    /**
     * @return {?}
     */
    NotificationComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.childComponentType) {
            if (this.childComponentType instanceof Type) {
                this.loadFromComponent(this.childComponentType);
            }
            else if (this.childComponentType instanceof TemplateRef) {
                this.loadFromTemplate(this.childComponentType);
            }
            else {
                this.createFromDefaultConfiguration(this.childComponentType);
            }
        }
        this.cdRef.detectChanges();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NotificationComponent.prototype.closeNotificationEsc = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.escKeyCloseable && event.key === 'Escape') {
            this.notificationRef.dismiss('escape');
        }
    };
    /**
     * @private
     * @param {?} conf
     * @return {?}
     */
    NotificationComponent.prototype.createFromDefaultConfiguration = /**
     * @private
     * @param {?} conf
     * @return {?}
     */
    function (conf) {
        this.containerRef.clear();
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(DefaultNotificationComponent);
        this.componentRef = this.containerRef.createComponent(componentFactory);
        this.componentRef.instance.defaultConfigurationNotification = conf;
        this.componentRef.instance.type = this.type;
    };
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    NotificationComponent.prototype.loadFromComponent = /**
     * @private
     * @param {?} content
     * @return {?}
     */
    function (content) {
        this.containerRef.clear();
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(content);
        this.componentRef = this.containerRef.createComponent(componentFactory);
    };
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    NotificationComponent.prototype.loadFromTemplate = /**
     * @private
     * @param {?} content
     * @return {?}
     */
    function (content) {
        this.containerRef.clear();
        /** @type {?} */
        var context = {
            $implicit: this.notificationRef
        };
        this.componentRef = this.containerRef.createEmbeddedView(content, context);
    };
    /**
     * @return {?}
     */
    NotificationComponent.prototype._setProperties = /**
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-notification');
        this._addClassToElement('fd-notification-custom-block');
        if (this.type) {
            this._addClassToElement('fd-notification--' + this.type);
        }
        if (this.size) {
            this._addClassToElement('fd-notification--' + this.size);
        }
    };
    NotificationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification',
                    template: "<ng-content select=\"fd-notification-header\" #element></ng-content>\n<ng-content select=\"fd-notification-body\"></ng-content>\n<ng-content select=\"fd-notification-footer\"></ng-content>\n<ng-content select=\"fd-notification-default\"></ng-content>\n<ng-container #vc></ng-container>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[attr.aria-labelledby]': 'ariaLabelledBy',
                        '[attr.aria-label]': 'ariaLabel',
                        'role': 'notification',
                        '[attr.id]': 'id',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n */.fd-notification{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 .5rem;border:0;max-width:60rem;background-color:#fff;background-color:var(--sapList_Background,#fff);box-shadow:0 0 0 .0625rem rgba(0,0,0,.1),0 .125rem .5rem 0 rgba(0,0,0,.1);box-shadow:var(--sapContent_Shadow0,0 0 0 .0625rem rgba(0,0,0,.1),0 .125rem .5rem 0 rgba(0,0,0,.1));border-radius:.25rem;border-left:.375rem solid #6a6d70;border-left-color:var(--sapNeutralBorderColor,#6a6d70)}.fd-notification:after,.fd-notification:before{box-sizing:inherit;font-size:inherit}.fd-notification[dir=rtl],[dir=rtl] .fd-notification{border-left:none;border-right:.375rem solid #6a6d70;border-right-color:var(--sapNeutralBorderColor,#6a6d70)}.fd-notification__header{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:flex;box-shadow:inset 0 -.0625rem #d9d9d9;box-shadow:inset 0 -.0625rem var(--sapPageHeader_BorderColor,#d9d9d9)}.fd-notification__header:after,.fd-notification__header:before{box-sizing:inherit;font-size:inherit}.fd-notification__indicator{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-notification__indicator:after,.fd-notification__indicator:before{box-sizing:inherit;font-size:inherit}.fd-notification__indicator--error{margin:1px 0 0;padding:1rem .5rem 0 1rem;color:#b00;color:var(--sapErrorColor,#b00)}.fd-notification__indicator--error:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE0B1\";font-size:16px;line-height:1}.fd-notification__indicator--success{margin:1px 0 0;padding:1rem .5rem 0 1rem;color:#107e3e;color:var(--sapSuccessColor,#107e3e)}.fd-notification__indicator--success:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE203\";font-size:16px;line-height:1}.fd-notification__indicator--warning{margin:1px 0 0;padding:1rem .5rem 0 1rem;color:#e9730c;color:var(--sapWarningColor,#e9730c)}.fd-notification__indicator--warning:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE201\";font-size:16px;line-height:1}.fd-notification__indicator--information{margin:1px 0 0;padding:1rem .5rem 0 1rem;color:#0a6ed1;color:var(--sapInformationColor,#0a6ed1)}.fd-notification__indicator--information:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE202\";font-size:16px;line-height:1}.fd-notification__title{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:1rem 0;margin:0 1rem 0 0;border:0;color:#32363a;color:var(--sapGroup_TitleTextColor,#32363a);font-weight:700;width:100%}.fd-notification__title:after,.fd-notification__title:before{box-sizing:inherit;font-size:inherit}.fd-notification__title[dir=rtl],[dir=rtl] .fd-notification__title{margin-right:0;margin-left:1rem}.fd-notification__title:first-child{padding-left:1rem}.fd-notification__title:first-child[dir=rtl],[dir=rtl] .fd-notification__title:first-child{padding-left:0;padding-right:1rem}.fd-notification__close{color:#32363a;color:var(--sapGroup_TitleTextColor,#32363a);position:relative;top:.45rem;right:.35rem;min-width:36px;min-height:36px;width:36px;height:36px}.fd-notification__close:hover{color:#32363a;color:var(--sapGroup_TitleTextColor,#32363a);border:none;background-color:#fff;background-color:var(--sapList_Background,#fff)}.fd-notification__close[dir=rtl],[dir=rtl] .fd-notification__close{right:-.35rem}.fd-notification__body{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:flex;align-items:flex-end;width:100%}.fd-notification__body:after,.fd-notification__body:before{box-sizing:inherit;font-size:inherit}.fd-notification__content{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:1rem;margin:0;border:0;display:flex;width:60%}.fd-notification__content:after,.fd-notification__content:before{box-sizing:inherit;font-size:inherit}.fd-notification__avatar{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 1rem 0 0;border:0;min-width:2rem;min-height:2rem}.fd-notification__avatar:after,.fd-notification__avatar:before{box-sizing:inherit;font-size:inherit}.fd-notification__avatar[dir=rtl],[dir=rtl] .fd-notification__avatar{margin-right:0;margin-left:1rem}.fd-notification__text{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 1rem 0 0;margin:0;border:0;width:100%;display:flex;flex-direction:column}.fd-notification__text:after,.fd-notification__text:before{box-sizing:inherit;font-size:inherit}.fd-notification__description{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 1em 0 0;margin:0 -1em 0 0;border:0;position:relative;line-height:1.45em;max-height:2.9em;overflow:hidden;color:#6a6d70;color:var(--sapContent_LabelColor,#6a6d70)}.fd-notification__description:after,.fd-notification__description:before{box-sizing:inherit;font-size:inherit}.fd-notification__description:before{position:absolute;content:\"...\";right:0;bottom:1px}.fd-notification__description:after{content:\"\";position:absolute;right:0;width:1em;height:1em;margin-top:.2em;background:#fff;background:var(--sapList_Background,#fff)}.fd-notification__metadata{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:.375rem 0 0;border:0;color:#6a6d70;color:var(--sapContent_LabelColor,#6a6d70)}.fd-notification__metadata:after,.fd-notification__metadata:before{box-sizing:inherit;font-size:inherit}.fd-notification__footer{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 .5rem .5rem 1rem;margin:0;border:0;width:40%;display:flex;align-items:center;justify-content:space-between}.fd-notification__footer:after,.fd-notification__footer:before{box-sizing:inherit;font-size:inherit}.fd-notification__footer>:not(:last-child){margin:.25rem}.fd-notification__actions{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 0 auto;border:0;display:flex}.fd-notification__actions:after,.fd-notification__actions:before{box-sizing:inherit;font-size:inherit}.fd-notification__actions>*{margin:.25rem}.fd-notification__actions[dir=rtl],[dir=rtl] .fd-notification__actions{margin-left:0}.fd-notification--error{border-left-color:#b00;border-left-color:var(--sapErrorBorderColor,#b00)}.fd-notification--error[dir=rtl],[dir=rtl] .fd-notification--error{border-right-color:#b00;border-right-color:var(--sapErrorBorderColor,#b00)}.fd-notification--success{border-left-color:#107e3e;border-left-color:var(--sapSuccessBorderColor,#107e3e)}.fd-notification--success[dir=rtl],[dir=rtl] .fd-notification--success{border-right-color:#107e3e;border-right-color:var(--sapSuccessBorderColor,#107e3e)}.fd-notification--warning{border-left-color:#e9730c;border-left-color:var(--sapWarningBorderColor,#e9730c)}.fd-notification--warning[dir=rtl],[dir=rtl] .fd-notification--warning{border-right-color:#e9730c;border-right-color:var(--sapWarningBorderColor,#e9730c)}.fd-notification--information{border-left-color:#0a6ed1;border-left-color:var(--sapInformationBorderColor,#0a6ed1)}.fd-notification--information[dir=rtl],[dir=rtl] .fd-notification--information{border-right-color:#0a6ed1;border-right-color:var(--sapInformationBorderColor,#0a6ed1)}.fd-notification--m .fd-notification__body,.fd-notification--s .fd-notification__body{flex-direction:column;align-items:flex-start;width:100%}.fd-notification--m .fd-notification__content,.fd-notification--s .fd-notification__content{width:100%;padding:1rem 1rem .5rem}.fd-notification--m .fd-notification__content:last-child,.fd-notification--s .fd-notification__content:last-child{padding-bottom:1rem}.fd-notification--m .fd-notification__footer,.fd-notification--s .fd-notification__footer{width:100%}.fd-notification--m{max-width:640px;min-width:320px}.fd-notification--s{width:320px}.fd-notification--group>*{border-radius:0}.fd-notification--group>:not(:first-child) .fd-notification__header{box-shadow:none}.fd-notification--group .fd-notification{margin-bottom:0}.fd-notification--group .fd-notification:first-child{border-radius:.25rem .25rem 0 0}.fd-notification--group .fd-notification:last-child{border-radius:0 0 .25rem .25rem}.fd-notification--group .fd-notification:only-child{border-radius:.25rem}.fd-notification--group .fd-notification__content{padding-top:.75rem;padding-bottom:.5rem}.fd-notification--group .fd-notification__title{padding:.75rem 0}.fd-notification--group .fd-notification__close{top:.1875rem;right:.1875rem}.fd-notification--group[dir=rtl] .fd-notification__close,[dir=rtl] .fd-notification--group .fd-notification__close{right:-.1875rem}.fd-notification--group .fd-notification__indicator--error,.fd-notification--group .fd-notification__indicator--information,.fd-notification--group .fd-notification__indicator--success,.fd-notification--group .fd-notification__indicator--warning{padding-top:.75rem;padding-bottom:.75rem}.fd-notification-custom-block{display:block}"]
                }] }
    ];
    /** @nocollapse */
    NotificationComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ComponentFactoryResolver },
        { type: ChangeDetectorRef },
        { type: NotificationRef, decorators: [{ type: Optional }] }
    ]; };
    NotificationComponent.propDecorators = {
        size: [{ type: Input }],
        type: [{ type: Input }],
        containerRef: [{ type: ViewChild, args: ['vc', { read: ViewContainerRef, static: false },] }],
        closeNotificationEsc: [{ type: HostListener, args: ['keyup', ['$event'],] }]
    };
    return NotificationComponent;
}(AbstractFdNgxClass));
export { NotificationComponent };
if (false) {
    /**
     * Size of notification, defined by user, s or m
     * @type {?}
     */
    NotificationComponent.prototype.size;
    /**
     * Type of Notification 'success' | 'warning' | 'information' | 'error'
     * @type {?}
     */
    NotificationComponent.prototype.type;
    /** @type {?} */
    NotificationComponent.prototype.containerRef;
    /** @type {?} */
    NotificationComponent.prototype.id;
    /** @type {?} */
    NotificationComponent.prototype.escKeyCloseable;
    /** @type {?} */
    NotificationComponent.prototype.focusTrapped;
    /** @type {?} */
    NotificationComponent.prototype.ariaLabelledBy;
    /** @type {?} */
    NotificationComponent.prototype.defaultNotificationConfiguration;
    /** @type {?} */
    NotificationComponent.prototype.ariaLabel;
    /** @type {?} */
    NotificationComponent.prototype.ariaDescribedBy;
    /** @type {?} */
    NotificationComponent.prototype.childComponentType;
    /** @type {?} */
    NotificationComponent.prototype.backdropClickCloseable;
    /** @type {?} */
    NotificationComponent.prototype.hasBackdrop;
    /** @type {?} */
    NotificationComponent.prototype.notificationPanelClass;
    /** @type {?} */
    NotificationComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    NotificationComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    NotificationComponent.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    NotificationComponent.prototype.cdRef;
    /**
     * @type {?}
     * @private
     */
    NotificationComponent.prototype.notificationRef;
}
