// @fundamental-ngx/core@0.13.2/esm2015/lib/panel/panel-body/panel-body.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/panel/panel-body/panel-body.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation } from '@angular/core';
/**
 * Main content of the panel can that hold lists, table, tree, text, form or any other information.
 *
 * ```html
 * <fd-panel>
 *     <fd-panel-body>
 *         Some text can go here!
 *     </fd-panel-body>
 * </fd-panel>
 * ```
 */
export class PanelBodyComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdPanelBodyClass = true;
        /**
         * Whether the edges of the panel should have bleeding padding.
         */
        this.bleed = false;
    }
}
PanelBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-panel-body',
                template: "<ng-content></ng-content>\n",
                host: {
                    '[class.fd-has-display-block]': 'true'
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PanelBodyComponent.propDecorators = {
    fdPanelBodyClass: [{ type: HostBinding, args: ['class.fd-panel__body',] }],
    bleed: [{ type: Input }, { type: HostBinding, args: ['class.fd-panel__body--bleed',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PanelBodyComponent.prototype.fdPanelBodyClass;
    /**
     * Whether the edges of the panel should have bleeding padding.
     * @type {?}
     */
    PanelBodyComponent.prototype.bleed;
}
