// @fundamental-ngx/core@0.13.2/esm2015/lib/form/fieldset/fieldset.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/fieldset/fieldset.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Used for easily displaying forms with a margin. Not necessary for fundamental forms to be functional.
 *
 * ```html
 * <div fd-fieldset
 *     <div fd-form-item>
 *         ...
 *     </div>
 * </div>
 * ```
 */
export class FieldsetComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdFieldClass = true;
    }
}
FieldsetComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: '[fd-fieldset]',
                template: `<ng-content></ng-content>`,
                encapsulation: ViewEncapsulation.None,
                styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-group\n*/.fd-fieldset,.fd-form-group{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 16px;border:0}.fd-fieldset:after,.fd-fieldset:before,.fd-form-group:after,.fd-form-group:before{box-sizing:inherit;font-size:inherit}.fd-fieldset:last-child,.fd-form-group:last-child{margin-bottom:0}.fd-form-group--inline{display:flex;justify-content:flex-start}.fd-form-group--inline :not(:last-child){margin-right:16px}.fd-form-group__item{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-form-group__item:after,.fd-form-group__item:before{box-sizing:inherit;font-size:inherit}/*!\n.fd-fieldset\n*/.fd-fieldset__legend{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 8px;border:0;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);line-height:16px;display:block}.fd-fieldset__legend:after,.fd-fieldset__legend:before{box-sizing:inherit;font-size:inherit}.fd-fieldset__legend.is-disabled{opacity:.4}"]
            }] }
];
FieldsetComponent.propDecorators = {
    fdFieldClass: [{ type: HostBinding, args: ['class.fd-fieldset',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    FieldsetComponent.prototype.fdFieldClass;
}
