// @fundamental-ngx/core@0.13.2/esm2015/lib/panel/panel-head/panel-head.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/panel/panel-head/panel-head.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Container for title and description.
 *
 * ```html
 * <fd-panel>
 *     <fd-panel-header>
 *         <fd-panel-head>
 *             <fd-panel-title>Title!</fd-panel-title>
 *         </fd-panel-head>
 *     </fd-panel-header>
 * </fd-panel>
 * ```
 */
export class PanelHeadComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdPanelHeadClass = true;
    }
}
PanelHeadComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-panel-head',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PanelHeadComponent.propDecorators = {
    fdPanelHeadClass: [{ type: HostBinding, args: ['class.fd-panel__head',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PanelHeadComponent.prototype.fdPanelHeadClass;
}
