// @fundamental-ngx/core@0.13.2/esm2015/lib/modal/modal-header/modal-header.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-header/modal-header.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
/**
 * Applies fundamental layout and styling to the contents of a modal header.
 *
 * ```html
 * <fd-modal-header>
 *     <h1 fd-modal-title>Title</h1>
 *     <button fd-modal-close-btn></button>
 * </fd-modal-header>
 * ```
 */
export class ModalHeaderComponent {
    constructor() {
        /**
         * @hidden
         */
        this.modalHeader = true;
    }
}
ModalHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-modal-header',
                template: "<ng-content></ng-content>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [':host {display: block;}']
            }] }
];
ModalHeaderComponent.propDecorators = {
    modalHeader: [{ type: HostBinding, args: ['class.fd-modal__header',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalHeaderComponent.prototype.modalHeader;
}
