// @fundamental-ngx/core@0.13.2/esm2015/lib/alert/alert-utils/alert-container.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/alert/alert-utils/alert-container.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
import { alertContainerNgIf } from './alert-animations';
export class AlertContainerComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdAlertContainerClass = true;
    }
}
AlertContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-alert-container',
                template: ``,
                host: {
                    '[@alertContainerNgIf]': ''
                },
                animations: [
                    alertContainerNgIf
                ],
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [`
        .fd-alert-container {
            position: fixed;
            display: flex;
            flex-direction: column;
            z-index: 5000;
            align-items: center;
            top: 0;
            right: 50%;
            left: 50%;
        }
    `]
            }] }
];
AlertContainerComponent.propDecorators = {
    fdAlertContainerClass: [{ type: HostBinding, args: ['class.fd-alert-container',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    AlertContainerComponent.prototype.fdAlertContainerClass;
}
