// @fundamental-ngx/core@0.13.2/esm5/lib/loading-spinner/loading-spinner.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/loading-spinner/loading-spinner.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * The component that represents a loading spinner.
 *
 * ```html
 * <fd-loading-spinner [loading]="true"></fd-loading-spinner>
 * ```
 */
var LoadingSpinnerComponent = /** @class */ (function () {
    function LoadingSpinnerComponent() {
        /**
         * Whether to display the loading indicator animation.
         */
        this.loading = false;
        /**
         * Aria label for the 'loading' spinner.
         */
        this.loadingLabel = 'Loading';
    }
    LoadingSpinnerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-loading-spinner',
                    template: "<div class=\"fd-spinner\" *ngIf=\"loading\" [attr.aria-label]=\"loadingLabel\">\n    <div class=\"fd-spinner__body\"></div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-spinner+(.is-hidden|[aria-hidden=true])\n\nInspired by line scale spinner from Load Awesome v1.1.0 (http:// github.danielcardoso.net/load-awesome/)\n* Copyright 2015 Daniel Cardoso <@DanielCardoso>\n* Licensed under MIT\n*/.fd-spinner{--fd-spinner-background-color:var(--fd-color-action-1);font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 auto;border:0;position:relative;width:29px;height:40px;display:flex;justify-content:space-between}.fd-spinner.is-hidden{display:none}.is-busy{position:relative;min-height:40px}.is-busy:before{position:absolute;background-color:#edeff0;background-color:var(--fd-color-background-1,#edeff0);opacity:.95;top:0;right:0;bottom:0;left:0;z-index:1;content:\"\"}.is-busy .fd-spinner{position:absolute;z-index:2;left:calc(50% - 14.5px);top:calc(50% - 20px)}.fd-spinner:after,.fd-spinner:before{box-sizing:inherit;font-size:inherit;content:\"\";position:relative;width:5px;height:100%;background-color:#0a6ed1;background-color:var(--fd-spinner-background-color,#0a6ed1)}.fd-spinner:before{-webkit-animation:1s -1s infinite line-scale;animation:1s -1s infinite line-scale}.fd-spinner .fd-spinner__body{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-spinner .fd-spinner__body:after,.fd-spinner .fd-spinner__body:before{box-sizing:inherit;font-size:inherit;content:\"\";position:absolute;width:5px;height:100%;background-color:#0a6ed1;background-color:var(--fd-spinner-background-color,#0a6ed1)}.fd-spinner .fd-spinner__body:before{left:8px;-webkit-animation:1s -.9s infinite line-scale;animation:1s -.9s infinite line-scale}.fd-spinner .fd-spinner__body:after{right:8px;-webkit-animation:1s -.8s infinite line-scale;animation:1s -.8s infinite line-scale}.fd-spinner:after{-webkit-animation:1s -.7s infinite line-scale;animation:1s -.7s infinite line-scale}@-webkit-keyframes line-scale{0%,40%,to{transform:scaleY(.4)}80%{transform:scaleY(1)}}@keyframes line-scale{0%,40%,to{transform:scaleY(.4)}80%{transform:scaleY(1)}}"]
                }] }
    ];
    LoadingSpinnerComponent.propDecorators = {
        loading: [{ type: Input }],
        loadingLabel: [{ type: Input }]
    };
    return LoadingSpinnerComponent;
}());
export { LoadingSpinnerComponent };
if (false) {
    /**
     * Whether to display the loading indicator animation.
     * @type {?}
     */
    LoadingSpinnerComponent.prototype.loading;
    /**
     * Aria label for the 'loading' spinner.
     * @type {?}
     */
    LoadingSpinnerComponent.prototype.loadingLabel;
}
