// @fundamental-ngx/core@0.13.2/esm5/lib/popover/popover.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/popover/popover.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, ViewEncapsulation, ContentChild, ChangeDetectionStrategy, HostBinding } from '@angular/core';
import { PopoverDirective } from './popover-directive/popover.directive';
import { PopoverDropdownComponent } from './popover-dropdown/popover-dropdown.component';
/** @type {?} */
var popoverUniqueId = 0;
/**
 * The popover is a wrapping component that accepts a *control* as well as a *body*.
 * The control is what will trigger the opening of the actual popover, which is called the body.
 * By default, popovers are triggered by click. This can be customized through the *triggers* input.
 * PopoverComponent is an abstraction of PopoverDirective.
 */
var PopoverComponent = /** @class */ (function () {
    function PopoverComponent() {
        /**
         * Whether the popover should have an arrow.
         */
        this.noArrow = true;
        /**
         * Whether the popover is disabled.
         */
        this.disabled = false;
        /**
         * Whether the popover should be treated as a dropdown.
         */
        this.isDropdown = false;
        /**
         * The trigger events that will open/close the popover.
         *  Accepts any [HTML DOM Events](https://www.w3schools.com/jsref/dom_obj_event.asp).
         */
        this.triggers = ['click'];
        /**
         * Whether the popover is open. Can be used through two-way binding.
         */
        this.isOpen = false;
        /**
         * List of additional classes that will be added to popover container element
         */
        this.additionalClasses = [];
        /**
         * The Popper.js options to attach to this popover.
         * See the [Popper.js Documentation](https://popper.js.org/popper-documentation.html) for details.
         */
        this.options = {
            placement: 'bottom-start',
            modifiers: {
                preventOverflow: {
                    enabled: true,
                    escapeWithReference: true,
                    boundariesElement: 'scrollParent'
                }
            }
        };
        /**
         * Whether the popover should be focusTrapped.
         */
        this.focusTrapped = false;
        /**
         * Whether the popover should close when a click is made outside its boundaries.
         */
        this.closeOnOutsideClick = true;
        /**
         * Whether the popover should close when the escape key is pressed.
         */
        this.closeOnEscapeKey = true;
        /**
         * Event emitted when the state of the isOpen property changes.
         */
        this.isOpenChange = new EventEmitter();
        /**
         * Id of the popover. If none is provided, one will be generated.
         */
        this.id = 'fd-popover-' + popoverUniqueId++;
    }
    /**
     * Toggles the popover open state.
     */
    /**
     * Toggles the popover open state.
     * @return {?}
     */
    PopoverComponent.prototype.toggle = /**
     * Toggles the popover open state.
     * @return {?}
     */
    function () {
        if (this.isOpen) {
            this.close();
        }
        else {
            this.open();
        }
    };
    /**
     * Closes the popover.
     */
    /**
     * Closes the popover.
     * @return {?}
     */
    PopoverComponent.prototype.close = /**
     * Closes the popover.
     * @return {?}
     */
    function () {
        if (this.isOpen) {
            this.isOpen = false;
            this.isOpenChange.emit(this.isOpen);
        }
    };
    /**
     * Opens the popover.
     */
    /**
     * Opens the popover.
     * @return {?}
     */
    PopoverComponent.prototype.open = /**
     * Opens the popover.
     * @return {?}
     */
    function () {
        if (!this.isOpen) {
            this.isOpen = true;
            this.isOpenChange.emit(this.isOpen);
        }
    };
    /**
     * Forces an update of the popover's positioning calculation.
     */
    /**
     * Forces an update of the popover's positioning calculation.
     * @return {?}
     */
    PopoverComponent.prototype.updatePopover = /**
     * Forces an update of the popover's positioning calculation.
     * @return {?}
     */
    function () {
        this.directiveRef.updatePopper();
    };
    /**
     * Function is called every time popover changes open attribute
     */
    /**
     * Function is called every time popover changes open attribute
     * @param {?} isOpen
     * @return {?}
     */
    PopoverComponent.prototype.openChanged = /**
     * Function is called every time popover changes open attribute
     * @param {?} isOpen
     * @return {?}
     */
    function (isOpen) {
        this.isOpenChange.emit(isOpen);
        this.updateDropdownIsOpen(isOpen);
    };
    /** Method that is called, when there is keydown event dispatched */
    /**
     * Method that is called, when there is keydown event dispatched
     * @param {?} event
     * @return {?}
     */
    PopoverComponent.prototype.handleKeydown = /**
     * Method that is called, when there is keydown event dispatched
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.key === 'ArrowDown' && event.altKey) {
            this.open();
        }
    };
    /** @hidden
     *  Function that allows us to control aria-expanded on dropdown child
     * */
    /**
     * @hidden
     *  Function that allows us to control aria-expanded on dropdown child
     *
     * @private
     * @param {?} isOpen
     * @return {?}
     */
    PopoverComponent.prototype.updateDropdownIsOpen = /**
     * @hidden
     *  Function that allows us to control aria-expanded on dropdown child
     *
     * @private
     * @param {?} isOpen
     * @return {?}
     */
    function (isOpen) {
        if (this.dropdownComponent) {
            this.dropdownComponent.isOpen = isOpen;
        }
    };
    PopoverComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-popover',
                    template: "<div #popoverContainer [ngClass]=\"addContainerClass ? addContainerClass : ''\">\n    <div\n        class=\"fd-popover__control\"\n        [attr.aria-expanded]=\"this.disabled ? false : isOpen\"\n        [attr.aria-disabled]=\"this.disabled\"\n        aria-haspopup=\"true\"\n        (keydown)=\"handleKeydown($event)\"\n        [fdPopover]=\"popoverBody\"\n        [additionalClasses]=\"additionalClasses\"\n        [(isOpen)]=\"isOpen\"\n        (isOpenChange)=\"openChanged($event)\"\n        [noArrow]=\"noArrow\"\n        [disabled]=\"disabled\"\n        [triggers]=\"triggers\"\n        [placement]=\"placement\"\n        [focusTrapped]=\"focusTrapped\"\n        [options]=\"options\"\n        [fillControlMode]=\"fillControlMode\"\n        [closeOnOutsideClick]=\"closeOnOutsideClick\"\n        [closeOnEscapeKey]=\"closeOnEscapeKey\"\n        [appendTo]=\"appendTo ? appendTo : popoverContainer\"\n    >\n        <ng-content select=\"fd-popover-control\"></ng-content>\n        <ng-content select=\"fd-dropdown-control\"></ng-content>\n    </div>\n    <ng-template #popoverBody>\n        <ng-container *ngTemplateOutlet=\"popoverBodyTpl\"></ng-container>\n    </ng-template>\n</div>\n\n<ng-template #popoverBodyTpl>\n    <div [attr.aria-hidden]=\"disabled ? true : !isOpen\">\n        <ng-content select=\"fd-popover-body\"></ng-content>\n    </div>\n</ng-template>\n",
                    host: {
                        '[class.fd-popover-custom]': 'true',
                        '[attr.id]': 'id'
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n */.fd-popover{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;position:relative;display:inline-block}.fd-popover:after,.fd-popover:before{box-sizing:inherit;font-size:inherit}.fd-popover__control{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;position:relative;cursor:pointer}.fd-popover__control:after,.fd-popover__control:before{box-sizing:inherit;font-size:inherit}.fd-popover__control.fd-input:last-child{margin-bottom:0}.fd-popover__body{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;border:none;position:absolute;white-space:nowrap;box-shadow:0 0 0 .0625rem rgba(0,0,0,.15),0 .625rem 1.875rem 0 rgba(0,0,0,.3);box-shadow:var(--sapContent_Shadow2,0 0 0 .0625rem rgba(0,0,0,.15),0 .625rem 1.875rem 0 rgba(0,0,0,.3));z-index:1000;border-radius:.25rem;border-radius:var(--sapElement_BorderCornerRadius,.25rem);background:#fff;background:var(--sapGroup_ContentBackground,#fff);opacity:1;visibility:visible;left:.0625rem;left:var(--sapGroup_BorderWidth,.0625rem);top:calc(100% + .5rem);transition:125ms;transform:translateY(0)}.fd-popover__body:after,.fd-popover__body:before{box-sizing:inherit;font-size:inherit}.fd-popover__body[dir=rtl],[dir=rtl] .fd-popover__body{left:auto;right:.0625rem;right:var(--sapGroup_BorderWidth,.0625rem)}.fd-popover__body[dir=rtl]:after,.fd-popover__body[dir=rtl]:before,[dir=rtl] .fd-popover__body:after,[dir=rtl] .fd-popover__body:before{right:.5rem}.fd-popover__body:before{height:0;width:0;border-style:solid;border-width:0 .5rem .5rem;border-bottom-color:#fff;border-bottom-color:var(--sapGroup_ContentBackground,#fff);border-left-color:transparent;border-right-color:transparent;content:\"\";position:absolute;top:calc(-.5rem + 1px);left:.5rem;z-index:1002}.fd-popover__body:after{height:0;width:0;border-style:solid;border-width:0 .5rem .5rem;border-bottom-color:#d9d9d9;border-bottom-color:var(--sapGroup_ContentBorderColor,#d9d9d9);border-left-color:transparent;border-right-color:transparent;content:\"\";position:absolute;top:-.5rem;left:.5rem;z-index:1001}.fd-popover__body-header{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 .5rem;margin:0;border:0;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:2.75rem;line-height:2.75rem;width:auto;background-color:#fff;background-color:var(--sapPageHeader_Background,#fff);box-shadow:0 0 .25rem 0 rgba(0,0,0,.15),inset 0 -.0625rem 0 0 #d9d9d9;box-shadow:var(--sapContent_HeaderShadow,0 0 .25rem 0 rgba(0,0,0,.15),inset 0 -.0625rem 0 0 #d9d9d9);border-radius:.25rem .25rem 0 0;border-radius:var(--sapElement_BorderCornerRadius,.25rem) var(--sapElement_BorderCornerRadius,.25rem) 0 0}.fd-popover__body-header:after,.fd-popover__body-header:before{box-sizing:inherit;font-size:inherit}.fd-popover__body-header--compact{height:2.5rem;line-height:2.5rem}.fd-popover__body-header--with-subheader{box-shadow:none}.fd-popover__body-subheader{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 .5rem;margin:0;border:0;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:2.75rem;line-height:2.75rem;width:auto;background-color:#fff;background-color:var(--sapPageHeader_Background,#fff);box-shadow:0 0 .25rem 0 rgba(0,0,0,.15),inset 0 -.0625rem 0 0 #d9d9d9;box-shadow:var(--sapContent_HeaderShadow,0 0 .25rem 0 rgba(0,0,0,.15),inset 0 -.0625rem 0 0 #d9d9d9);border-radius:0}.fd-popover__body-subheader:after,.fd-popover__body-subheader:before{box-sizing:inherit;font-size:inherit}.fd-popover__body-subheader--compact{height:2rem;line-height:2rem}.fd-popover__body-footer{font-size:.875rem;font-size:var(--sapFontSize,.875rem);color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 .5rem;margin:0;border:0;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;height:2.75rem;line-height:2.75rem;width:auto;min-width:4rem;background-color:#fff;background-color:var(--sapPageFooter_Background,#fff);border-top:.0625rem solid #d9d9d9;border-top:.0625rem solid var(--sapPageFooter_BorderColor,#d9d9d9);border-radius:0 0 .25rem .25rem;border-radius:0 0 var(--sapElement_BorderCornerRadius,.25rem) var(--sapElement_BorderCornerRadius,.25rem)}.fd-popover__body-footer:after,.fd-popover__body-footer:before{box-sizing:inherit;font-size:inherit}.fd-popover__body-footer--compact{height:2.5rem;line-height:2.5rem}.fd-popover__body--left{right:auto;left:.0625rem;left:var(--sapGroup_BorderWidth,.0625rem)}.fd-popover__body--left[dir=rtl],[dir=rtl] .fd-popover__body--left{right:.0625rem;right:var(--sapGroup_BorderWidth,.0625rem);left:auto}.fd-popover__body--left[dir=rtl]:after,.fd-popover__body--left[dir=rtl]:before,[dir=rtl] .fd-popover__body--left:after,[dir=rtl] .fd-popover__body--left:before{right:.5rem;left:unset}.fd-popover__body--left:after,.fd-popover__body--left:before{left:.5rem}.fd-popover__body--right{left:auto;right:.0625rem;right:var(--sapGroup_BorderWidth,.0625rem)}.fd-popover__body--right[dir=rtl],[dir=rtl] .fd-popover__body--right{right:auto;left:.0625rem;left:var(--sapGroup_BorderWidth,.0625rem)}.fd-popover__body--right[dir=rtl]:after,.fd-popover__body--right[dir=rtl]:before,[dir=rtl] .fd-popover__body--right:after,[dir=rtl] .fd-popover__body--right:before{left:.5rem;right:unset}.fd-popover__body--right:after,.fd-popover__body--right:before{left:auto;right:.5rem}.fd-popover__body--no-arrow{box-shadow:0 0 0 .0625rem rgba(0,0,0,.15),0 .125rem .5rem 0 rgba(0,0,0,.3);box-shadow:var(--sapContent_Shadow1,0 0 0 .0625rem rgba(0,0,0,.15),0 .125rem .5rem 0 rgba(0,0,0,.3));top:100%}.fd-popover__body--no-arrow:after,.fd-popover__body--no-arrow:before{display:none}.fd-popover__body--no-arrow.fd-popover__body--left{left:.0625rem;left:var(--sapGroup_BorderWidth,.0625rem);right:auto}.fd-popover__body--no-arrow.fd-popover__body--left[dir=rtl],.fd-popover__body--no-arrow.fd-popover__body--right,[dir=rtl] .fd-popover__body--no-arrow.fd-popover__body--left{right:.0625rem;right:var(--sapGroup_BorderWidth,.0625rem);left:auto}.fd-popover__body--no-arrow.fd-popover__body--right[dir=rtl],[dir=rtl] .fd-popover__body--no-arrow.fd-popover__body--right{left:.0625rem;left:var(--sapGroup_BorderWidth,.0625rem);right:auto}.fd-popover__body.is-hidden,.fd-popover__body[aria-hidden=true]{z-index:1;opacity:0;visibility:hidden}.fd-popover__popper{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;border:none;position:absolute;white-space:nowrap;box-shadow:0 0 0 .0625rem rgba(0,0,0,.15),0 .625rem 1.875rem 0 rgba(0,0,0,.3);box-shadow:var(--sapContent_Shadow2,0 0 0 .0625rem rgba(0,0,0,.15),0 .625rem 1.875rem 0 rgba(0,0,0,.3));z-index:1000;border-radius:.25rem;border-radius:var(--sapElement_BorderCornerRadius,.25rem);background:#fff;background:var(--sapGroup_ContentBackground,#fff);opacity:1}.fd-popover__popper:after,.fd-popover__popper:before{box-sizing:inherit;font-size:inherit}.fd-popover__popper--no-arrow{margin:0!important}.fd-popover__popper--no-arrow .fd-popover__arrow{display:none}.fd-popover__popper .fd-popover__arrow{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0 solid transparent;position:absolute;width:0;height:0}.fd-popover__popper .fd-popover__arrow:after,.fd-popover__popper .fd-popover__arrow:before{box-sizing:inherit;font-size:inherit}.fd-popover__popper .fd-popover__arrow:after{position:absolute;content:\"\";border-style:solid;border-color:transparent}.fd-popover__popper[x-out-of-boundaries]{display:none}.fd-popover__popper[x-placement^=top]{margin-bottom:.5rem}.fd-popover__popper[x-placement^=top] .fd-popover__arrow{bottom:-.5rem;margin:0 .5rem;border-width:.5rem .5rem 0;border-top-color:#d9d9d9;border-top-color:var(--sapGroup_ContentBorderColor,#d9d9d9)}.fd-popover__popper[x-placement^=top] .fd-popover__arrow:after{bottom:1px;margin:0 -.5rem;border-width:.5rem .5rem 0;border-top-color:#fff;border-top-color:var(--sapGroup_ContentBackground,#fff)}.fd-popover__popper[x-placement^=bottom]{margin-top:.5rem}.fd-popover__popper[x-placement^=bottom] .fd-popover__arrow{top:-.5rem;margin:0 .5rem;border-width:0 .5rem .5rem;border-bottom-color:#d9d9d9;border-bottom-color:var(--sapGroup_ContentBorderColor,#d9d9d9)}.fd-popover__popper[x-placement^=bottom] .fd-popover__arrow:after{top:1px;margin:0 -.5rem;border-width:0 .5rem .5rem;border-bottom-color:#fff;border-bottom-color:var(--sapGroup_ContentBackground,#fff)}.fd-popover__popper[x-placement^=left]{margin-right:.5rem}.fd-popover__popper[x-placement^=left] .fd-popover__arrow{right:-.5rem;margin:.5rem 0;border-width:.5rem 0 .5rem .5rem;border-left-color:#d9d9d9;border-left-color:var(--sapGroup_ContentBorderColor,#d9d9d9)}.fd-popover__popper[x-placement^=left] .fd-popover__arrow:after{right:1px;margin:-.5rem 0;border-width:.5rem 0 .5rem .5rem;border-left-color:#fff;border-left-color:var(--sapGroup_ContentBackground,#fff)}.fd-popover__popper[x-placement^=right]{margin-left:.5rem}.fd-popover__popper[x-placement^=right] .fd-popover__arrow{left:-.5rem;margin:.5rem 0;border-width:.5rem .5rem .5rem 0;border-right-color:#d9d9d9;border-right-color:var(--sapGroup_ContentBorderColor,#d9d9d9)}.fd-popover__popper[x-placement^=right] .fd-popover__arrow:after{left:1px;margin:-.5rem 0;border-width:.5rem .5rem .5rem 0;border-right-color:#fff;border-right-color:var(--sapGroup_ContentBackground,#fff)}/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-dropdown+()\n    .fd-dropdown__control+([disabled])\n*/.fd-dropdown{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;position:relative;display:inline-block}.fd-dropdown:after,.fd-dropdown:before{box-sizing:inherit;font-size:inherit}.fd-dropdown__control{position:relative;margin-left:0;width:100%;text-align:left}.fd-dropdown__control[dir=rtl]:after,[dir=rtl] .fd-dropdown__control:after{border-radius:4px 0 0 4px}.fd-dropdown__control[dir=rtl]:not(.fd-button--standard):after,[dir=rtl] .fd-dropdown__control:not(.fd-button--standard):after{border-right-color:var(--fd-color-neutral-4,#89919a);border-left:none;border-right:1px solid #89919a}.fd-dropdown__control[dir=rtl].fd-button,[dir=rtl] .fd-dropdown__control.fd-button{padding-left:0;padding-right:12px}.fd-dropdown__control:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;font-size:14px;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE1E2\";transition:transform 125ms ease-in;color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1);width:34px;height:34px;margin-top:-3px;padding-top:10px;border-left-color:var(--fd-color-neutral-4,#89919a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);border-top-right-radius:4px;border-bottom-right-radius:4px;border-left:1px solid #89919a}.fd-dropdown__control--no-border{border:none}.fd-dropdown__control.fd-button--xs:after{height:21.5px;padding-top:5px}.fd-dropdown__control.fd-button--s:after{font-size:14px;line-height:1.42857;margin-top:-3px;height:26px;padding-top:4.8px}.fd-dropdown__control.fd-button--compact:after{font-size:14px;line-height:1.42857;height:26px;padding-top:3px}.fd-dropdown__control.fd-button--l:after{font-size:14px;line-height:1.42857;margin-top:-3px;width:38px;height:38px;padding-top:10px}.fd-dropdown__control.fd-button{color:#32363a;color:var(--fd-color-text-1,#32363a);border-color:#89919a;border-color:var(--fd-color-neutral-4,#89919a);padding-right:0}.fd-dropdown__control.fd-button .is-hover,.fd-dropdown__control.fd-button:hover{color:#32363a;color:var(--fd-color-text-1,#32363a);background-color:#fff;background-color:var(--fd-color-background-2,#fff);border-color:#085caf;border-color:var(--fd-color-action-hover,#085caf);box-shadow:none}.fd-dropdown__control.fd-button.is-active,.fd-dropdown__control.fd-button:active,.fd-dropdown__control.fd-button[aria-expanded=true]{color:#32363a;color:var(--fd-color-text-1,#32363a);background-color:#fff;background-color:var(--fd-color-background-2,#fff);box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-hover,#085caf)}.fd-dropdown__control.fd-button--standard.is-active,.fd-dropdown__control.fd-button--standard:active,.fd-dropdown__control.fd-button--standard:hover{color:#51555a;color:var(--fd-color-text-2,#51555a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);border-color:#0a7e3e;border-color:var(--fd-color-status-4,#0a7e3e);box-shadow:none}.fd-dropdown__control.fd-button--standard:after{width:auto;margin-right:10px;color:#32363a;color:var(--fd-color-text-1,#32363a);background-color:transparent;border-left:none}.fd-dropdown__control.fd-button--standard[dir=rtl],[dir=rtl] .fd-dropdown__control.fd-button--standard{padding-left:12px}.fd-dropdown__group,.fd-dropdown__list{list-style:none;margin-bottom:0}.fd-dropdown__title{display:block;margin-bottom:0;padding:8px 24px;font-size:12px;line-height:1.33333;font-weight:400;text-transform:uppercase;color:#6a6d70;background:#fafafa}.fd-dropdown__item{display:block;padding:10px 24px;color:#51555a}.fd-dropdown__item:hover{color:#51555a;background-color:#fafafa}.fd-dropdown__item.is-selected,.fd-dropdown__item[aria-selected=true]{color:#51555a;background-color:rgba(10,110,209,.07)}.fd-popover-custom{margin-right:0;display:inline-block}.fd-popover-custom .fd-dropdown{position:static}.fd-popover-custom--disabled{pointer-events:none}fd-popover-body .fd-menu{width:auto;display:block}"]
                }] }
    ];
    PopoverComponent.propDecorators = {
        directiveRef: [{ type: ViewChild, args: [PopoverDirective, { static: false },] }],
        dropdownComponent: [{ type: ContentChild, args: [PopoverDropdownComponent, { static: false },] }],
        noArrow: [{ type: Input }],
        addContainerClass: [{ type: Input }],
        disabled: [{ type: Input }, { type: HostBinding, args: ['class.fd-popover-custom--disabled',] }],
        isDropdown: [{ type: Input }],
        appendTo: [{ type: Input }],
        triggers: [{ type: Input }],
        placement: [{ type: Input }],
        isOpen: [{ type: Input }],
        additionalClasses: [{ type: Input }],
        options: [{ type: Input }],
        focusTrapped: [{ type: Input }],
        fillControlMode: [{ type: Input }],
        closeOnOutsideClick: [{ type: Input }],
        closeOnEscapeKey: [{ type: Input }],
        isOpenChange: [{ type: Output }],
        id: [{ type: Input }]
    };
    return PopoverComponent;
}());
export { PopoverComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PopoverComponent.prototype.directiveRef;
    /**
     * @hidden
     * @type {?}
     */
    PopoverComponent.prototype.dropdownComponent;
    /**
     * Whether the popover should have an arrow.
     * @type {?}
     */
    PopoverComponent.prototype.noArrow;
    /**
     * Whether the popover container needs an extra class for styling.
     * @type {?}
     */
    PopoverComponent.prototype.addContainerClass;
    /**
     * Whether the popover is disabled.
     * @type {?}
     */
    PopoverComponent.prototype.disabled;
    /**
     * Whether the popover should be treated as a dropdown.
     * @type {?}
     */
    PopoverComponent.prototype.isDropdown;
    /**
     * The element to which the popover should be appended.
     * @type {?}
     */
    PopoverComponent.prototype.appendTo;
    /**
     * The trigger events that will open/close the popover.
     *  Accepts any [HTML DOM Events](https://www.w3schools.com/jsref/dom_obj_event.asp).
     * @type {?}
     */
    PopoverComponent.prototype.triggers;
    /**
     * The placement of the popover. It can be one of: top, top-start, top-end, bottom,
     *  bottom-start, bottom-end, right, right-start, right-end, left, left-start, left-end.
     * @type {?}
     */
    PopoverComponent.prototype.placement;
    /**
     * Whether the popover is open. Can be used through two-way binding.
     * @type {?}
     */
    PopoverComponent.prototype.isOpen;
    /**
     * List of additional classes that will be added to popover container element
     * @type {?}
     */
    PopoverComponent.prototype.additionalClasses;
    /**
     * The Popper.js options to attach to this popover.
     * See the [Popper.js Documentation](https://popper.js.org/popper-documentation.html) for details.
     * @type {?}
     */
    PopoverComponent.prototype.options;
    /**
     * Whether the popover should be focusTrapped.
     * @type {?}
     */
    PopoverComponent.prototype.focusTrapped;
    /**
     * Preset options for the popover body width.
     * * `at-least` will apply a minimum width to the body equivalent to the width of the control.
     * * `equal` will apply a width to the body equivalent to the width of the control.
     * * Leave blank for no effect.
     * @type {?}
     */
    PopoverComponent.prototype.fillControlMode;
    /**
     * Whether the popover should close when a click is made outside its boundaries.
     * @type {?}
     */
    PopoverComponent.prototype.closeOnOutsideClick;
    /**
     * Whether the popover should close when the escape key is pressed.
     * @type {?}
     */
    PopoverComponent.prototype.closeOnEscapeKey;
    /**
     * Event emitted when the state of the isOpen property changes.
     * @type {?}
     */
    PopoverComponent.prototype.isOpenChange;
    /**
     * Id of the popover. If none is provided, one will be generated.
     * @type {?}
     */
    PopoverComponent.prototype.id;
}
