// @fundamental-ngx/core@0.13.2/esm2015/lib/modal/modal-utils/modal-container.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-utils/modal-container.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { modalFadeNgIf } from './modal-animations';
export class ModalContainer {
}
ModalContainer.decorators = [
    { type: Component, args: [{
                selector: 'fd-modal-container',
                template: ``,
                host: {
                    '[@modal-fade]': '',
                    '[class.fd-modal-container]': 'true'
                },
                animations: [
                    modalFadeNgIf
                ],
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [`
        .fd-modal-container {
            position: fixed;
            top: 0;
            bottom: 0;
            right: 0;
            left: 0;
            display: flex;
            align-items: center;
            justify-content: center;
            z-index: 1000;
        }
    `]
            }] }
];
