// @fundamental-ngx/core@0.13.2/esm2015/lib/modal/modal-utils/modal-directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-utils/modal-directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding } from '@angular/core';
/**
 * Directive that applies fundamental modal styling to a header.
 *
 * ```html
 * <h1 fd-modal-title>Modal Title</h1>
 * ```
 */
export class ModalTitleDirective {
    constructor() {
        /**
         * @hidden
         */
        this.modalTitle = true;
    }
}
ModalTitleDirective.decorators = [
    { type: Directive, args: [{
                // TODO to be discussed
                // tslint:disable-next-line:directive-selector
                selector: '[fd-modal-title]'
            },] }
];
ModalTitleDirective.propDecorators = {
    modalTitle: [{ type: HostBinding, args: ['class.fd-modal__title',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalTitleDirective.prototype.modalTitle;
}
/**
 * Directive that applies fundamental modal styling to a button.
 *
 * ```html
 * <button fd-modal-close-btn></button>
 * ```
 */
export class ModalCloseButtonDirective {
    constructor() {
        /**
         * @hidden
         */
        this.ariaLabel = 'close';
        /**
         * @hidden
         */
        this.sapIconDeclineClass = true;
        /**
         * @hidden
         */
        this.sapIconLClass = true;
        /**
         * @hidden
         */
        this.lightButtonClass = true;
        /**
         * @hidden
         */
        this.modalCloseClass = true;
    }
}
ModalCloseButtonDirective.decorators = [
    { type: Directive, args: [{
                // TODO to be discussed
                // tslint:disable-next-line:directive-selector
                selector: '[fd-modal-close-btn]'
            },] }
];
ModalCloseButtonDirective.propDecorators = {
    ariaLabel: [{ type: HostBinding, args: ['attr.aria-label',] }],
    sapIconDeclineClass: [{ type: HostBinding, args: ['class.sap-icon--decline',] }],
    sapIconLClass: [{ type: HostBinding, args: ['class.sap-icon--l',] }],
    lightButtonClass: [{ type: HostBinding, args: ['class.fd-button--light',] }],
    modalCloseClass: [{ type: HostBinding, args: ['class.fd-modal__close',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalCloseButtonDirective.prototype.ariaLabel;
    /**
     * @hidden
     * @type {?}
     */
    ModalCloseButtonDirective.prototype.sapIconDeclineClass;
    /**
     * @hidden
     * @type {?}
     */
    ModalCloseButtonDirective.prototype.sapIconLClass;
    /**
     * @hidden
     * @type {?}
     */
    ModalCloseButtonDirective.prototype.lightButtonClass;
    /**
     * @hidden
     * @type {?}
     */
    ModalCloseButtonDirective.prototype.modalCloseClass;
}
