// @fundamental-ngx/core@0.13.2/esm5/lib/panel/panel-filters/panel-filters.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/panel/panel-filters/panel-filters.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Panel level filters that is specific to the data being displayed within the panel.
 *
 * ```html
 * <fd-panel>
 *     <fd-panel-filters>
 *         Some text can go here!
 *     </fd-panel-filters>
 * </fd-panel>
 * ```
 */
var PanelFiltersComponent = /** @class */ (function () {
    function PanelFiltersComponent() {
        /**
         * @hidden
         */
        this.fdPanelFiltersClass = true;
    }
    PanelFiltersComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-panel-filters',
                    template: "<ng-content></ng-content>\n",
                    host: {
                        '[class.fd-has-display-block]': 'true'
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PanelFiltersComponent.propDecorators = {
        fdPanelFiltersClass: [{ type: HostBinding, args: ['class.fd-panel__filters',] }]
    };
    return PanelFiltersComponent;
}());
export { PanelFiltersComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    PanelFiltersComponent.prototype.fdPanelFiltersClass;
}
