// @fundamental-ngx/core@0.13.2/esm5/lib/modal/modal-body/modal-body.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-body/modal-body.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
/**
 * Applies fundamental layout and styling to the contents of a modal body.
 *
 * ```html
 * <fd-modal-body>
 *     <div>Modal body content</div>
 * </fd-modal-body>
 * ```
 */
var ModalBodyComponent = /** @class */ (function () {
    function ModalBodyComponent() {
        /**
         * @hidden
         */
        this.modalBody = true;
    }
    ModalBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-modal-body',
                    template: "<ng-content></ng-content>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        :host {\n            display: block;\n            overflow: auto;\n            flex-grow: 1;\n        }\n    "]
                }] }
    ];
    ModalBodyComponent.propDecorators = {
        modalBody: [{ type: HostBinding, args: ['class.fd-modal__body',] }]
    };
    return ModalBodyComponent;
}());
export { ModalBodyComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalBodyComponent.prototype.modalBody;
}
