// @fundamental-ngx/core@0.13.2/esm5/lib/token/token.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/token/token.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Output, ViewChild, ViewEncapsulation, Input, ChangeDetectionStrategy } from '@angular/core';
/**
 * A token is used to represent contextualizing information.
 * They can be useful to show applied filters, selected values for form fields or object metadata.
 */
var TokenComponent = /** @class */ (function () {
    /** @hidden */
    function TokenComponent(elRef) {
        this.elRef = elRef;
        /**
         * Whether the token is disabled.
         */
        this.disabled = false;
        /**
         * Emitted when the *x* icon is clicked. Specifically, any pseudo-element.
         */
        this.onCloseClick = new EventEmitter();
    }
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    TokenComponent.prototype.clickHandler = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.contentContainer && !this.disabled) {
            if (this.elRef.nativeElement.contains(event.target) && !this.contentContainer.nativeElement.contains(event.target)) {
                this.onCloseClick.emit();
            }
        }
    };
    TokenComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-token',
                    template: "<span class=\"fd-token-content\" #contentContainer>\n    <ng-content></ng-content>\n</span>\n\n",
                    host: {
                        class: 'fd-token',
                        '[class.fd-token__disabled]': 'disabled',
                        'role': 'button'
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-token+(--no-border)\n    .fd-token__content+()\n    .fd-token__title+()\n*/.fd-token{--fd-token-color:var(--fd-color-text-2, #51555a);--fd-token-background-color:var(--fd-color-neutral-1, #fafafa);--fd-token-border-color:var(--fd-color-neutral-3, #d9d9d9);font-size:var(--sapFontSize,.875rem);color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:1px solid #d9d9d9;font-size:12px;font-weight:400;color:#51555a;color:var(--fd-token-color,#51555a);background-color:#fafafa;background-color:var(--fd-token-background-color,#fafafa);line-height:24px;vertical-align:middle;display:inline-block;border-radius:4px;border-color:var(--fd-token-border-color,#d9d9d9)}.fd-token:after,.fd-token:before{box-sizing:inherit;font-size:inherit}.fd-token:after{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;display:inline-block;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE03E\";font-size:12px;color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1);margin-left:8px;vertical-align:bottom;line-height:24px;cursor:pointer;padding-right:8px}.fd-token[dir=rtl]:before,[dir=rtl] .fd-token:before{font-family:SAP-icons;font-style:normal;font-weight:400;text-align:center;text-decoration:inherit;text-transform:none;text-rendering:optimizeLegibility;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;speak:none;content:\"\uE03E\";font-size:12px;color:#0a6ed1;color:var(--fd-color-action-1,#0a6ed1);vertical-align:bottom;line-height:24px;cursor:pointer;display:inline;margin-left:0;margin-right:8px}.fd-token[dir=rtl]:after,[dir=rtl] .fd-token:after{content:none}.fd-token-content{display:inline-block;padding-left:8px}.fd-token__disabled::after{cursor:not-allowed}"]
                }] }
    ];
    /** @nocollapse */
    TokenComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    TokenComponent.propDecorators = {
        contentContainer: [{ type: ViewChild, args: ['contentContainer', { static: false },] }],
        disabled: [{ type: Input }],
        onCloseClick: [{ type: Output }],
        clickHandler: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return TokenComponent;
}());
export { TokenComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    TokenComponent.prototype.contentContainer;
    /**
     * Whether the token is disabled.
     * @type {?}
     */
    TokenComponent.prototype.disabled;
    /**
     * Emitted when the *x* icon is clicked. Specifically, any pseudo-element.
     * @type {?}
     */
    TokenComponent.prototype.onCloseClick;
    /**
     * @type {?}
     * @private
     */
    TokenComponent.prototype.elRef;
}
