// @fundamental-ngx/core@0.13.2/esm5/lib/form/form-message/form-message.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/form-message/form-message.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * Form message. Intended to be displayed with a form control for validation purposes.
 */
var FormMessageComponent = /** @class */ (function () {
    function FormMessageComponent() {
        /**
         * Whether to apply compact mode to the message.
         */
        this.compact = false;
    }
    FormMessageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-form-message',
                    template: "<span\n    class=\"fd-form-message\"\n    [ngClass]=\"[type ? ' fd-form-message--' + type : '', compact ? ' fd-form-message--' + compact : '']\"\n>\n    <ng-content></ng-content>\n</span>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-message(--help, --error, --warning)\n*/.fd-form-message{font-size:var(--sapFontSize,.875rem);line-height:1.42857;font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:.5rem;margin:-.25rem 0 0;border:0;z-index:3;clear:both;height:auto;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;min-width:6rem;max-width:22rem;font-size:.75rem;font-size:var(--sapFontSmallSize,.75rem);font-weight:400;line-height:normal;color:#32363a;color:var(--sapTextColor,#32363a);overflow:hidden;text-overflow:ellipsis;border-radius:0!important;white-space:pre-line!important;display:block}.fd-form-message:after,.fd-form-message:before{box-sizing:inherit;font-size:inherit}.fd-form-message[dir=rtl],[dir=rtl] .fd-form-message{right:0}.fd-form-message--compact{margin-top:-.1875rem}.fd-form-message--error{background-color:#ffebeb;background-color:var(--sapErrorBackground,#ffebeb)}.fd-form-message--warning{background-color:#fef7f1;background-color:var(--sapWarningBackground,#fef7f1)}.fd-form-message--success{background-color:#f1fdf6;background-color:var(--sapSuccessBackground,#f1fdf6)}.fd-form-message--information{background-color:#f5faff;background-color:var(--sapInformationBackground,#f5faff)}.fd-form-message+.fd-input{margin-top:0}"]
                }] }
    ];
    FormMessageComponent.propDecorators = {
        type: [{ type: Input }],
        compact: [{ type: Input }]
    };
    return FormMessageComponent;
}());
export { FormMessageComponent };
if (false) {
    /**
     * Type of the message. Can be 'success' | 'error' | 'warning' | 'information'
     * @type {?}
     */
    FormMessageComponent.prototype.type;
    /**
     * Whether to apply compact mode to the message.
     * @type {?}
     */
    FormMessageComponent.prototype.compact;
}
