// @fundamental-ngx/core@0.13.2/esm2015/lib/modal/modal-body/modal-body.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-body/modal-body.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
/**
 * Applies fundamental layout and styling to the contents of a modal body.
 *
 * ```html
 * <fd-modal-body>
 *     <div>Modal body content</div>
 * </fd-modal-body>
 * ```
 */
export class ModalBodyComponent {
    constructor() {
        /**
         * @hidden
         */
        this.modalBody = true;
    }
}
ModalBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-modal-body',
                template: "<ng-content></ng-content>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [`
        :host {
            display: block;
            overflow: auto;
            flex-grow: 1;
        }
    `]
            }] }
];
ModalBodyComponent.propDecorators = {
    modalBody: [{ type: HostBinding, args: ['class.fd-modal__body',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalBodyComponent.prototype.modalBody;
}
