// @fundamental-ngx/core@0.13.2/esm5/lib/localizator-editor/localization-editor.directives.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/localizator-editor/localization-editor.directives.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, Input, TemplateRef, ViewContainerRef } from '@angular/core';
/**
 * Directive which is used along with input elements, inside the localization editor item or main.
 *  ```html
 *  <fd-localization-editor-item>
 *      <input fd-localization-editor-input type="text" placeholder="EN">
 *  </fd-localization-editor-item>
 *  ```
 */
var LocalizationEditorInputDirective = /** @class */ (function () {
    function LocalizationEditorInputDirective() {
        /**
         * @hidden
         */
        this.fdInputGroupInputClass = true;
    }
    LocalizationEditorInputDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-localization-editor-input]'
                },] }
    ];
    LocalizationEditorInputDirective.propDecorators = {
        fdInputGroupInputClass: [{ type: HostBinding, args: ['class.fd-input-group__input',] }]
    };
    return LocalizationEditorInputDirective;
}());
export { LocalizationEditorInputDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorInputDirective.prototype.fdInputGroupInputClass;
}
/**
 * Directive which is used along with textarea elements, inside the localization editor item or main.
 *  ```html
 *  <fd-localization-editor-item>
 *      <textarea fd-localization-editor-input placeholder="EN">
 *  </fd-localization-editor-item>
 *  ```
 */
var LocalizationEditorTextareaDirective = /** @class */ (function () {
    function LocalizationEditorTextareaDirective() {
        /**
         * @hidden
         */
        this.fdInputGroupInputClass = true;
        /**
         * @hidden
         */
        this.fdLocalizationEditorTextareaClass = true;
    }
    LocalizationEditorTextareaDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-localization-editor-textarea]',
                },] }
    ];
    LocalizationEditorTextareaDirective.propDecorators = {
        compact: [{ type: HostBinding, args: ['class.fd-input--compact',] }],
        fdInputGroupInputClass: [{ type: HostBinding, args: ['class.fd-input-group__input',] }],
        fdLocalizationEditorTextareaClass: [{ type: HostBinding, args: ['class.fd-localization-editor-textarea',] }]
    };
    return LocalizationEditorTextareaDirective;
}());
export { LocalizationEditorTextareaDirective };
if (false) {
    /**
     * @hidden
     *  Variable is controlled by parent component and define if there should be used compact mode
     *
     * @type {?}
     */
    LocalizationEditorTextareaDirective.prototype.compact;
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorTextareaDirective.prototype.fdInputGroupInputClass;
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorTextareaDirective.prototype.fdLocalizationEditorTextareaClass;
}
/**
 * Directive which is used to add complex content, which will be displayed in the add-on space.
 *  ```html
 *  <fd-localization-editor-item>
 *      <ng-template fd-localization-editor-label>
 *          <fd-icon [glyph]="field.glyph"></fd-icon>
 *      </ng-template>
 *      <textarea fd-localization-editor-input placeholder="EN">
 *  </fd-localization-editor-item>
 *  ```
 *
 */
var LocalizationEditorLabel = /** @class */ (function () {
    function LocalizationEditorLabel() {
    }
    LocalizationEditorLabel.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-localization-editor-label]',
                },] }
    ];
    return LocalizationEditorLabel;
}());
export { LocalizationEditorLabel };
/**
 * Directive that is used to wrap whole localization field inside li element.
 *  ```html
 *  <li fd-localization-editor-element>
 *      <fd-localization-editor-item>
 *          <textarea fd-localization-editor-input placeholder="EN">
 *      </fd-localization-editor-item>
 *  </li>
 *  ```
 *
 */
var LocalizationEditorElement = /** @class */ (function () {
    function LocalizationEditorElement() {
        /**
         * @hidden
         */
        this.fdLocalizationEditorLanguage = true;
    }
    LocalizationEditorElement.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-localization-editor-element]',
                },] }
    ];
    LocalizationEditorElement.propDecorators = {
        fdLocalizationEditorLanguage: [{ type: HostBinding, args: ['class.fd-localization-editor__language',] }]
    };
    return LocalizationEditorElement;
}());
export { LocalizationEditorElement };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorElement.prototype.fdLocalizationEditorLanguage;
}
/**
 * Not for external use. Portal to render the complex title template.
 */
var LocalizationEditorLoadLabel = /** @class */ (function () {
    /** @hidden */
    function LocalizationEditorLoadLabel(viewRef) {
        this.viewRef = viewRef;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    LocalizationEditorLoadLabel.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.viewRef.clear();
        this.contentRef = this.viewRef.createEmbeddedView(this.content);
    };
    LocalizationEditorLoadLabel.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-localization-editor-load-label]'
                },] }
    ];
    /** @nocollapse */
    LocalizationEditorLoadLabel.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    LocalizationEditorLoadLabel.propDecorators = {
        content: [{ type: Input, args: ['fd-localization-editor-load-label',] }]
    };
    return LocalizationEditorLoadLabel;
}());
export { LocalizationEditorLoadLabel };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorLoadLabel.prototype.content;
    /**
     * @hidden
     * @type {?}
     * @private
     */
    LocalizationEditorLoadLabel.prototype.contentRef;
    /**
     * @type {?}
     * @private
     */
    LocalizationEditorLoadLabel.prototype.viewRef;
}
