// @fundamental-ngx/core@0.13.2/esm5/lib/form/form-group/form-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form/form-group/form-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation, ElementRef, Input } from '@angular/core';
import { applyCssClass } from '../../utils/public_api';
/**
 * Represents a form group element.
 *
 * ```html
 * <fd-form-group>
 *     <div fd-form-item [isCheck]="true">
 *         <input fd-form-control type="radio" checked>
 *         <fd-form-label>Option One</fd-form-label>
 *     </div>
 * </fd-form-group>
 * ```
 */
var FormGroupComponent = /** @class */ (function () {
    /** @hidden */
    function FormGroupComponent(_elementRef) {
        this._elementRef = _elementRef;
        /**
         * @hidden
         */
        this.fdFormGroupClass = true;
        this._isInline = false;
    }
    Object.defineProperty(FormGroupComponent.prototype, "isInline", {
        /** Determines if form items should be displayed inline or not
         * Default value is set to false;
         */
        set: /**
         * Determines if form items should be displayed inline or not
         * Default value is set to false;
         * @param {?} inline
         * @return {?}
         */
        function (inline) {
            this._isInline = inline;
            this.buildComponentCssClass();
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    FormGroupComponent.prototype.ngAfterViewInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.buildComponentCssClass();
    };
    /** This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with @applyCssClass decorator
     */
    /**
     * This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with \@applyCssClass decorator
     * @return {?}
     */
    FormGroupComponent.prototype.buildComponentCssClass = /**
     * This method is responsible for building a css class based on current state
     *  It is implementation of CssClassBuilder interface and
     *  should be used with \@applyCssClass decorator
     * @return {?}
     */
    function () {
        return this._isInline ? 'fd-form-group--inline' : '';
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    FormGroupComponent.prototype.elementRef = /**
     * @hidden
     * @return {?}
     */
    function () {
        return this._elementRef;
    };
    FormGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-form-group',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/*!\n * Fundamental Library Styles v0.4.1\n * Copyright (c) 2020 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-group\n*/.fd-form-group{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 16px;border:0}.fd-form-group:after,.fd-form-group:before{box-sizing:inherit;font-size:inherit}.fd-form-group:last-child{margin-bottom:0}.fd-form-group--inline{display:flex;justify-content:flex-start}.fd-form-group--inline :not(:last-child){margin-right:16px}.fd-form-group__item{font-size:.875rem;font-size:var(--sapFontSize,.875rem);line-height:1.42857;color:#32363a;color:var(--sapTextColor,#32363a);font-family:\"72\",\"72full\",Arial,Helvetica,sans-serif;font-family:var(--sapFontFamily, \"72\", \"72full\", Arial, Helvetica, sans-serif);font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0}.fd-form-group__item:after,.fd-form-group__item:before{box-sizing:inherit;font-size:inherit}"]
                }] }
    ];
    /** @nocollapse */
    FormGroupComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FormGroupComponent.propDecorators = {
        fdFormGroupClass: [{ type: HostBinding, args: ['class.fd-form-group',] }],
        isInline: [{ type: Input }]
    };
    tslib_1.__decorate([
        applyCssClass,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", String)
    ], FormGroupComponent.prototype, "buildComponentCssClass", null);
    return FormGroupComponent;
}());
export { FormGroupComponent };
if (false) {
    /** @type {?} */
    FormGroupComponent.prototype.class;
    /**
     * @hidden
     * @type {?}
     */
    FormGroupComponent.prototype.fdFormGroupClass;
    /**
     * @type {?}
     * @private
     */
    FormGroupComponent.prototype._isInline;
    /**
     * @type {?}
     * @private
     */
    FormGroupComponent.prototype._elementRef;
}
