// @fundamental-ngx/core@0.13.2/esm5/lib/modal/modal-footer/modal-footer.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-footer/modal-footer.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
/**
 * Applies fundamental layout and styling to the contents of a modal footer.
 *
 * ```html
 * <fd-modal-footer>
 *     <button>Do action</button>
 * </fd-modal-footer>
 * ```
 */
var ModalFooterComponent = /** @class */ (function () {
    function ModalFooterComponent() {
        /**
         * @hidden
         */
        this.modalFooter = true;
    }
    ModalFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-modal-footer',
                    template: "<div class=\"fd-modal__actions\">\n    <ng-content></ng-content>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        :host {\n            display: block;\n            border-top: 1px solid #eeeeef;\n        }\n    "]
                }] }
    ];
    ModalFooterComponent.propDecorators = {
        modalFooter: [{ type: HostBinding, args: ['class.fd-modal__footer',] }]
    };
    return ModalFooterComponent;
}());
export { ModalFooterComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ModalFooterComponent.prototype.modalFooter;
}
