// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification-body/notification-body.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-body/notification-body.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation } from '@angular/core';
var NotificationBodyComponent = /** @class */ (function () {
    function NotificationBodyComponent() {
        /**
         * @hidden
         */
        this.fdNotificationHeaderBody = true;
    }
    NotificationBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification-body',
                    template: "<ng-content select=\"[fd-notification-content]\"></ng-content>\n<ng-content select=\"fd-notification-footer\"></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [""]
                }] }
    ];
    NotificationBodyComponent.propDecorators = {
        fdNotificationHeaderBody: [{ type: HostBinding, args: ['class.fd-notification__body',] }]
    };
    return NotificationBodyComponent;
}());
export { NotificationBodyComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationBodyComponent.prototype.fdNotificationHeaderBody;
}
