// @fundamental-ngx/core@0.13.2/esm5/lib/modal/modal-utils/modal-config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-utils/modal-config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ModalConfig = /** @class */ (function () {
    function ModalConfig() {
        /**
         * Aria label for the modal component element.
         */
        this.ariaLabel = null;
        /**
         * Id of the element that labels the modal.
         */
        this.ariaLabelledBy = null;
        /**
         * Id of the element that describes the modal.
         */
        this.ariaDescribedBy = null;
        /**
         * Whether the modal should have a backdrop.
         */
        this.hasBackdrop = true;
        /**
         * Global classes to apply to the backdrop.
         */
        this.backdropClass = '';
        /**
         * Whether clicking on the backdrop should close the modal. Only works if hasBackdrop is true.
         */
        this.backdropClickCloseable = true;
        /**
         * Global classes to apply to the modal panel.
         */
        this.modalPanelClass = '';
        /**
         * Whether the escape key should close the modal.
         */
        this.escKeyCloseable = true;
        /**
         * Whether the modal should be focus trapped.
         */
        this.focusTrapped = true;
        /**
         * The container that the modal is appended to. By default, it is appended to the body.
         */
        this.container = 'body';
    }
    return ModalConfig;
}());
export { ModalConfig };
if (false) {
    /**
     * Id for the modal component. If omitted, a unique one is generated.
     * @type {?}
     */
    ModalConfig.prototype.id;
    /**
     * Width of the modal.
     * @type {?}
     */
    ModalConfig.prototype.width;
    /**
     * Height of the modal.
     * @type {?}
     */
    ModalConfig.prototype.height;
    /**
     * Minimum width of the modal.
     * @type {?}
     */
    ModalConfig.prototype.minWidth;
    /**
     * Minimum height of the modal.
     * @type {?}
     */
    ModalConfig.prototype.minHeight;
    /**
     * Maximum width of the modal.
     * @type {?}
     */
    ModalConfig.prototype.maxWidth;
    /**
     * Maximum height of the modal.
     * @type {?}
     */
    ModalConfig.prototype.maxHeight;
    /**
     * Position of the modal.
     * @type {?}
     */
    ModalConfig.prototype.position;
    /**
     * Aria label for the modal component element.
     * @type {?}
     */
    ModalConfig.prototype.ariaLabel;
    /**
     * Id of the element that labels the modal.
     * @type {?}
     */
    ModalConfig.prototype.ariaLabelledBy;
    /**
     * Id of the element that describes the modal.
     * @type {?}
     */
    ModalConfig.prototype.ariaDescribedBy;
    /**
     * Whether the modal should have a backdrop.
     * @type {?}
     */
    ModalConfig.prototype.hasBackdrop;
    /**
     * Global classes to apply to the backdrop.
     * @type {?}
     */
    ModalConfig.prototype.backdropClass;
    /**
     * Whether clicking on the backdrop should close the modal. Only works if hasBackdrop is true.
     * @type {?}
     */
    ModalConfig.prototype.backdropClickCloseable;
    /**
     * Global classes to apply to the modal panel.
     * @type {?}
     */
    ModalConfig.prototype.modalPanelClass;
    /**
     * Whether the escape key should close the modal.
     * @type {?}
     */
    ModalConfig.prototype.escKeyCloseable;
    /**
     * Whether the modal should be focus trapped.
     * @type {?}
     */
    ModalConfig.prototype.focusTrapped;
    /**
     * The container that the modal is appended to. By default, it is appended to the body.
     * @type {?}
     */
    ModalConfig.prototype.container;
    /**
     * Data to pass along to the content through the ModalRef.
     * @type {?}
     */
    ModalConfig.prototype.data;
}
