// @fundamental-ngx/core@0.13.2/esm5/lib/modal/modal-utils/modal-backdrop.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-utils/modal-backdrop.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, HostBinding, HostListener, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
import { modalFadeNgIf } from './modal-animations';
import { ModalRef } from './modal-ref';
var ModalBackdrop = /** @class */ (function (_super) {
    tslib_1.__extends(ModalBackdrop, _super);
    function ModalBackdrop(elRef, modalRef) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        _this.modalRef = modalRef;
        _this.backdropClass = '';
        _this.backdropClickCloseable = true;
        _this.overlayMain = true;
        _this.overlayModal = true;
        return _this;
    }
    /**
     * @return {?}
     */
    ModalBackdrop.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._setProperties();
    };
    /**
     * @return {?}
     */
    ModalBackdrop.prototype._setProperties = /**
     * @return {?}
     */
    function () {
        if (this.backdropClass) {
            this._addClassToElement(this.backdropClass);
        }
    };
    /**
     * @return {?}
     */
    ModalBackdrop.prototype.closeModal = /**
     * @return {?}
     */
    function () {
        if (this.backdropClickCloseable) {
            this.modalRef.dismiss('backdrop');
        }
    };
    ModalBackdrop.decorators = [
        { type: Component, args: [{
                    selector: 'fd-modal-overlay',
                    template: "",
                    host: {
                        'tabindex': '-1',
                        '[@modal-fade]': ''
                    },
                    animations: [
                        modalFadeNgIf
                    ],
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    ModalBackdrop.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ModalRef }
    ]; };
    ModalBackdrop.propDecorators = {
        overlayMain: [{ type: HostBinding, args: ['class.fd-overlay',] }],
        overlayModal: [{ type: HostBinding, args: ['class.fd-overlay--modal',] }],
        closeModal: [{ type: HostListener, args: ['click',] }]
    };
    return ModalBackdrop;
}(AbstractFdNgxClass));
export { ModalBackdrop };
if (false) {
    /** @type {?} */
    ModalBackdrop.prototype.backdropClass;
    /** @type {?} */
    ModalBackdrop.prototype.backdropClickCloseable;
    /** @type {?} */
    ModalBackdrop.prototype.overlayMain;
    /** @type {?} */
    ModalBackdrop.prototype.overlayModal;
    /**
     * @type {?}
     * @private
     */
    ModalBackdrop.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    ModalBackdrop.prototype.modalRef;
}
