// @fundamental-ngx/core@0.13.2/esm5/lib/modal/modal-utils/modal-container.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-utils/modal-container.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { modalFadeNgIf } from './modal-animations';
var ModalContainer = /** @class */ (function () {
    function ModalContainer() {
    }
    ModalContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fd-modal-container',
                    template: "",
                    host: {
                        '[@modal-fade]': '',
                        '[class.fd-modal-container]': 'true'
                    },
                    animations: [
                        modalFadeNgIf
                    ],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        .fd-modal-container {\n            position: fixed;\n            top: 0;\n            bottom: 0;\n            right: 0;\n            left: 0;\n            display: flex;\n            align-items: center;\n            justify-content: center;\n            z-index: 1000;\n        }\n    "]
                }] }
    ];
    return ModalContainer;
}());
export { ModalContainer };
