// @fundamental-ngx/core@0.13.2/esm5/lib/shellbar/shellbar-actions/shellbar-actions.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shellbar/shellbar-actions/shellbar-actions.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, HostListener, Input, QueryList, ViewEncapsulation, ContentChild, ViewChild, ChangeDetectionStrategy } from '@angular/core';
import { ShellbarActionComponent } from '../shellbar-action/shellbar-action.component';
import { ShellbarUserMenuComponent } from '../user-menu/shellbar-user-menu.component';
import { ComboboxComponent } from '../../combobox/combobox.component';
/**
 * The component that represents shellbar actions.
 * It is a container wrapper for all product actions and links (required element).
 * ```html
 * <fd-shellbar-actions [user]="user"
 *                      [userMenu]="userMenu"
 *                      [productSwitcher]="productSwitcher">
 *        <button fd-button [fdType]="'standard'">Custom Button</button>
 *
 *        <fd-shellbar-action *ngFor="let action of actions"
 *                            [glyph]="action.glyph"
 *                            [callback]="action.callback"
 *                            [label]="action.label"
 *                            [notificationCount]="action.notificationCount"
 *                            [notificationLabel]="action.notificationLabel">
 *        </fd-shellbar-action>
 * </fd-shellbar-actions>
 * ```
 */
var ShellbarActionsComponent = /** @class */ (function () {
    function ShellbarActionsComponent() {
        /**
         * @hidden
         */
        this.actionsCollapsed = false;
        /**
         * @hidden
         */
        this.showCollapsedProducts = false;
        /**
         * When set to true, popover list will be closed after selecting the option
         */
        this.closePopoverOnSelect = false;
        /**
         * Label for the collapsed item menu.
         */
        this.collapsedItemMenuLabel = 'Collapsed Item Menu';
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ShellbarActionsComponent.prototype.onResize = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.actionsCollapsed = window.innerWidth < 1024;
    };
    /**
     * @hidden
     */
    /**
     * @hidden
     * @param {?} item
     * @param {?} event
     * @return {?}
     */
    ShellbarActionsComponent.prototype.actionClicked = /**
     * @hidden
     * @param {?} item
     * @param {?} event
     * @return {?}
     */
    function (item, event) {
        this.triggerItems();
        item.callback(event);
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ShellbarActionsComponent.prototype.triggerItems = /**
     * @hidden
     * @return {?}
     */
    function () {
        if (this.closePopoverOnSelect) {
            if (this.userComponentView) {
                this.userComponentView.close();
            }
            if (this.userComponent) {
                this.userComponent.close();
            }
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ShellbarActionsComponent.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.onResize();
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    ShellbarActionsComponent.prototype.ngAfterContentChecked = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        this.totalNotifications = 0;
        this.shellbarActions.forEach((/**
         * @param {?} action
         * @return {?}
         */
        function (action) {
            if (action.notificationCount && typeof action.notificationCount === 'number') {
                _this.totalNotifications = _this.totalNotifications + action.notificationCount;
            }
        }));
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    ShellbarActionsComponent.prototype.toggleCollapsedProducts = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.showCollapsedProducts = !this.showCollapsedProducts;
    };
    Object.defineProperty(ShellbarActionsComponent.prototype, "userItem", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.userComponent) {
                return this.userComponent.user;
            }
            else {
                return this.user;
            }
        },
        enumerable: true,
        configurable: true
    });
    ShellbarActionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-shellbar-actions',
                    template: "<div *ngIf=\"comboboxComponent\" class=\"fd-shellbar__action fd-shellbar__action--desktop\">\n    <div class=\"fd-search-input\">\n        <ng-content select=\"fd-combobox\"></ng-content>\n    </div>\n</div>\n\n<div *ngIf=\"actionsCollapsed && shellbarActions.length\"\n        class=\"fd-shellbar__action fd-shellbar__action--mobile\">\n    <div class=\"fd-shellbar-collapse\">\n        <fd-popover [options]=\"{placement: 'bottom-end'}\"\n                    [focusTrapped]=\"false\">\n            <fd-popover-control>\n                <div class=\"fd-shellbar-collapse--control\">\n                    <button fd-button [glyph]=\"'overflow'\" class=\"fd-shellbar__button\"\n                            [attr.aria-label]=\"collapsedItemMenuLabel\">\n                        <span *ngIf=\"totalNotifications\"\n                                class=\"fd-counter fd-counter--notification\">{{totalNotifications}}</span>\n                    </button>\n                </div>\n            </fd-popover-control>\n            <fd-popover-body>\n                <fd-menu>\n                    <ul fd-menu-list\n                        *ngIf=\"!showCollapsedProducts\">\n                        <li fd-menu-item\n                            *ngFor=\"let action of shellbarActions\"\n                            (click)=\"actionClicked(action, $event)\">\n                            <fd-icon fd-menu-item-addon [glyph]=\"action.glyph\"></fd-icon>\n                            {{action.label}}\n                        </li>\n                    </ul>\n                    <ul fd-menu-list\n                        *ngIf=\"showCollapsedProducts\">\n                        <li (click)=\"toggleCollapsedProducts($event)\">\n                            <span fd-menu-item class=\"sap-icon--nav-back\"></span>\n                        </li>\n                    </ul>\n                </fd-menu>\n            </fd-popover-body>\n        </fd-popover>\n    </div>\n</div>\n<ng-content></ng-content>\n<ng-content select=\"fd-shellbar-action\"></ng-content>\n<ng-container *ngIf=\"userItem\">\n    <ng-container *ngIf=\"!userComponent\">\n        <fd-shellbar-user-menu (itemClicked)=\"triggerItems()\"\n                               [options]=\"{placement: 'bottom-end'}\"\n                               [focusTrapped]=\"false\"\n                               [user]=\"user\"\n                               [userMenu]=\"userMenu\"\n        ></fd-shellbar-user-menu>\n    </ng-container>\n    <ng-container *ngIf=\"userComponent\">\n        <ng-content select=\"fd-shellbar-user-menu\"></ng-content>\n    </ng-container>\n</ng-container>\n\n\n<ng-content select=\"fd-product-switch\"></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".fd-search-input--closed .fd-popover__popper{width:300px!important;left:-264px!important}fd-shellbar-actions{display:flex}"]
                }] }
    ];
    ShellbarActionsComponent.propDecorators = {
        user: [{ type: Input }],
        userMenu: [{ type: Input }],
        closePopoverOnSelect: [{ type: Input }],
        collapsedItemMenuLabel: [{ type: Input }],
        shellbarActions: [{ type: ContentChildren, args: [ShellbarActionComponent,] }],
        userComponent: [{ type: ContentChild, args: [ShellbarUserMenuComponent, { static: false },] }],
        userComponentView: [{ type: ViewChild, args: [ShellbarUserMenuComponent, { static: false },] }],
        comboboxComponent: [{ type: ContentChild, args: [ComboboxComponent, { static: false },] }],
        onResize: [{ type: HostListener, args: ['window:resize', [],] }]
    };
    return ShellbarActionsComponent;
}());
export { ShellbarActionsComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.actionsCollapsed;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.showCollapsedProducts;
    /**
     * The user data.
     * @type {?}
     */
    ShellbarActionsComponent.prototype.user;
    /**
     * The user menu data.
     * @type {?}
     */
    ShellbarActionsComponent.prototype.userMenu;
    /**
     * When set to true, popover list will be closed after selecting the option
     * @type {?}
     */
    ShellbarActionsComponent.prototype.closePopoverOnSelect;
    /**
     * Label for the collapsed item menu.
     * @type {?}
     */
    ShellbarActionsComponent.prototype.collapsedItemMenuLabel;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.shellbarActions;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.userComponent;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.userComponentView;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.comboboxComponent;
    /**
     * @hidden
     * @type {?}
     */
    ShellbarActionsComponent.prototype.totalNotifications;
}
