// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification-utils/default-notification/default-notification.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-utils/default-notification/default-notification.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var DefaultNotificationComponent = /** @class */ (function () {
    function DefaultNotificationComponent() {
    }
    /**
     * @return {?}
     */
    DefaultNotificationComponent.prototype.handleCloseButtonClick = /**
     * @return {?}
     */
    function () {
        if (this.defaultConfigurationNotification.closeButtonCallback) {
            this.defaultConfigurationNotification.closeButtonCallback();
        }
    };
    /**
     * @return {?}
     */
    DefaultNotificationComponent.prototype.handleApproveButtonClick = /**
     * @return {?}
     */
    function () {
        if (this.defaultConfigurationNotification.approveCallback) {
            this.defaultConfigurationNotification.approveCallback();
        }
    };
    /**
     * @return {?}
     */
    DefaultNotificationComponent.prototype.handleCancelButtonClick = /**
     * @return {?}
     */
    function () {
        if (this.defaultConfigurationNotification.cancelCallback) {
            this.defaultConfigurationNotification.cancelCallback();
        }
    };
    DefaultNotificationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-default-notification',
                    template: "<fd-notification-header [closeButton]=\"!!defaultConfigurationNotification?.closeButtonCallback\"\n                        (closeButtonClick)=\"handleCloseButtonClick()\" [type]=\"type\">\n    <h3 fd-notification-title *ngIf=\"defaultConfigurationNotification?.title\">\n        {{defaultConfigurationNotification?.title}}\n    </h3>\n</fd-notification-header>\n<fd-notification-body>\n    <div fd-notification-content>\n        <div fd-notification-avatar *ngIf=\"defaultConfigurationNotification?.avatar\">\n            <ng-container *ngTemplateOutlet=\"defaultConfigurationNotification?.avatar\"></ng-container>\n        </div>\n        <div fd-notification-text>\n            <div fd-notification-description *ngIf=\"defaultConfigurationNotification?.description\">\n                {{defaultConfigurationNotification?.description}}\n            </div>\n            <div fd-notification-metadata *ngIf=\"defaultConfigurationNotification?.metadata\">\n                {{defaultConfigurationNotification?.metadata}}\n            </div>\n        </div>\n    </div>\n    <fd-notification-footer>\n        <button fd-button [options]=\"'light'\" *ngIf=\"defaultConfigurationNotification?.moreInfo\">\n            {{defaultConfigurationNotification?.moreInfo}}\n        </button>\n        <div fd-notification-actions>\n            <button fd-button [fdType]=\"'positive'\"\n                    *ngIf=\"defaultConfigurationNotification?.approve\"\n                    (click)=\"handleApproveButtonClick()\">\n                {{defaultConfigurationNotification?.approve}}\n            </button>\n            <button fd-button [fdType]=\"'negative'\"\n                    *ngIf=\"defaultConfigurationNotification?.cancel\"\n                    (click)=\"handleCancelButtonClick()\">\n                {{defaultConfigurationNotification?.cancel}}\n            </button>\n        </div>\n    </fd-notification-footer>\n</fd-notification-body>\n"
                }] }
    ];
    return DefaultNotificationComponent;
}());
export { DefaultNotificationComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    DefaultNotificationComponent.prototype.defaultConfigurationNotification;
    /**
     *
     * @type {?}
     */
    DefaultNotificationComponent.prototype.type;
}
