// @fundamental-ngx/core@0.13.2/esm5/lib/datetime-picker/models/fd-datetime.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datetime-picker/models/fd-datetime.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Default date with time model used by the fundamental components.
 */
import { FdDate } from '../../calendar/models/fd-date';
var FdDatetime = /** @class */ (function () {
    /**
     * Constructor to build a FdDateTime object from a FdDate and TimeObject.
     * @param date the FdDate object.
     * @param time the TimeObject object.
     */
    function FdDatetime(date, time) {
        this.date = date;
        this.time = time;
    }
    /**
     * Static function to get the current date in FdDateTime form.
     */
    /**
     * Static function to get the current date in FdDateTime form.
     * @return {?}
     */
    FdDatetime.getToday = /**
     * Static function to get the current date in FdDateTime form.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = new Date();
        /** @type {?} */
        var time = { hour: date.getHours(), minute: date.getMinutes(), second: date.getSeconds() };
        return new FdDatetime(FdDate.getToday(), time);
    };
    /**
     * Get Luxon date object converted to string from FdDate.
     */
    /**
     * Get Luxon date object converted to string from FdDate.
     * @return {?}
     */
    FdDatetime.prototype.toLocaleDateString = /**
     * Get Luxon date object converted to string from FdDate.
     * @return {?}
     */
    function () {
        if (this.toDate() && this.isTimeValid() && this.isDateValid()) {
            return this.toDate().toLocaleString();
        }
        else {
            return null;
        }
    };
    /**
     * Method that checks validity of time on FdDateTime object.
     */
    /**
     * Method that checks validity of time on FdDateTime object.
     * @return {?}
     */
    FdDatetime.prototype.isTimeValid = /**
     * Method that checks validity of time on FdDateTime object.
     * @return {?}
     */
    function () {
        if (!this.time ||
            this.hour === undefined ||
            this.minute === undefined ||
            this.second === undefined ||
            isNaN(this.hour) ||
            isNaN(this.minute) ||
            isNaN(this.second)) {
            return false;
        }
        if (this.hour > 23 || this.hour < 0) {
            return false;
        }
        if (this.minute > 59 || this.minute < 0) {
            return false;
        }
        if (this.second > 59 || this.second < 0) {
            return false;
        }
        return true;
    };
    /**
     * Method that checks validity of date on FdDateTime object.
     */
    /**
     * Method that checks validity of date on FdDateTime object.
     * @return {?}
     */
    FdDatetime.prototype.isDateValid = /**
     * Method that checks validity of date on FdDateTime object.
     * @return {?}
     */
    function () {
        return this.date && this.date.isDateValid();
    };
    Object.defineProperty(FdDatetime.prototype, "year", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.date) {
                return this.date.year;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FdDatetime.prototype, "month", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.date) {
                return this.date.month;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FdDatetime.prototype, "day", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.date) {
                return this.date.day;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FdDatetime.prototype, "hour", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.time) {
                return this.time.hour;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FdDatetime.prototype, "minute", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.time) {
                return this.time.minute;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FdDatetime.prototype, "second", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.time) {
                return this.time.second;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Get native date object from FdDate.
     */
    /**
     * Get native date object from FdDate.
     * @return {?}
     */
    FdDatetime.prototype.toDate = /**
     * Get native date object from FdDate.
     * @return {?}
     */
    function () {
        return new Date(this.year, this.month - 1, this.day, this.hour, this.minute, this.second);
    };
    return FdDatetime;
}());
export { FdDatetime };
if (false) {
    /** @type {?} */
    FdDatetime.prototype.date;
    /** @type {?} */
    FdDatetime.prototype.time;
}
