// @fundamental-ngx/core@0.13.2/esm2015/lib/menu/menu-item.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/menu/menu-item.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding } from '@angular/core';
/**
 * The directive that represents a menu item.
 */
export class MenuItemDirective {
    /**
     * @hidden
     * @param {?} itemEl
     */
    constructor(itemEl) {
        this.itemEl = itemEl;
        /**
         * @hidden
         */
        this.fdMenuItemClass = true;
    }
    /**
     * @return {?}
     */
    focus() {
        this.itemEl.nativeElement.focus();
    }
    /**
     * @return {?}
     */
    click() {
        this.itemEl.nativeElement.click();
    }
}
MenuItemDirective.decorators = [
    { type: Directive, args: [{
                // TODO to be discussed
                // tslint:disable-next-line:directive-selector
                selector: '[fd-menu-item]',
            },] }
];
/** @nocollapse */
MenuItemDirective.ctorParameters = () => [
    { type: ElementRef }
];
MenuItemDirective.propDecorators = {
    fdMenuItemClass: [{ type: HostBinding, args: ['class.fd-menu__item',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MenuItemDirective.prototype.fdMenuItemClass;
    /** @type {?} */
    MenuItemDirective.prototype.itemEl;
}
