// @fundamental-ngx/core@0.13.2/esm5/lib/shellbar/user-menu/shellbar-user-menu.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shellbar/user-menu/shellbar-user-menu.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { PopoverComponent } from '../../popover/popover.component';
/**
 * This Component extends popover component and passes all of the options and events from outside to popover component
 * and Vice Versa
 *
 */
var ShellbarUserMenuComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ShellbarUserMenuComponent, _super);
    function ShellbarUserMenuComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Event emitted on item click
         */
        _this.itemClicked = new EventEmitter();
        return _this;
    }
    /**
     * @hidden
     */
    /**
     * @hidden
     * @param {?} item
     * @param {?} event
     * @return {?}
     */
    ShellbarUserMenuComponent.prototype.itemClick = /**
     * @hidden
     * @param {?} item
     * @param {?} event
     * @return {?}
     */
    function (item, event) {
        this.itemClicked.emit();
        if (item.callback) {
            item.callback(event);
        }
    };
    ShellbarUserMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-shellbar-user-menu',
                    template: "<div class=\"fd-shellbar__action fd-shellbar__action--show-always\" *ngIf=\"user\">\n    <div class=\"fd-user-menu\">\n        <fd-popover [options]=\"options || {placement: 'bottom-end'}\"\n                    [closeOnEscapeKey]=\"closeOnEscapeKey\"\n                    [closeOnOutsideClick]=\"closeOnOutsideClick\"\n                    [(isOpen)]=\"isOpen\"\n                    [triggers]=\"triggers\"\n                    (isOpenChange)=\"openChanged($event)\"\n                    [disabled]=\"disabled\"\n                    [isDropdown]=\"isDropdown\"\n                    [focusTrapped]=\"false\">\n            <fd-popover-control>\n                <div class=\"fd-user-menu__control\">\n                    <span fd-identifier *ngIf=\"!user.image\" [colorAccent]=\"user.colorAccent\" [circle]=\"true\" [size]=\"'xs'\">\n                        {{user.initials}}\n                    </span>\n                    <span fd-identifier *ngIf=\"user.image\" [colorAccent]=\"user.colorAccent\" [circle]=\"true\" [size]=\"'xs'\"\n                          [ngStyle]=\"{'background-image': 'url(' + user.image + ')'}\">\n                    </span>\n                </div>\n            </fd-popover-control>\n            <fd-popover-body>\n                <fd-menu>\n                    <ul fd-menu-list>\n                        <li fd-menu-item\n                            *ngFor=\"let item of userMenu\"\n                            (click)=\"itemClick(item, $event)\">\n                            {{item.text}}\n                        </li>\n                    </ul>\n                </fd-menu>\n            </fd-popover-body>\n        </fd-popover>\n    </div>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    ShellbarUserMenuComponent.propDecorators = {
        user: [{ type: Input }],
        userMenu: [{ type: Input }],
        itemClicked: [{ type: Output }]
    };
    return ShellbarUserMenuComponent;
}(PopoverComponent));
export { ShellbarUserMenuComponent };
if (false) {
    /**
     * The user data.
     * @type {?}
     */
    ShellbarUserMenuComponent.prototype.user;
    /**
     * The user menu data.
     * @type {?}
     */
    ShellbarUserMenuComponent.prototype.userMenu;
    /**
     * Event emitted on item click
     * @type {?}
     */
    ShellbarUserMenuComponent.prototype.itemClicked;
}
