// @fundamental-ngx/core@0.13.2/esm5/lib/localizator-editor/localization-editor-main/localization-editor-main.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/localizator-editor/localization-editor-main/localization-editor-main.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Input, Component, ViewEncapsulation } from '@angular/core';
import { LocalizationEditorItemComponent } from '../localization-editor-item/localization-editor-item.component';
/**
 *  Component that represents the field which is always visible and is rendered outside the popover.
 *  ```html
 *  <fd-localization-editor-main [label]="'EN'">
 *       <input fd-localization-editor-input type="text" placeholder="EN">
 *  </fd-localization-editor-main>
 *  ```
 */
var LocalizationEditorMainComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LocalizationEditorMainComponent, _super);
    function LocalizationEditorMainComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Whether AddOn Button should be focusable, set to true by default
         */
        _this.buttonFocusable = true;
        return _this;
    }
    LocalizationEditorMainComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-localization-editor-main',
                    template: "<fd-input-group [compact]=\"compact\"\n                [state]=\"state\"\n                [inputTemplate]=\"input\">\n    <span fd-input-group-addon\n          [button]=\"true\"\n          [compact]=\"compact\"\n          [state]=\"state\"\n          [type]=\"type\">\n        <button class=\"fd-localization-editor__button\"\n                fd-button aria-haspopup=\"true\"\n                [attr.tabindex]=\"buttonFocusable ? 0 : -1\"\n                [options]=\"'light'\"\n                [compact]=\"compact\"\n                [attr.aria-expanded]=\"expanded\">\n            <ng-container *ngIf=\"labelTemplate\">\n                <ng-container [fd-localization-editor-load-label]=\"labelTemplate\"></ng-container>\n            </ng-container>\n            <ng-container *ngIf=\"!labelTemplate\">\n                {{label}}\n            </ng-container>\n        </button>\n    </span>\n</fd-input-group>\n<ng-template #input>\n    <ng-content select=\"[fd-localization-editor-input]\"></ng-content>\n    <ng-content select=\"[fd-localization-editor-textarea]\"></ng-content>\n</ng-template>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    LocalizationEditorMainComponent.propDecorators = {
        state: [{ type: Input }],
        buttonFocusable: [{ type: Input }]
    };
    return LocalizationEditorMainComponent;
}(LocalizationEditorItemComponent));
export { LocalizationEditorMainComponent };
if (false) {
    /**
     * @hidden
     * This variable is controlled by parent component
     *
     * @type {?}
     */
    LocalizationEditorMainComponent.prototype.expanded;
    /**
     *  The state of the form control - applies css classes.
     *  Can be `valid`, `invalid`, `warning`, `information` or blank for default.
     * @type {?}
     */
    LocalizationEditorMainComponent.prototype.state;
    /**
     * Whether AddOn Button should be focusable, set to true by default
     * @type {?}
     */
    LocalizationEditorMainComponent.prototype.buttonFocusable;
}
