// @fundamental-ngx/core@0.13.2/esm5/lib/notification/notification-utils/notification-container.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/notification/notification-utils/notification-container.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
var NotificationContainer = /** @class */ (function () {
    function NotificationContainer() {
    }
    NotificationContainer.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification-container',
                    template: "",
                    host: {
                        '[class.fd-notification-container]': 'true'
                    },
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        .fd-notification-container {\n            position: fixed;\n            display: flex;\n            flex-direction: column;\n            z-index: 5000;\n            align-items: center;\n            top: 1rem;\n            right: 1rem;\n        }\n    "]
                }] }
    ];
    return NotificationContainer;
}());
export { NotificationContainer };
